% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/gen_dist.R
\name{gen_dist}
\alias{gen_dist}
\title{Calculate genetic distances}
\usage{
gen_dist(
  gen = NULL,
  dist_type = "euclidean",
  plink_file = NULL,
  plink_id_file = NULL,
  npc_selection = "auto",
  criticalpoint = 2.0234
)
}
\arguments{
\item{gen}{path to vcf file, a \code{vcfR} type object, or a dosage matrix}

\item{dist_type}{the type of genetic distance to calculate (options: \code{"euclidean"} (default), \code{"bray_curtis"}, \code{"dps"} for proportion of shared alleles (requires vcf), \code{"plink"}, or \code{"pc"} for PC-based)}

\item{plink_file}{if \code{"plink"} dist_type is used, path to plink distance file (typically ".dist"; required only for calculating plink distance)}

\item{plink_id_file}{if \code{"plink"} dist_type is used, path to plink id file (typically ".dist.id"; required only for calculating plink distance)}

\item{npc_selection}{if \code{dist_type = "pc"}, how to perform K selection (options: \code{"auto"} for automatic selection based on significant eigenvalues from Tracy-Widom test (default), or \code{"manual"} to examine PC screeplot and enter no. PCs into console)}

\item{criticalpoint}{if \code{dist_type = "pc"} used with \code{npc_selection = "auto"}, the critical point for the significance threshold for the Tracy-Widom test within the PCA (defaults to 2.0234 which corresponds to an alpha of 0.01)}
}
\value{
pairwise distance matrix for given distance metric
}
\description{
Calculate genetic distances
}
\details{
Euclidean and Bray-Curtis distances calculated using the ecodist package: Goslee, S.C. and Urban, D.L. 2007. The ecodist package for dissimilarity-based analysis of ecological data. Journal of Statistical Software 22(7):1-19. DOI:10.18637/jss.v022.i07.
Proportions of shared alleles calculated using the adegenet package: Jombart T. and Ahmed I. (2011) adegenet 1.3-1: new tools for the analysis of genome-wide SNP data. Bioinformatics. doi:10.1093/bioinformatics/btr521.
For calculating proportions of shared alleles, missing values are ignored (i.e., prop shared alleles calculated from present values; no scaling performed)
}
