<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A Landscape Genomic Analysis Toolkit in R • algatr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A Landscape Genomic Analysis Toolkit in R">
<meta name="description" content="What the package does (one paragraph).">
<meta property="og:description" content="What the package does (one paragraph).">
<meta property="og:image" content="https://thewanglab.github.io/algatr/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">algatr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference Manual</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="articles/data_processing_vignette.html">Genetic data processing</a></li>
    <li><a class="dropdown-item" href="articles/enviro_data_vignette.html">Environmental data processing</a></li>
    <li><a class="dropdown-item" href="articles/masking_vignette.html">Masking</a></li>
    <li><a class="dropdown-item" href="articles/gen_dist_vignette.html">Genetic distances</a></li>
    <li><a class="dropdown-item" href="articles/TESS_vignette.html">TESS</a></li>
    <li><a class="dropdown-item" href="articles/MMRR_vignette.html">MMRR</a></li>
    <li><a class="dropdown-item" href="articles/GDM_vignette.html">GDM</a></li>
    <li><a class="dropdown-item" href="articles/RDA_vignette.html">RDA</a></li>
    <li><a class="dropdown-item" href="articles/LFMM_vignette.html">LFMM</a></li>
    <li><a class="dropdown-item" href="articles/wingen_vignette.html">Wingen</a></li>
    <li><a class="dropdown-item" href="articles/alazygatr_vignette.html">Alazygatr</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="algatr-">algatr <a class="anchor" aria-label="anchor" href="#algatr-"></a>
</h1>
</div>
<p><a href="https://github.com/TheWangLab/algatr/actions/workflows/build-test.yml" class="external-link"><img src="https://github.com/TheWangLab/algatr/actions/workflows/build-test.yml/badge.svg?branch=main" alt="build"></a> <a href="https://img.shields.io/badge/license-MIT-blue" class="external-link"><img src="https://img.shields.io/badge/license-MIT-blue" alt="license: MIT"></a> <a href="https://zenodo.org/doi/10.5281/zenodo.10535497" class="external-link"><img src="https://zenodo.org/badge/523130808.svg" alt="Zenodo DOI"></a></p>
<p><strong>A</strong> <strong>L</strong>andscape <strong>G</strong>enomic <strong>A</strong>nalysis <strong>T</strong>oolkit in <strong>R</strong> (<strong>algatr</strong>) was built to provide researchers with a step-by-step, start-to-finish pipeline to perform core landscape genomic analysis methods with their data.</p>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<hr>
<p>Please cite our paper if you use this package:</p>
<p><strong>Chambers, E.A., Bishop, A.P., &amp; Wang, I.J. (2023). Individual-based landscape genomics for conservation: An analysis pipeline. <em>Molecular Ecology Resources.</em><a href="https://doi.org/10.1111/1755-0998.13884" class="external-link">https://doi.org/10.1111/1755-0998.13884.</a></strong></p>
<p>Because algatr makes use of existing software, please also be sure to cite the papers for the relevant corresponding method:</p>
<table class="table">
<colgroup>
<col width="7%">
<col width="22%">
<col width="70%">
</colgroup>
<tbody>
<tr>
<td><code>wingen</code></td>
<td>Bishop et al. (2023)</td>
<td>DOI: <a href="https://doi.org/10.1111/2041-210X.14090" class="external-link">10.1111/2041-210X.14090</a>
</td>
</tr>
<tr>
<td><code>TESS</code></td>
<td>Caye et al. (2016)</td>
<td>DOI: <a href="https://doi.org/10.1111/1755-0998.12471" class="external-link">10.1111/1755-0998.12471</a>
</td>
</tr>
<tr>
<td><code>MMRR</code></td>
<td>Wang (2013)</td>
<td>DOI: <a href="https://doi.org/10.1111/evo.12134" class="external-link">10.1111/evo.12134</a>
</td>
</tr>
<tr>
<td><code>GDM</code></td>
<td>Ferrier et al. (2007)</td>
<td>DOI: <a href="https://doi.org/10.1111/j.1472-4642.2007.00341.x" class="external-link">10.1111/j.1472-4642.2007.00341.x</a>
</td>
</tr>
<tr>
<td></td>
<td>Fitzpatrick et al. (2022)</td>
<td>gdm: Generalized Dissimilarity Modeling. R package version 1.5.0-9.1.</td>
</tr>
<tr>
<td><code>RDA</code></td>
<td>Capblancq &amp; Forester (2021)</td>
<td>DOI: <a href="https://doi.org/10.1111/2041-210X.13722" class="external-link">10.1111/2041-210X.13722</a>
</td>
</tr>
<tr>
<td><code>LFMM</code></td>
<td>Caye et al. (2019)</td>
<td>DOI: <a href="https://doi.org/10.1093/molbev/msz008" class="external-link">10.1093/molbev/msz008</a>
</td>
</tr>
<tr>
<td></td>
<td>Jumentier (2021)</td>
<td>lfmm: Latent Factor Mixed Models. R package version 1.1.</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<hr>
<p>algatr can be installed with the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"TheWangLab/algatr"</span><span class="op">)</span></span></code></pre></div>
<p>The algatr package depends on many packages for all of the different methods implemented. To ensure that algatr can still be installed even if one of the many dependencies breaks, the packages required for each method are suggested rather than imported. To install all of the suggested packages set <code>dependencies = TRUE</code> when running <code>install_github()</code>. To install just the packages required for each method, use the <code>*_packages()</code> functions below. This is helpful if you’re only interested in using a subset of the methods provided and don’t want to install unnecessary packages.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install algatr</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"TheWangLab/algatr"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thewanglab.github.io/algatr/">algatr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Option 1: Install all of the packages for algatr</span></span>
<span><span class="fu"><a href="reference/alazygatr_packages.html">alazygatr_packages</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Option 2: Install subsets of packages based on what methods you want to use:</span></span>
<span><span class="co">## Genetic distance processing:</span></span>
<span><span class="fu"><a href="reference/gen_dist_packages.html">gen_dist_packages</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## Genetic data processing:</span></span>
<span><span class="fu"><a href="reference/data_processing_packages.html">data_processing_packages</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## Environmental and geographic data processing:</span></span>
<span><span class="fu"><a href="reference/envirodata_packages.html">envirodata_packages</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## LFMM:</span></span>
<span><span class="fu"><a href="reference/lfmm_packages.html">lfmm_packages</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## RDA:</span></span>
<span><span class="fu"><a href="reference/rda_packages.html">rda_packages</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## MMRR:</span></span>
<span><span class="fu"><a href="reference/mmrr_packages.html">mmrr_packages</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## GDM:</span></span>
<span><span class="fu"><a href="reference/gdm_packages.html">gdm_packages</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## TESS:</span></span>
<span><span class="fu"><a href="reference/tess_packages.html">tess_packages</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## wingen:</span></span>
<span><span class="fu"><a href="reference/wingen_packages.html">wingen_packages</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>If you’re installing on Ubuntu, you may run into issues installing the rmapshaper package; scroll to the bottom of the README for more information.</p>
<div class="section level3">
<h3 id="docker">Docker<a class="anchor" aria-label="anchor" href="#docker"></a>
</h3>
<p>algatr can be run using <a href="https://docs.docker.com/get-started/" class="external-link">Docker</a>, in which case prior installation of package dependencies is not required. First, install Docker, and then start algatr within a Docker container:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">--rm</span> <span class="at">-ti</span> ghcr.io/thewanglab/algatr</span></code></pre></div>
<p>You can also run the container in an RStudio instance:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">--rm</span> <span class="at">-ti</span> <span class="at">-e</span> PASSWORD=yourpassword <span class="at">-p</span> 8787:8787 ghcr.io/thewanglab/algatr</span></code></pre></div>
<p>Open localhost:8787 in your browser and log in with username:rstudio and password:yourpassword (substitute yourpassword for whatever password you would like)</p>
</div>
<div class="section level3">
<h3 id="codespaces">Codespaces<a class="anchor" aria-label="anchor" href="#codespaces"></a>
</h3>
<p>You can also try out algatr using GitHub codespaces — a cloud-based development environment with everything you need installed and ready to go. To start a codespace, click the green “Code” button at the top of the repository and select “Open with Codespaces”. This will open a new codespace with the algatr repository loaded. If you want to make changes to any of the files, you will need to fork the repository to save them. GitHub Free personal accounts automatically come with 120 codespace core hours per month free. For more information about codespaces see <a href="https://docs.github.com/en/codespaces/overview" class="external-link">here</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr>
<p>Landscape genetics (or genomics) combines the fields of landscape ecology and population genetics to understand how environmental variation affects spatial genetic variation. Thus, at its most basic, any landscape genetics method requires genetic and environmental data as input.</p>
<p>The methods contained within algatr will perform analyses on:</p>
<ul>
<li><p>Genomic-scale datasets (i.e., those generated using reduced representation or whole genome sequencing)</p></li>
<li><p>Datasets with individual-based sampling (<em>it will also work on population-based sampling schemes</em>)</p></li>
</ul>
<p>algatr makes use of several existing packages and methods, and we provide citations to these packages (and corresponding publications) whenever possible. Many of these packages have extensive documentation and excellent additional resources, which we provide links to in the corresponding vignettes. We have added functionality to each of these methods within algatr which we discuss in each of the methods’ vignettes.</p>
<p>Other than input data processing functions, the main functions within algatr are named with the pattern <code>[method]_do_everything()</code>. As the name implies, these functions will take you from your raw input data through to generating results, tables, and figures from the analysis. For example, <code><a href="reference/gdm_do_everything.html">gdm_do_everything()</a></code> will run generalized dissimilarity modeling while also generating a GDM map, a table with results, and several other outputs. All of the <code>[method]_do_everything</code> functions has a <code>quiet</code> argument which, when set to <code>"TRUE"</code>, will not automatically print figures and outputs.</p>
<p>To better understand what’s going on under the hood of these <code>[method]_do_everything()</code> functions, the algatr vignettes provide a line-by-line breakdown of the individual user-facing functions contained within the <code>[method]_do_everything()</code> function to (a) increase a user’s understanding of how the method actually works, and (b) allow users with more customizability in how they run their own analysis, if so desired. <strong>We strongly encourage all researchers to only use the</strong> <code>[method]_do_everything()</code> <strong>functions as an initial first-pass examination of their data; users should follow the workflows provided in the vignettes for more nuanced parameter control and to generally better understand how each of these methods works.</strong></p>
<p>When deciding on methods to have within algatr, we found it best to first identify the questions that these methods seek to answer, and we think this is a good framework for anyone (particularly beginner landscape genomicists) to think about landscape genomic methods. These questions fall into four broad categories of analyses:</p>
<table style="width:99%;" class="table">
<colgroup>
<col style="width: 17%">
<col style="width: 16%">
<col style="width: 32%">
<col style="width: 24%">
<col style="width: 7%">
</colgroup>
<thead><tr class="header">
<th>
Question
</th>
<th>
Category
</th>
<th>
Method
</th>
<th>
Vignette
</th>
<th>
algatr function
</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
How do we delineate population units for management?
</td>
<td>
Population structure
</td>
<td>
TESS (Caye et al. 2016)
</td>
<td>
<a href="https://thewanglab.github.io/algatr/articles/TESS_vignette.html">TESS vignette</a>
</td>
<td>
<code><a href="reference/tess_do_everything.html">tess_do_everything()</a></code>
</td>
</tr>
<tr class="even">
<td>
How is genetic variation distributed?
</td>
<td>
Genetic diversity
</td>
<td>
Moving windows of genetic diversity; wingen (Bishop et al. 2023)
</td>
<td>
<a href="https://thewanglab.github.io/algatr/articles/wingen_vignette.html">wingen vignette</a>
</td>
<td>
<code><a href="reference/wingen_do_everything.html">wingen_do_everything()</a></code>
</td>
</tr>
<tr class="odd">
<td>
What are the drivers of population connectivity?
</td>
<td>
Isolation by distance/isolation by environment (IBD/IBE)
</td>
<td>
<p>
Multiple matrix regression with randomization; MMRR (Wang 2013)
</p>
<p>
Generalized dissimilarity modeling; GDM (Ferrier et al. 2007; Freedman et al. 2010; Fitzpatrick &amp; Keller 2015)
</p>
</td>
<td>
<p>
<a href="https://thewanglab.github.io/algatr/articles/MMRR_vignette.html">MMRR vignette</a>
</p>
<p>
<a href="https://thewanglab.github.io/algatr/articles/GDM_vignette.html">GDM vignette</a>
</p>
</td>
<td>
<p>
<code><a href="reference/mmrr_do_everything.html">mmrr_do_everything()</a></code>
</p>
<p>
<code><a href="reference/gdm_do_everything.html">gdm_do_everything()</a></code>
</p>
</td>
</tr>
<tr class="even">
<td>
How can we identify and protect adaptive genetic variation?
</td>
<td>
Genotype-environment associations (GEA)
</td>
<td>
<p>
Redundancy analysis; RDA (Capblancq &amp; Forester 2021)
</p>
<p>
Latent factor mixed models; LFMM (Caye et al. 2019)
</p>
</td>
<td>
<p>
<a href="https://thewanglab.github.io/algatr/articles/RDA_vignette.html">RDA vignette</a>
</p>
<p>
<a href="https://thewanglab.github.io/algatr/articles/LFMM_vignette.html">LFMM vignette</a>
</p>
</td>
<td>
<p>
<code><a href="reference/rda_do_everything.html">rda_do_everything()</a></code>
</p>
<p>
<code><a href="reference/lfmm_do_everything.html">lfmm_do_everything()</a></code>
</p>
</td>
</tr>
</tbody>
</table>
<p>You can also view the knitted versions of this README and all of algatr’s vignettes on the <strong>package’s pkgdown website <a href="https://thewanglab.github.io/algatr/index.html">here</a>.</strong></p>
<div class="section level3">
<h3 id="the-example-dataset">The example dataset<a class="anchor" aria-label="anchor" href="#the-example-dataset"></a>
</h3>
<hr>
<p>As an example for the algatr vignettes, we’ll be using the <em>Sceloporus</em> RADseq dataset from <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/mec.15836" class="external-link">Bouzid et al. 2022</a>. Although the original dataset contains &gt;6K SNPs and 108 individuals across Western North America, we’ve pruned the dataset down to 1,000 SNPs and only those individuals collected in California (53 individuals) to make all analyses run a bit faster. This dataset also makes for a nice test dataset because the 53 individuals were collected from 53 separate localities (i.e., individual-based sampling was used) across the state the California.</p>
<p>There are four objects loaded within the example dataset:</p>
<ul>
<li><p><code>liz_coords</code>: sampling coordinates for 53 individuals (from 53 localities)</p></li>
<li><p><code>liz_vcf</code>: the vcfR object containing variant information</p></li>
<li><p><code>liz_gendist</code>: a matrix of genetic distances generated from the vcf file (<em>distances were calculated using Plink</em>)</p></li>
<li><p><code>CA_env</code>: a RasterStack object with three PC environmental layers</p></li>
</ul>
<p>Load the example dataset to take a look:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/load_algatr_example.html">load_algatr_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ---------------- example dataset ----------------</span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; Objects loaded: </span></span>
<span><span class="co">#&gt; *liz_vcf* vcfR object (1000 loci x 53 samples) </span></span>
<span><span class="co">#&gt; *liz_gendist* genetic distance matrix (Plink Distance) </span></span>
<span><span class="co">#&gt; *liz_coords* dataframe with x and y coordinates </span></span>
<span><span class="co">#&gt; *CA_env* RasterStack with example environmental layers </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span>
<span><span class="co">#&gt; </span></span></code></pre></div>
<p>Let’s take a look at the environmental layers included in the example dataset. These were generated by performing a raster PCA on 19 bioclimatic variables (obtained from the World Clim database) and retaining the top 3 PCs. Let’s take a look at the rasters:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">CA_env</span>, col <span class="op">=</span> <span class="fu">turbo</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, axes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-rasters-1.png" width="100%"></p>
<p>We can combine all three PCs into a single map by scaling each of the rasters such that they each correspond to either R, G, or B using the <code><a href="reference/scaleRGB.html">scaleRGB()</a></code> function and subsequently map using the <code>plotRGB()</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">env</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/scaleRGB.html">scaleRGB</a></span><span class="op">(</span><span class="va">CA_env</span><span class="op">)</span></span>
<span><span class="fu">plotRGB</span><span class="op">(</span><span class="va">env</span>, r <span class="op">=</span> <span class="fl">1</span>, g <span class="op">=</span> <span class="fl">2</span>, b <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add sampling localities on top of this</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="va">liz_coords</span>, pch <span class="op">=</span> <span class="fl">19</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-RGB-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="your-ngs-data">Your NGS data<a class="anchor" aria-label="anchor" href="#your-ngs-data"></a>
</h3>
<hr>
<p>To generate the above files for your own dataset, you’ll need the following:</p>
<ul>
<li><p>Sampling coordinates (<em>always in longitude, latitude [x,y] order; refer to <code>liz_coords</code> for formatting</em>)</p></li>
<li><p>Genetic data in vcf file format (<em>this is the most standard file format for reduced representation or whole genome sequencing data</em>)</p></li>
<li><p>Environmental data layers of your choice</p></li>
</ul>
<p><strong>Be sure that the ordering of your individuals across your coordinate and genetic data files are consistent!</strong></p>
</div>
<div class="section level3">
<h3 id="next-steps">Next steps<a class="anchor" aria-label="anchor" href="#next-steps"></a>
</h3>
<p>Read in your vcf file using the <code>read.vcfR()</code> function in the vcfR package. To see how this works, we can load the example dataset vcf like so:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu">read.vcfR</span><span class="op">(</span><span class="fu">here</span><span class="op">(</span><span class="st">"inst"</span>, <span class="st">"extdata"</span>, <span class="st">"liz_test.vcf"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scanning file to determine attributes.</span></span>
<span><span class="co">#&gt; File attributes:</span></span>
<span><span class="co">#&gt;   meta lines: 6</span></span>
<span><span class="co">#&gt;   header_line: 7</span></span>
<span><span class="co">#&gt;   variant count: 1000</span></span>
<span><span class="co">#&gt;   column count: 62</span></span>
<span><span class="co">#&gt; Meta line 6 read in.</span></span>
<span><span class="co">#&gt; All meta lines processed.</span></span>
<span><span class="co">#&gt; gt matrix initialized.</span></span>
<span><span class="co">#&gt; Character matrix gt created.</span></span>
<span><span class="co">#&gt;   Character matrix gt rows: 1000</span></span>
<span><span class="co">#&gt;   Character matrix gt cols: 62</span></span>
<span><span class="co">#&gt;   skip: 0</span></span>
<span><span class="co">#&gt;   nrows: 1000</span></span>
<span><span class="co">#&gt;   row_num: 0</span></span>
<span><span class="co">#&gt; Processed variant 1000Processed variant: 1000</span></span>
<span><span class="co">#&gt; All variants processed</span></span></code></pre></div>
<p>You’ll now want to do some processing of these data, such as file conversions and LD-pruning (see the <a href="https://thewanglab.github.io/algatr/articles/data_processing_vignette.html"><strong>data processing vignette</strong></a>), environmental data checking and manipulation (see the <a href="https://thewanglab.github.io/algatr/articles/enviro_data_vignette.html"><strong>environmental data processing vignette</strong></a>), and calculating genetic distances (see the <a href="https://thewanglab.github.io/algatr/articles/gen_dist_vignette.html"><strong>genetic distances vignette</strong></a>).</p>
<table style="width:98%;" class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 38%">
<col style="width: 15%">
<col style="width: 25%">
<col style="width: 12%">
</colgroup>
<thead><tr class="header">
<th>
</th>
<th>
R package
</th>
<th>
algatr function
</th>
<th>
Input files
</th>
<th>
Required arguments
</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
TESS
</td>
<td>
<a href="https://bcm-uga.github.io/TESS3_encho_sen/index.html" class="external-link">tess3r</a>
</td>
<td>
<code><a href="reference/tess_do_everything.html">tess_do_everything()</a></code>
</td>
<td>
<ul>
<li>
Genotype dosage matrix
</li>
<li>
Environmental layers
</li>
<li>
Sampling coordinates
</li>
</ul>
</td>
<td>
<ul>
<li>
<p>
<code>gen</code>
</p>
</li>
<li>
<p>
<code>coords</code>
</p>
</li>
<li>
<p>
<code>Kvals</code>
</p>
</li>
</ul>
</td>
</tr>
<tr class="even">
<td>
MMRR
</td>
<td>
algatr
</td>
<td>
<code><a href="reference/mmrr_do_everything.html">mmrr_do_everything()</a></code>
</td>
<td>
<ul>
<li>
<p>
Genetic distance matrix
</p>
</li>
<li>
<p>
Environmental layers
</p>
</li>
<li>
<p>
Sampling coordinates
</p>
</li>
</ul>
</td>
<td>
<ul>
<li>
<p>
<code>gendist</code>
</p>
</li>
<li>
<p>
<code>coords</code>
</p>
</li>
<li>
<p>
<code>env</code>
</p>
</li>
</ul>
</td>
</tr>
<tr class="odd">
<td>
GDM
</td>
<td>
<a href="https://cran.r-project.org/web/packages/gdm/index.html" class="external-link">gdm</a>
</td>
<td>
<code><a href="reference/gdm_do_everything.html">gdm_do_everything()</a></code>
</td>
<td>
<ul>
<li>
Genetic distance matrix
</li>
<li>
Environmental layers
</li>
<li>
Sampling coordinates
</li>
</ul>
</td>
<td>
<ul>
<li>
<p>
<code>gendist</code>
</p>
</li>
<li>
<p>
<code>coords</code>
</p>
</li>
<li>
<p>
<code>envlayers</code>
</p>
</li>
</ul>
</td>
</tr>
<tr class="even">
<td>
RDA
</td>
<td>
<a href="https://cran.r-project.org/web/packages/vegan/index.html" class="external-link">vegan</a>
</td>
<td>
<code><a href="reference/rda_do_everything.html">rda_do_everything()</a></code>
</td>
<td>
<ul>
<li>
<p>
Genotype dosage matrix
</p>
</li>
<li>
<p>
Environmental layers
</p>
</li>
<li>
<p>
Sampling coordinates
</p>
</li>
</ul>
</td>
<td>
<ul>
<li>
<p>
<code>gen</code>
</p>
</li>
<li>
<p>
<code>env</code>
</p>
</li>
</ul>
</td>
</tr>
<tr class="odd">
<td>
LFMM
</td>
<td>
<a href="https://rdrr.io/bioc/LEA/" class="external-link">LEA</a>
</td>
<td>
<code><a href="reference/lfmm_do_everything.html">lfmm_do_everything()</a></code>
</td>
<td>
<ul>
<li>
Genotype dosage matrix
</li>
<li>
Environmental layers
</li>
<li>
Sampling coordinates
</li>
</ul>
</td>
<td>
<ul>
<li>
<p>
<code>gen</code>
</p>
</li>
<li>
<p>
<code>env</code>
</p>
</li>
</ul>
</td>
</tr>
<tr class="even">
<td>
wingen
</td>
<td>
<a href="https://github.com/AnushaPB/wingen" class="external-link">wingen</a>
</td>
<td>
<code><a href="reference/wingen_do_everything.html">wingen_do_everything()</a></code>
</td>
<td>
<ul>
<li>
VCF file
</li>
<li>
Raster layer (<em>though not necessary</em>)
</li>
<li>
Sampling coordinates
</li>
</ul>
</td>
<td>
<ul>
<li>
<p>
<code>gen</code>
</p>
</li>
<li>
<p>
<code>lyr</code>
</p>
</li>
<li>
<p>
<code>coords</code>
</p>
</li>
</ul>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h3>
<hr>
<p>Do alligators alligate? You can run through all of algatr’s functionality which we’ve coded up into a single function, <code><a href="reference/do_everything_for_me.html">do_everything_for_me()</a></code>. The <code>do_everything</code> functions provide a useful first pass at analyses, but each of algatr’s methods should be thoroughly tested and parameterized based on input data. Given this, we <strong><em>strongly</em></strong> advise against using this function for any final analyses.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># do_everything_for_me(liz_vcf, liz_coords, CA_env)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="installing-on-ubuntu">Installing on Ubuntu<a class="anchor" aria-label="anchor" href="#installing-on-ubuntu"></a>
</h3>
<p>You may run into issues installing the rmapshaper package if you’re using a Ubuntu system. If so, run the following line of code before attempting to install the package:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>sudo apt<span class="sc">-</span>get install <span class="sc">-</span>y libprotobuf<span class="sc">-</span>dev protobuf<span class="sc">-</span>compiler libjq<span class="sc">-</span>dev</span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing algatr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Anusha Bishop <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1731-8683" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Anne Chambers <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-7369-0108" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Ian Wang <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-2554-9414" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anusha Bishop, Anne Chambers, Ian Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
