---
title: "data_processing"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{data_processing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Processing genomic data files

```{r setup, warning = FALSE, message = FALSE}
library(algatr)
library(here)
library(vcfR)
library(SNPRelate)
library(SeqArray)
```

For landscape genomic analyses, we need to process our input data files in various ways, which are used throughout the analyses in this package. It is important to understand what this data processing is doing, so in this vignette, we'll do the following:

-   `vcf_to_dosage()` to convert a vcf file to a dosage matrix

-   `ld_prune()` to prune variants that are in linkage disequilibrium with one another; this primarily uses the `snpgdsLDpruning()` function within the SNPRelate package ([Zheng et al. 2012](10.1093/bioinformatics/bts606); see [here](https://www.bioconductor.org/packages/release/bioc/manuals/SNPRelate/man/SNPRelate.pdf) for further information)

## Converting a vcf to a dosage matrix with `vcf_to_dosage()`

------------------------------------------------------------------------

Recall the way genotypes are encoded within a vcf: `"0/0"` denotes individual is homozygous for the reference allele, `"0/1"` for heterozygote, and `"1/1"` for a homozygote for the alternate allele; `NA` is missing data.

```{r vcf}
# Load the example data
load_example()

# Look at genotypes for five individuals at eight sites:
liz_vcf@gt[1:8,2:6]
```

A dosage matrix contains each row as an individual (unlike a vcf) and each column as a site. Importantly, genotypes are now encoded with a single number. For a diploid organism, there are three possibilities for coding genotypes in a dosage matrix: 0, 1, or 2 (corresponding to 0/0, 1/1, and 0/1 from the vcf, respectively). `NA` still represents missing data. We can convert our vcf to a dosage matrix using the `vcf_to_dosage()` function:

```{r dosage}
dosage <- vcf_to_dosage(liz_vcf)

# Look at genotypes for five individuals at five sites:
dosage[1:5,1:8]
```

## Removing sites that are in linkage disequilibrium using `ld_prune()`

------------------------------------------------------------------------

We want to prune any linked sites from our data to not bias results. To do so, we'll use the `ld_prune()` function (which largely uses the `snpgdsLDpruning()` function within the SNPRelate package). The function works by first removing any variants that occur  below a given minor allele frequency, and then by pruning any variants that have a given correlation using a sliding window approach. The main arguments within this function are as follows:

-   `ld.threshold` specifies the LD pruning threshold (the default is the correlation coefficient value; thus, the higher the `ld.threshold` value, fewer variants will be pruned)
-   `slide.max.n` specifies the maximum number of SNPs in the sliding window (defaults to 100)
-   `maf` specifies the minor allele frequency (defaults to 0.05)

```{r LD-pruning int files}
# ld_prune() takes in the path to the vcf file, not the vcf object. The vcf used to generate the test dataset is in the extdata directory
# Leave the ld.threshold default (0.6), but change slide.max.n:
# vcf <- system.file("extdata", "liz_test.vcf", mustWork = TRUE)
ld_prune(vcf = "inst/extdata/liz_test.vcf", out_name = "liz", out_format = "vcf", slide.max.n=10)

# The above should have generated a new directory (liz_LDpruned) containing several files
# Let's take a look at the LD-pruned vcf:
vcf_ldpruned <- vcfR::read.vcfR(here("liz_LDpruned", "liz_LDpruned_r0.6_n10.vcf"))

# How many SNPs were retained after LD-pruning?
nrow(vcf_ldpruned@gt)
```

You'll notice that a number of intermediate files (contained within a new directory, liz_LDpruned), were created. This is because within the `ld_prune()` function, several file conversions are occurring, and any intermediate files produced from this conversion are stored in the new directory. There is also a logfile that is produced that provides output from the LD pruning.

We can specify not to create these intermediate files, but rather to just produce the end vcf, using the `save_output` argument. If we do this, the function will simply return the vcf object. Let's also change the `ld.threshold` argument for this run.

```{r LD-pruning no int files}
vcf_ldpruned <- ld_prune(vcf = "inst/extdata/liz_test.vcf", out_name = "liz", out_format = "vcf", ld.threshold = 0.9, slide.max.n=10, save_output = FALSE)

# How many sites were retained from this run, with a higher LD threshold correlation value?
nrow(vcf_ldpruned@gt)
```
