---
title: "Data_processing_vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{data_processing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Processing genomic data files

```{r setup, warning = FALSE, message = FALSE}
library(algatr)
library(here)
library(vcfR)
library(SNPRelate)
library(SeqArray)
library(wingen)

devtools::load_all()
```

For landscape genomic analyses, we first need to process our input data files in various ways. It is important to understand what this data processing is doing, so in this vignette, we'll do the following:

-   `vcf_to_dosage()` to convert a vcf file to a dosage matrix

-   `ld_prune()` to prune variants that are in linkage disequilibrium with one another; this primarily uses the `snpgdsLDpruning()` function within the SNPRelate package ([Zheng et al. 2012](https://academic.oup.com/bioinformatics/article/28/24/3326/245844); see [here](https://www.bioconductor.org/packages/release/bioc/manuals/SNPRelate/man/SNPRelate.pdf) for further information)

## Converting a vcf to a dosage matrix with `vcf_to_dosage()`

------------------------------------------------------------------------

Recall the way genotypes are encoded within a vcf: `"0/0"` denotes individual is homozygous for the reference allele, `"0/1"` for heterozygote, and `"1/1"` for a homozygote for the alternate allele; `NA` is missing data.

```{r vcf}
# Load the example data
load_algatr_example()

# Look at genotypes for five individuals at eight sites:
liz_vcf@gt[1:8, 2:6]
```

A dosage matrix contains each row as an individual (unlike a vcf) and each column as a site. Importantly, genotypes are now encoded with a single number. For a diploid organism, there are three possibilities for coding genotypes in a dosage matrix: 0, 1, or 2 (corresponding to 0/0, 1/1, and 0/1 from the vcf, respectively). `NA` still represents missing data. We can convert our vcf to a dosage matrix using the `vcf_to_dosage()` function:

```{r dosage}
dosage <- vcf_to_dosage(liz_vcf)

# Look at genotypes for five individuals at five sites:
dosage[1:5, 1:8]
```

## Removing sites that are in linkage disequilibrium using `ld_prune()`

------------------------------------------------------------------------

For some analyses, we may want to prune out any linked sites from our data to not bias results. This is particularly the case with methods that estimate population structure (e.g., TESS), as population structure can be overinflated as more SNPs appear to independently support the same pattern. For other analyses - such as genotype-environment association methods like RDA and LFMM - SNPs in LD are sometimes removed prior to running such analyses, but we do not recommend this because SNPs most strongly associated with environmental variables may have been pruned out. However, failing to account for linkage may result in overinflated conclusions about environmental associations. A solution to this is to perform LD-pruning after having run GEA methods; one approach to do so is to retain SNPs with the strongest association (i.e., lowest p-value) and removing all others that are in linkage with that particular SNP (see the approach described in Capblancq & Forester 2021 for more information).

To perform LD-pruning, we can use the `ld_prune()` function (which largely uses the `snpgdsLDpruning()` function within the SNPRelate package). The function works by first removing any variants that occur below a given minor allele frequency, and then by pruning any variants that have a given correlation using a sliding window approach.

The main arguments within this function are as follows:

-   `ld.threshold` specifies the LD pruning threshold (the default is the correlation coefficient value; thus, the higher the `ld.threshold` value, fewer variants will be pruned)

-   `slide.max.n` specifies the maximum number of SNPs in the sliding window (defaults to 100)

-   `maf` specifies the minor allele frequency (defaults to 0.05)

-   `vcf` specifies the path to the vcf file that you would like to be pruned

-   `out_name` specifies the prefix of output file names

-   `out_format` specifies the file type (options are `"vcf"` for vcf and gds files [default], or `"plink"` for ped and map files)

-   `save_output` specifies whether you would like a new directory made containing all output (and intermediate) files; default is `TRUE.` If set to `FALSE`, the function will simply return a vcf object.

The way to call this function is as follows:

`vcf_ldpruned <- ld_prune(vcf = "test.vcf", out_name = "liz", out_format = "vcf", ld.threshold = 0.9, slide.max.n = 10)`

We will not be running this function within the vignette to avoid generating intermediate files, but the example line above creates a new directory called liz_LDpruned, and it contains five files: a LOGFILE, with information on how many sites were pruned and how many remain, three gds files (pre- and post-pruned SNP GDS files and a GDS SeqArray file), and the pruned vcf. All of these output files are named such that the `ld.threshold` and `slide.max.n` parameter settings are contained in file names.

## Additional documentation and citations

------------------------------------------------------------------------

|                 | Citation/URL                                                                                                                                                                                           | Details                                                                                  |
|-------------------|-----------------------------|-------------------------|
| Associated code | [Zheng et al. 2012](https://academic.oup.com/bioinformatics/article/28/24/3326/245844); further details [here](https://www.bioconductor.org/packages/release/bioc/manuals/SNPRelate/man/SNPRelate.pdf) | algatr uses the `snpgdsLDpruning()` function within the SNPRelate package for LD-pruning |

Retrieve SNPRelate's vignette:

```{r package vignette}
vignette("SNPRelate")
```
