---
title: "gen_dist_vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Gendist_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Calculating genetic distances

```{r setup, warning = FALSE, message = FALSE}
library(algatr)
library(here)
library(vcfR)
library(ecodist)
library(adegenet)
```

For several landscape genomics analyses, a pairwise genetic distance matrix is required as input. There are a variety of ways to calculate genetic distance, and the main function in this package, `gen_dist()`, allows a user to choose from five different genetic distances and saves square distance matrices as csv files. The five genetic distance metrics are as follows:

1.  Euclidean distance (the `"euclidean"` argument), which uses the `distance()` function within the [ecodist](https://cran.r-project.org/web/packages/ecodist/index.html) package

2.  Bray-Curtis distance (the `"bray-curtis"` argument), which uses the `distance()` function within the [ecodist](https://cran.r-project.org/web/packages/ecodist/index.html) package

3.  Proportion of shared alleles (the `"dps"` argument), which uses the `propShared()` function within the [adegenet](https://cran.r-project.org/web/packages/adegenet/index.html) package

4.  PC-based distance (the `"pc"` argument), which uses the `prcomp()` function within the [stats](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/00Index.html) package, and selects the number of PCs using a Tracy-Widom test

5.  Processing distances generated using Plink (the `"plink"` argument)

A good place to start to understand these metrics is [Shirk et al., 2017](https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12684), who tested a number of different genetic distance metrics (including most of the above) and compared results for use in landscape genetics analyses.

This package also allows a user to look at the relationship between distance metrics using the `gen_dist_corr()` function, and to visualize one genetic distance metric with a heatmap using the `gen_dist_hm()` function.

## Load example dataset

```{r load example}
load_example()
```

## Calculate genetic distances with `gen_dist_calc()`

------------------------------------------------------------------------

Euclidean, Bray-Curtis, and PC-based genetic distance metrics ***require no missing data***. A simple imputation (to the median or mean, depending on the distance metric) is built into the `gen_dist_calc()` function mostly for creating test datasets; we ***highly discourage*** using this form of imputation in any of your analyses.

To calculate Euclidean distances between our samples:

```{r euc dists}
# Calculate Euclidean distance matrix
euc_dists <- gen_dist(liz_vcf, dist_type = "euclidean")

head(euc_dists)
```

Now, let's process Plink distances:

```{r plink dists}
plink_dists <- gen_dist(plink_file = here("inst", "extdata", "liz_test.dist"), plink_id_file = here("inst", "extdata", "liz_test.dist.id"), dist_type = "plink")
# TODO: switch above to system.file for raw plink dist files
```

Finally, let's calculate PC-based distances using a significance threshold of 0.01 (the default):

```{r pc dists}
pc_dists <- gen_dist(vcf = here("inst", "extdata", "liz_test.vcf"), dist_type = "pc")
# TODO: switch above to system.file
```

## Compare genetic distance results

------------------------------------------------------------------------

### Plot comparisons between two genetic distances with `gen_dist_corr()`

To compare between the genetic distances you just produced, we can plot them:

```{r gendist corr}
# Plot some pairwise comparisons
p_euc_plink <- gen_dist_corr(euc_dists, plink_dists)
p_pc_plink <- gen_dist_corr(pc_dists, plink_dists)

# Show all plots as panels on a single plot
plot_grid(p_euc_plink, p_pc_plink, nrow=1)
```

### Plot heatmap of one of the genetic distances with `gen_dist_hm()`

Let's look at a heatmap scaled to one of the genetic distance metrics we calculated.

```{r gendist hm}

```
