---
title: "RDA"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{RDA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, eval = FALSE}
# Install packages
rda_packages()
```

# Redundancy analysis (RDA)

```{r setup, warning=FALSE, message=FALSE}
library(algatr)
library(dplyr)
library(raster)
library(vegan)
```

**If using RDA, please cite the following: Capblancq T., Forester B.R. (2021) Redundancy analysis: A Swiss Army Knife for landscape genomics. Methods in Ecology and Evolution 12:2298-2309. DOI:10.1111/2041-210X.13722**

Redundancy analysis (RDA) is a genotype-environment association (GEA) method that uses constrained ordination to detect outlier loci that are significantly associated with environmental variables. It does so by combining regression (in which response variables are genetic data, in our case, while explanatory variables are environmental data) with ordination (a PCA). Importantly, RDA is multivariate (i.e., multiple loci can be considered at a time), which is appealing in cases where multilocus selection may be occurring.

Model selection is performed by starting with a null model wherein the response is only explained by an intercept, and environmental variables are added until the amount of variance explained by a model that includes all variables is reached. By doing so, this method minimizes the *redundancy* among explanatory variables. Importantly, we can also perform an RDA that accounts for *covariables* (or "conditioning variables") such as those that account for neutral population structure. When covariables are included, this is called a **partial RDA**. In algatr, population structure is quantified by running a PCA on the genetic data, and selecting a number of PC axes to best represent this type of neutral variation.

Some of the earliest examples of uses of RDA for identifying environment-associated loci are [Lasky et al. 2012](https://onlinelibrary.wiley.com/doi/10.1111/j.1365-294X.2012.05709.x) and [Forester et al. 2016](https://onlinelibrary.wiley.com/doi/10.1111/mec.13476). Since then, there have been several reviews and walkthroughs of the method that provide additional information, including workflows and comparisons with other GEA methods (e.g., [Forester et al. 2018](https://onlinelibrary.wiley.com/doi/full/10.1111/mec.14584)). Much of algatr's code is adapted from one such paper, [Capblancq & Forester 2021](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722), the code of which is available [here](https://github.com/Capblancq/RDA-landscape-genomics). algatr's general workflow roughly follows the framework of Capblancq & Forester (2021).

RDA cannot take in missing values. Imputation based on the per-site median is commonly performed, but there are several other ways researchers can deal with missing values. For example, algatr contains the `str_impute()` function to impute missing values based on population structure using the `LEA::impute()` function. However, here, we'll opt to impute to the median, but strongly urge researchers to use extreme caution when using this form of simplistic imputation. We mainly provide code to impute on the median for testing datasets and highly discourage its use in further analyses (please use `str_impute()` instead!).

The general workflow to perform an RDA with algatr is as follows:

-   **Simple vs. partial RDAs:** We can run a simple RDA (no covariables) or partial RDA (conditioning covariables included) using `rda_run()`

-   **Variable selection:** Both simple and partial RDAs can be performed considering all variables (`"full"` model) or by performing variable selection to determine variables that contribute most to genetic variance (`"best"` model)

-   **Variance partitioning:** We can perform variance partitioning considering only variables that contribute significantly to genetic variance using `rda_varpart()`

-   **Detecting outlier loci:** Based on the results from above, we can detect outlier loci with `rda_getoutliers()`

### Read in and process genetic data

Running an RDA requires two data files for input: a genotype dosage matrix (the `gen` argument) and the environmental values extracted at sampling coordinates (the `env` argument). Let's first convert our vcf to a dosage matrix using the `vcf_to_dosage()` function. *N.B.: our code assumes that sample IDs from our genetic data and our coords are in the same order; be sure to check this before moving forward!*

```{r gen data, warning = FALSE}
load_algatr_example()
# Convert from vcf to dosage matrix:
gen <- vcf_to_dosage(liz_vcf)
```

As mentioned above, running an RDA requires that your genotype matrix contains no missing values. Let's impute missing values based on the per-site median. *N.B.: this type of simplistic imputation is strongly not recommended for downstream analyses and is used here for example's sake!*

```{r impute}
# Are there NAs in the data?
gen[1:5, 1:5]
gen <- simple_impute(gen)
# Check that NAs are gone
gen[1:5, 1:5]
```

### Process environmental data

Let's extract environmental variables using the `extract()` function from the raster package. We also need to standardize environmental variables. This is particularly important if we are using (for example) bioclimatic variables as input, as units of measurement are completely different (e.g., mm for precipitation vs. degrees Celsius for temperature). To do so, we'll use the `scale()` function within the raster package.

```{r enviro data}
# Extract environmental vars
env <- raster::extract(CA_env, liz_coords)

# Standardize environmental variables and make into dataframe
env <- scale(env, center = TRUE, scale = TRUE)
env <- data.frame(env)
```

## Running simple and partial RDAs using `rda_run()`, with and without variable selection

------------------------------------------------------------------------

The main function within algatr to perform an RDA is `rda_run()`, which uses the `rda()` function within the vegan package.

### Run a simple RDA with no variable selection

A simple RDA is one in which our model will not account for covariables in the model, and is the default for `rda_run()`. Let's first run a simple RDA on all environmental variables (i.e., no variable selection). This is specified using the `model = "full"` argument.

```{r full model, warning = FALSE, message = FALSE}
mod_full <- rda_run(gen, env, model = "full")
```

The resulting object is large, containing 12 elements relevant to the RDA. Let's take a look at what function was called. We can see that all environmental variables (CA_rPCA1, 2, and 3) were included in the model, and that geography or structure were not included in the model (and there were no conditioning variables).

```{r full call}
mod_full$call
```

Now, let's take a look at the summary of this model. One of the most important parts of this object is the partitioning of variance. Within an RDA, the term "inertia" can be interpreted as variance, and our results show us that the amount of variance explained by explanatory variables alone (constrained inertia) is only 9.768%. Unconstrained inertia is the residual variance, which is very high (90.232%). The summary also provides site scores, site constraints, and biplot scores.

```{r full summary}
head(summary(mod_full))
```

One of the most relevant statistics to present from an RDA model is the adjusted R2 value. The R2 value *must* be adjusted; if not, it is biased because it will always increase if independent variables are added (i.e., there is no penalization for adding independent variables that aren't significantly affecting dependent variables within the model). The adjusted R2 value from the full (global) model can also help us determine the stopping point when we go on to do variable selection. As we can see, the unadjusted R2 is 0.098, while the adjusted value is 0.042. Be sure to always report *adjusted* R2 values.

```{r full Radj}
RsquareAdj(mod_full)
```

### Run a simple RDA with variable selection

Now, let's run an RDA model with variable selection by specifying the `"best"` model within `rda_run()`. Variable selection occurs using the `ordiR2step()` function in the vegan package, and is a forward selection method that begins with a null model and adds explanatory variables one at a time until genetic variance is maximized. To perform forward selection, we need to specify stopping criteria (i.e., when to stop adding more explanatory variables). There are two primary ways to specify stopping criteria: parameters involved in a permutation-based significance test (two parameters: the limits of permutation p-values for adding (`Pin`) or dropping (`Pout`) a term to the model, and the number of permutations used in the estimation of adjusted R2 (`R2permutations`)), and whether to use the adjusted R2 value as the stopping criterion (`R2scope`; only models with adjusted R2 values lower than specified are accepted).

```{r best model, warning = FALSE, message = FALSE}
mod_best <- rda_run(gen, env,
  model = "best",
  Pin = 0.05,
  R2permutations = 1000,
  R2scope = T
)
```

Let's look at our best model, and see how it compares to our full model. There is also an additional element in the `mod_best` object, which are the results from an ANOVA-like permutation test. When we look at the call, we can now see that only environmental PCs 2 and 3 are considered in the model, meaning that PC1 is not considered to be significantly associated with genetic variance. Importantly, our adjusted R2 value is 0.039 (compared to 0.042 for the full model), which tells us that the model with only two of the environmental variables explains nearly as much as the RDA with all three.

```{r best model stats}
mod_best$call
mod_best$anova
RsquareAdj(mod_best)
```

Following variable selection, we want to retain two environmental variables (CA_rPCA2 and CA_rPCA3) for all subsequent analyses.

### Run a partial RDA with geography as a covariable

A partial RDA, as described at the beginning of the vignette, is when covariables (or conditioning variables) are incorporated into the RDA. Within `rda_run()`, to correct for geography, users can set `correctGEO = TRUE`, in which case sampling coordinates (`coords`) must also be provided. Let's now run a partial RDA, without any variable selection (the full model) and view the results.

```{r partial rda, warning=FALSE, message=FALSE}
mod_pRDA_geo <- rda_run(gen, env, liz_coords,
  model = "full",
  correctGEO = TRUE,
  correctPC = FALSE
)
```

Let's take a look at our partial RDA results. Note that now, the call includes conditioning variables x and y, which correspond to our latitude and longitude values. Conditioning variables are specified within the `rda()` function using the `Condition()` function. Finally, when we look at the summary of the resulting object, we can see that along with total, constrained, and unconstrained inertia (or variance) as with our simple RDA, there is an additional row for conditioned inertia because our partial RDA includes conditioning variables. Interestingly, the conditioned inertia is higher than constrained inertia, suggesting that geography is more associated with genetic variance than our environmental variables are.

```{r pRDA results}
anova(mod_pRDA_geo)
RsquareAdj(mod_pRDA_geo) # 0.0305
head(summary(mod_pRDA_geo))
```

### Run a partial RDA with population genetic structure and geography as covariables

Lastly, let's run a partial RDA with both geography (`correctGEO = TRUE`) and population structure (`correctPC = TRUE`) as conditioning variables. Within algatr, population structure is quantified using a PCA, and a user can specify the number of axes to retain in the RDA model. In this case, we'll retain three axes, which we can specify using the `nPC` argument. If this argument is set to `"manual"`, a screeplot will be printed and a user can manually enter the number of PCs that they feel best describe the data.

```{r pRDA struct+geo}
mod_pRDA_gs <- rda_run(gen, env, liz_coords,
  model = "full",
  correctGEO = TRUE,
  correctPC = TRUE,
  nPC = 3
)
```

Now, we see that even more inertia/variance is explained by the conditioning variables.

```{r pRDA struct+geo summaryl}
head(summary(mod_pRDA_gs))
```

## Variance partitioning with partial RDA

------------------------------------------------------------------------

In many cases, we want to understand the independent contributions of each explanatory variable in combination with our covariables. Also, we can understand how much confounded variance there is (i.e., variation that is explained by the combination of multiple explanatory variables). The best way to do this is to run partial RDAs on each set of explanatory variables and examine the inertia of each pRDA result as it compares to the full model. In this case, the full model is an RDA that considers neutral processes as explanatory variables rather than conditioning variables (covariables).

The call for this full model looks like this: `rda(gen ~ env + covariable, data)`, rather than a true partial RDA which would take the form: `rda(gen ~ env + Condition(covariable), data)`.

In algatr, we can run variance partitioning using the function `rda_varpart()`, which will run the full model (as above) and then will run successive pRDAs which are conditioned on three successive sets of variables. These variable sets are significant environmental variables (in our case, environmental PCs 2 and 3), population genetic structure (once again determined using a PC with a user-specified number of PC axes), and geography (sampling coordinates). This function will then provide us with information on how variance is partitioned in the form of a table containing relevant statistics; this table is nearly identical to Table 2 from [Capblancq & Forester 2021](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722). The `rda_varpart()` parameters are identical to those from `rda_run()`. As with `rda_run()`, we can provide a number for the number of PCs to consider for population structure (`nPC`), or this argument can be set to `"manual"` if a user would like to manually enter a number into the console after viewing the screeplot. We can then use `rda_varpart_table()` to visualize these results in an aesthetically pleasing way; users can specify whether to display the function call in the table (`call_col`; defaults to `FALSE`) and the number of digits to round to `digits` (the default is 2).

```{r varpart, warning=FALSE, message=FALSE}
varpart <- rda_varpart(gen, env, liz_coords,
  Pin = 0.05, R2permutations = 1000,
  R2scope = T, nPC = 3
)
rda_varpart_table(varpart, call_col = TRUE)
```

From the above table, we can see that the full model only explains 27% of the total variance in our genetic data. Furthermore, our pure environmental model explains only a small proportion of the variance ("inertia") contained within the full model (9%), and that population structure explains substantially more (44%). There is also a large amount of variance (37%) within the full model that is confounded, meaning that variables within the three variable sets explain a substantial amount of the full model's variance *when in combination with one another*. This is a relatively common occurrence (see Capblancq & Forester 2021), and users should always keep in mind how collinearity may affect their results (and account for this, if possible; see algatr's environmental data vignette for more information on how to do so). However, accounting for collinearity may also elevate false positive rates, while failing to account for it at all can lead to elevated false negative rates.

In our case, it is clear from our variance partitioning analysis that much of the variation in our genetic data can explained by population structure, and thus users may want to further explore more sophisticated approaches for quantifying demographic history or population structure before moving forward with outlier loci detection. As with most analyses, careful consideration of the input data and the biology of the system are critical to draw robust and reliable conclusions.

## Identifying candidate SNPs with `rda_getoutliers()`

------------------------------------------------------------------------

Having run variance partitioning on our data, we can now move forward with detecting outlier loci that are significantly associated with environmental variables. To do so, we'll first run a partial RDA on only those significant environmental variables, and we'll also want to add population structure as a covariable (i.e., a conditioning variable). We'll only retain two PC axes for our measurement of population structure.

```{r pRDA final}
mod_pRDA <- rda_run(gen, env, model = "best", correctPC = TRUE, nPC = 2)
```

Given this partial RDA, only environmental PC2 is considered significant.

```{r pRDA summary}
mod_pRDA$anova
```

Now that we have the RDA model, we can discover SNPs that are outliers. In algatr, this can be done using two different methods to detect outlier loci, specified by the `outlier_method` argument. The first uses Z-scores for outlier detection, and is from [Forester et al. 2018](https://onlinelibrary.wiley.com/doi/abs/10.1111/mec.14584); their code can be found [here](https://popgen.nescent.org/2018-03-27_RDA_GEA.html). The other method transforms RDA loadings into p-values, and was developed by [Capblancq et al. 2018](https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906) and is further described in [Capblancq & Forester 2021](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722); code developed to run this is associated with Capblancq and Forester 2021 and can be found [here](https://github.com/Capblancq/RDA-landscape-genomics). We provide brief explanations for each of these methods below.

**N.B.: Some researchers may prefer to do MAF filtering and linkage disequilibrium pruning during outlier detection (e.g., Capblancq & Forester 2021). In this case, only one SNP per contig is selected based on the highest loading. The algatr pipeline does this filtering initially when processing genetic data; see data processing vignette for more information.**

### Identifying outliers using the Z-scores method

To understand how the Z-scores method works to detect outliers, let's first take a look at the loadings of SNPs on the first two RDA axes using the `rda_plot()` function. This will show us histograms for each of the selected RDA axes. SNPs that are not associated with environmental variables will fall in the center of these distributions, with outliers falling at the edges. The Z-score method works by setting a Z-score (number of standard deviations) threshold beyond which SNPs will be considered outliers (see [here](https://popgen.nescent.org/2018-03-27_RDA_GEA.html) for further details).

If we set this value too stringently, we minimize our false positive rate but risk detecting false negatives; if it's set too high, we increase our false positive rate. Researchers will want to adjust this parameter based on their tolerance for detecting false positives, but also the strength of selection under which outliers are subject to (assuming that SNPs that are more significantly associated with environmental variables are also under stronger selection).

Because only one environmental variable was associated with our genetic data, we only have one resulting RDA axis.

```{r loadings, warning=FALSE, message=FALSE}
rda_plot(mod_pRDA, axes = "all", binwidth = 20)
```

Now, let's discover outliers with the `rda_getoutliers()` function. The method is specified using the `outlier_method` argument, and the number of standard deviations to detect outliers is specified by the `z` argument; it defaults to 3. The resulting object is a dataframe containing the SNP name, the RDA axis it's associated with, and the loading along this axis. A screeplot for the RDA axes is automatically generated (but can be switched off using the `plot` argument).

```{r z-scores}
rda_sig_z <- rda_getoutliers(mod_pRDA, naxes = "all", outlier_method = "z", z = 3, plot = FALSE)

# How many outlier SNPs were detected?
length(rda_sig_z$rda_snps)
```

### Identifying outliers using the p-value method

Another method to detect outliers is a p-value method wherein rather than using the loadings along RDA axes to identify extreme outliers (as with the Z-scores method above), the loadings are transformed into p-values. To do so, Mahalanobis distances are calculated for each locus, scaled based on a genomic inflation factor, from which p-values and q-values can be computed. We then adjust our p-values (based on a false discovery rate, for example; this adjustment is done using the `p.adjust()` function) and subsequently identify outliers based on a user-specified significance threshold. Rather than detecting outliers based on a p-value significance threshold, one could also detect outliers based on q-values (which are also provided as output in algatr), as is done in [Capblancq et al. 2018](https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906) and [Capblancq & Forester 2021](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722). To do so, we would select loci that fall below a user-specified threshold based on the proportion of false positives we allow. For example, if we allow 10% or fewer outliers to be false positives, outliers could be identified based on loci that have q-values =\< 0.1.

Now, let's detect outliers using this method. To do so, along with `outlier_method`, we need to specify `p_adj`, which is the method by which p-values are adjusted, as well as `sig`, which is our significance threshold. We will also tell the function not to produce a screeplot.

One limitation of the p-value outlier method is that you cannot calculate outliers given fewer than two RDA axes (as is the case with our partial RDA results). So, let's move forward with this outlier detection method but using the results from our simple RDA model with variable selection, `mod_best`.

```{r cand snps}
rda_sig_p <- rda_getoutliers(mod_best, naxes = "all", outlier_method = "p", p_adj = "fdr", sig = 0.01, plot = FALSE)

# How many outlier SNPs were detected?
length(rda_sig_p$rda_snps)
```

As mentioned above, we could also identify outliers based on q-values. To do so, we can extract the q-values from the resulting object from `rda_getoutliers()`.

```{r q-values}
# Extract SNP names; choices is number of axes
snp_names <- rownames(scores(mod_best, choices = 2, display = "species"))

# Identify outliers that have q-values < 0.1
q_sig <-
  rda_sig_p$rdadapt %>%
  mutate(snp_names = snp_names) %>%
  filter(q.values <= 0.1)

# How many outlier SNPs were detected?
nrow(q_sig)
```

Note the differences between the two methods: using the p-value method, we found many more outlier SNPs than with the Z-scores method (101 vs. 19). Even more outliers were found when only using q-values as a significant threshold (170). We can also see that 19 SNPs were detected using all three methods.

```{r intersect}
Reduce(intersect, list(
  q_sig$snp_names,
  rda_sig_p$rda_snps,
  rda_sig_z$rda_snps
))
```

## Visualizing RDA results with `rda_plot()`

------------------------------------------------------------------------

algatr provides options for visualizing RDA results in two ways with the `rda_plot()` function: a biplot and a Manhattan plot.

### RDA biplot

Let's build a biplot (or some call it a triplot, since it also shows loadings of environmental variables) with RDA axes 1 and 2, specified using the `rdaplot` argument. In this case, we also need to specify the number of axes to display using the `biplot_axes` argument. The default of `rda_plot()` is to plot all variables, regardless of whether they were considered significant or not.

```{r biplot}
rda_plot(mod_best, rda_sig_p$rda_snps, biplot_axes = c(1, 2), rdaplot = TRUE, manhattan = FALSE)

# In the case of our partial RDA, there was only one RDA axis, so a histogram is generated
rda_plot(mod_pRDA, rda_sig_z$rda_snps, rdaplot = TRUE, manhattan = FALSE, binwidth = 0.01)
```

### Manhattan plot

Let's now build a Manhattan plot with a significance level of 0.05 using the `manhattan` argument; if this is set to TRUE, we must also provide a significance threshold value using the `sig` argument (the default is 0.05). Be aware that outliers were detected with a different p-value threshold which is why there are grey points above our threshold line (i.e., data points are colorized according to the threshold of the model, not the visualization).

```{r Manhattan}
rda_plot(mod_best, rda_sig_p$rda_snps, rda_sig_p$pvalues, rdaplot = FALSE, manhattan = TRUE)
```

## Interpreting RDA results

------------------------------------------------------------------------

As discussed in [Capblancq & Forester 2021](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722), users should be aware that ***"statistical selection of variables optimizes variance explained but cannot, by itself, identify the ecological or mechanistic drivers of genetic variation."***

### Identifying environmental associations with `rda_cor()` and `rda_table()`

Now that we've identified outlier loci, it is useful to be able to identify how much they correlate to each environmental variable. To do so, we can run a correlation test of each SNP using `rda_cor()` and summarize results in a customizable table using `rda_table()`.

```{r simple results}
# Extract genotypes for outlier SNPs
rda_snps <- rda_sig_p$rda_snps
rda_gen <- gen[, rda_snps]

# Run correlation test
cor_df <- rda_cor(rda_gen, env)

# Make a table from these results (displaying only the first 5 rows):
rda_table(cor_df, nrow = 5)

# Order by the strength of the correlation
rda_table(cor_df, order = TRUE, nrow = 5)

# Only retain the top variable for each SNP based on the strength of the correlation
rda_table(cor_df, top = TRUE, nrow = 5)

# Display results for only one environmental variable
rda_table(cor_df, var = "CA_rPCA2", nrow = 5)
```

## Running RDA with `rda_do_everything()`

------------------------------------------------------------------------

The algatr package also has an option to run all of the above functionality in a single function, `rda_do_everything()`. This function is similar in structure to `rda_run()`, but also performs imputation (`impute` option), can generate figures, the correlation test table (`cortest` option; default is TRUE), and variance partitioning (`varpart` option; default is FALSE). **Please be aware that the `do_everything()` functions are meant to be exploratory. We do not recommend their use for final analyses unless certain they are properly parameterized.**

The resulting object from `rda_do_everything()` contains:

-   Candidate SNPs: `rda_snps`

-   Data frame of environmental associations with outlier SNPs: `cor_df`

-   RDA model specifics: `rda_mod`

-   Outlier test results: `rda_outlier_test`

-   Relevant R-squared values: `rsq`

-   ANOVA results: `anova`

-   List of all p-values: `pvalues`

-   Variance partitioning results: `varpart`

Let's first run a simple RDA, imputing to the median, with variable selection and the p-value outlier method:

```{r simple RDA do everything, warning=FALSE, message=FALSE, results='asis', fig.align='center'}
results <- rda_do_everything(liz_vcf, CA_env, liz_coords,
                             impute = "simple",
                             correctGEO = FALSE,
                             correctPC = FALSE,
                             outlier_method = "p",
                             sig = 0.05,
                             p_adj = "fdr",
                             cortest = TRUE,
                             varpart = FALSE
)
```

Now, let's run a partial RDA, imputing based on 3 sNMF clusters, correcting for both geography and structure, and instead using the Z-score method to detect outlier loci, including variance partitioning. We'll keep the number of PCs for population structure set to the default, which is three axes (`nPC`):

```{r pRDA do everything, warning=FALSE, message=FALSE, results='asis', fig.align='center'}
results <- rda_do_everything(liz_vcf, CA_env, liz_coords,
                             impute = "structure",
                             K_impute = 3,
                             model = "full",
                             correctGEO = TRUE,
                             correctPC = TRUE,
                             nPC = 3,
                             varpart = TRUE,
                             outlier_method = "z",
                             z = 3
)
```

## Additional documentation and citations

------------------------------------------------------------------------

|                         | Citation/URL                                                                                                                                                                            | Details                                                                                             |
|-------------------|----------------------------|-------------------------|
| Associated literature   | [Capblancq & Forester 2021](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722); [Github repository for code](https://github.com/Capblancq/RDA-landscape-genomics) | Paper describing RDA methodology; also provides walkthrough and description of `rdadapt()` function |
| Associated literature   | [Capblancq et al. 2018](https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906)                                                                                                | Description of p-value method of detecting outlier loci                                             |
| R package documentation | [Oksanen et al. 2022.](https://cran.r-project.org/web/packages/vegan/index.html)                                                                                                        | algatr uses the `rda()` and `ordiR2step()` functions in the vegan package                           |
| Associated literature   | [Forester et al. 2018](https://onlinelibrary.wiley.com/doi/full/10.1111/mec.14584)                                                                                                      | Review of genotype-environment association methods                                                  |
| Associated tutorial     | [Forester et al.](https://popgen.nescent.org/2018-03-27_RDA_GEA.html)                                                                                                                   | Walkthrough of performing an RDA with the Z-scores outlier detection method                         |
| Associated vignette     | `vignette("intro-vegan")`                                                                                                                                                               | Introduction to ordination using the vegan package (including using the `rda()` function            |
| Associated vignette     | `vignette("partitioning")`                                                                                                                                                              | Introduction to variance partitioning in vegan                                                      |
