---
title: "RDA_vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{RDA_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

### Redundancy analysis (RDA)

```{r setup, warning=FALSE, message=FALSE}
library(algatr)
library("here")
library("gt")
# RDA
library("vegan")
library("raster")
library("robust")
library("tidyverse")
# BiocManager::install("qvalue")
library("qvalue")
devtools::load_all()
```

RDA is a method to detect outlier loci that are significantly associated with environmental variables.

Importantly, RDA cannot take in missing values. Imputation based on the per-site median is commonly performed, but there are several other ways researchers can deal with missing values. Here, we'll explore imputing on the median, but strongly urge researchers to use extreme caution when using this form of simplistic imputation. We mainly provide code to impute on the median for testing datasets and highly discourage its use in further analyses.

### Read in and process input data

Running an RDA requires three data files for input: a genotype dosage matrix (the `gen` argument), a genotype dosage matrix (the `gen` argument) and the environmental values extracted at sampling coordinates (the `env` argument). If you do not have extracted environmental values (as with the test dataset), `lfmm_do_everything()` will extract values for us. For clarity's sake here, we'll do the conversion from a vcf to dosage matrix using the `vcf_to_dosage()` function. 

```{r, warning = FALSE}
# Load vcf, genetic dist matrix, and coordinates for 53 inds, and three environmental layers for test dataset
load_example()
# Convert from vcf to dosage matrix:
gen <- vcf_to_dosage(liz_vcf)
# Also, our code assumes that sample IDs from gen and coords are the same order; be sure to check this before moving forward!
```

### Run a simple RDA

The RDA analysis in algatr can be done using two different methods to detect outlier loci, specified by the `outlier_method` argument. The first, using p values, is described in [Capblancq and Forester 2021](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722); their code can be found [here](https://github.com/Capblancq/RDA-landscape-genomics). The other method, relying on z-scores for outlier detection, is from [Forester et al. 2018](https://onlinelibrary.wiley.com/doi/abs/10.1111/mec.14584); their code can e found [here](https://popgen.nescent.org/2018-03-27_RDA_GEA.html).

Let's start with using the p-value method to detect outlier loci:

```{r, warning = FALSE, message = FALSE}
results <- rda_do_everything(gen, CA_env, liz_coords, 
                            correctGEO = FALSE, 
                            correctPC = FALSE,
                            outlier_method = "p",
                            sig = 0.05, 
                            padj_method = "fdr")
```

Now, let's run the same analysis but instead using the z-scores method to detect outlier loci:

```{r}
results <- rda_do_everything(gen, env, coords, 
                            correctGEO = FALSE, 
                            correctPC = FALSE,
                            outlier_method = "z",
                            z = 3, 
                            padj_method = "fdr")
```

## Look at simple RDA results

The results from `rda_do_everything()` contain:

1. Candidate SNPs: `rda_snps`

2. Data frame of environmental associations with outlier SNPs: `cor_df`

3. RDA model specifics: `rda_mod`

4. Outlier test results: `rda_outlier_test`

5. Relevant R-squared values: `rsq`

6. ANOVA results: `anova`

7. List of all p-values: `pvalues`

In our case, we can see that there were 138 outlier SNPs detected.

```{r simple results}
summary(results)
```

### Visualize results

## Biplot

Let's build a biplot with RDA axes 1 and 2. The default of `rda_plot()` is to plot all variables.

```{r biplot simple}
rda_plot(results$rda_mod, results$rda_snps, biplot_axes = c(1,2), rdaplot = TRUE, manhattan = FALSE)
```

## Manhattan plot

Let's now build a Manhattan plot with a significance level of 0.05.

```{r Manhattan simple}
rda_plot(results$rda_mod, results$rda_snps, results$pvalues, sig = 0.05, manhattan = TRUE, rdaplot = FALSE)
```

### Run a partial RDA

A partial RDA describes incorporating covariables into a redundancy analysis. For more information, there is an excellent description of the difference between a simple and partial RDA in [Capblancq and Forester 2021](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722). Built within the `rda_do_everything()` function is the option to account for geographic distance and XXX as possible covariables (or the combination of the two).

To condition on geographic sampling coordinates, use the `correctGEO` argument:

```{r, warning = TRUE, message = FALSE}
results <- rda_do_everything(gen, 
                            env, 
                            coords, 
                            model = "full",
                            correctGEO = TRUE, 
                            correctPC = FALSE,
                            sig = 0.05, 
                            padj_method = "fdr")
```

To condition on PCs built using a PCA of all genotypes, use the `correctPC` argument:

```{r, warning = TRUE, message = FALSE}
results <- rda_do_everything(gen, 
                            env, 
                            coords, 
                            model = "full",
                            correctGEO = FALSE, 
                            correctPC = TRUE,
                            sig = 0.05, 
                            padj_method = "fdr")
```

To condition both on PCs and geographic coordinates:

```{r, warning = TRUE, message = FALSE}
results <- rda_do_everything(gen, 
                            env, 
                            coords, 
                            model = "full",
                            correctGEO = TRUE, 
                            correctPC = TRUE,
                            sig = 0.05, 
                            padj_method = "fdr")
```

Let's look at how these results compare.

```{r}
# TODO [EAC] fill in 
```

## Variable selection

```{r, warning = TRUE, message = FALSE}
results <- rda_do_everything(gen, 
                            env, 
                            coords, 
                            model = "best",
                            correctGEO = FALSE, 
                            correctPC = FALSE,
                            sig = 0.05, 
                            padj_method = "fdr")
```

## Separate functions

### 1. Prepare data
```{r}
# Extract and standardize environmental variables
env <- raster::extract(CA_env, coords)
env <- scale(env, center = TRUE, scale = TRUE) 
env <- data.frame(env)
    
# Remove any NA values
if(any(is.na(env))){
  warning("NA values found in env data, removing rows with NAs for RDA")
  gen <- gen[complete.cases(env),]
  coords <- coords[complete.cases(env),]
  env <- env[complete.cases(env),]
}
if(any(is.na(gen))){
  warning("NA values found in gen data, removing rows with NAs for RDA")
  env <- env[complete.cases(gen),]
  coords <- coords[complete.cases(gen),]
  gen <- gen[complete.cases(gen),]
}
```

### 2. Run RDA
```{r}
mod <- rda_run(gen, env, correctGEO = FALSE, correctPC = FALSE)
```


### 3. Identify candidate snps
```{r}

# Running with all axes and z method
rda_sig <- rda_getoutliers(mod, naxes = "all", outlier_method = "z", z = 3)
rda_snps <- rda_sig[["rda_snps"]]
length(rda_snps)

# Running with all axes and p method
rda_sig <- rda_getoutliers(mod, naxes = "all", outlier_method = "p", padj_method = "fdr", sig = 0.01)
rda_snps <- rda_sig[["rda_snps"]]
pvalues <- rda_sig[["pvalues"]]
length(rda_snps)

```

### 4. Identify environmental associations
```{r}
rda_gen <- gen[,rda_snps]
cor_df <- rda_cor(rda_gen, env)

# You can make a pretty table from these results (here we are displaying only the first 5 rows):
rda_table(cor_df, nrow = 5)

# you can order by the strength of the correlation
rda_table(cor_df, order = TRUE, nrow = 5)

# you can choose to only keep the top variable for each SNP the table by the strength of the correlation
rda_table(cor_df, top = TRUE, nrow = 5)

# you can display only select environmental variables
rda_table(cor_df, var = "CA_rPCA1", nrow = 5)

# putting it all together we can display the top 5, ordered SNPS for each variable
rda_table(cor_df, top = TRUE, order = TRUE, var = "CA_rPCA1", nrow = 5)
rda_table(cor_df, top = TRUE, order = TRUE, var = "CA_rPCA2", nrow = 5)
rda_table(cor_df, top = TRUE, order = TRUE, var = "CA_rPCA3", nrow = 5)
```

### 5. Plot results

```{r, fig.width = 6, fig.height = 5}
rda_plot(mod, rda_snps, biplot_axes = c(1,3), rdaplot = TRUE)
rda_plot(mod, rda_snps, biplot_axes = c(2,3), rdaplot = TRUE)
rda_plot(mod, rda_snps, pvalues, sig = 0.05, manhattan = TRUE, rdaplot = FALSE)

# note: if only one axis is provided, a histogram will be generated
rda_plot(mod, rda_snps, axes = 1, rdaplot = TRUE, binwidth = 0.01)
```

