---
title: "wingen-vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{wingen-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# wingen

```{r setup, warning = FALSE, message = FALSE}
library(algatr)
library(wingen)
library(raster)
library(vcfR)
library(viridis)
```

wingen is a package that uses a moving window approach to create continuous maps of genetic diversity. Please see [Bishop et al. 2022](REFER) for the paper describing the method, and refer to the [wingen package](https://github.com/AnushaPB/wingen) for further documentation.

There are three key components within wingen:

-   `window_gd()` to create a map of genetic diversity

-   `krig_gd()` for spatial interpolation (smoothing) of the resulting map using kriging

-   `mask_gd()` to mask out any undersampled areas (or generally areas that are not of interest)

### Load example data

```{r test data}
load_example()

# For wingen, let's just consider PC1 of CA_env
envlayer <- CA_env$CA_rPCA1
```

## Moving window calculations using `window_gd()` and `preview_gd()`

------------------------------------------------------------------------

The `window_gd()` function takes in a vcf file (the `vcf` argument), sampling coordinates (`coords` argument), and a RasterLayer for the window to move across (the `lyr` argument), and the genetic diversity summary statistic to calculate (the `stat` argument); choices are allelic richness (`"allelic.richness"` or `"biallelic.richness"`), nucleotide diversity (`"pi"`), or average heterozygosity (`"Ho"`).

There are a number of additional arguments within this function, and the wingen package documentation provides extensive explanations (and guidelines for testing) these parameters. Briefly, `fact` aggregates the input raster layer to allow for faster computational speeds by a given factor, `wdim` specifies the window dimensions, `rarify` specifies whether the user would like samples to be rarified (which then requires users to specify `rarify_n` and `rarify_nit`), and finally `parallel` specifies whether you would like to parallelize.

### Preview the moving window

First, we can get an idea of what the size of the cell and moving window look like using the `preview_gd()` function. wingen calculates the specified diversity statistic from samples that fall within the window, and then assigns the focal cell within that window a value according to the calculated diversity. 

```{r, fig.width = 5, fig.height = 5, cache = TRUE}
preview_gd(envlayer,
           liz_coords,
           wdim = 10, 
           fact = 5,
           sample_count = TRUE, 
           min_n = 2)
```

### Run the moving window

Next, let's run the moving window function (`window_gd()`) using the parameters specified above.

```{r moving window, fig.width = 5, fig.height = 5, cache = TRUE}
wgd <- window_gd(liz_vcf,
          liz_coords,
          envlayer,
          stat = "pi",
          wdim = 10,
          fact = 5,
          rarify = FALSE,
          L = 100)
```

### Plot wingen results

wingen allows for plotting maps in two ways: `plot_gd()` plots the genetic diversity layer, while `plot_count()` plots the sample count layer.

```{r plot window}
plot_gd(wgd, main = "Window pi")
plot_count(wgd, main = "Window sample counts")
```

### Krige results

Kriging is a type of spatial interpolation to allow for smoother maps of genetic diversity. We can krige our moving window map using the `krig_gd()` function.

### Mask results

## Running wingen with `wingen_do_everything()`

------------------------------------------------------------------------

The algatr package also has an option to run all of the above functionality in a single function, `wingen_do_everything()`. This function will TODO XXXXX. The resulting object looks identical to the output object from TODO XXXX.

```{r do everything}
# wingen_do_everything()
```
