<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate genetic distances — gen_dist • algatr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate genetic distances — gen_dist"><meta name="description" content="Calculate genetic distances"><meta property="og:description" content="Calculate genetic distances"><meta property="og:image" content="https://thewanglab.github.io/algatr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">algatr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference Manual</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/data_processing_vignette.html">Genetic data processing</a></li>
    <li><a class="dropdown-item" href="../articles/enviro_data_vignette.html">Environmental data processing</a></li>
    <li><a class="dropdown-item" href="../articles/masking_vignette.html">Masking</a></li>
    <li><a class="dropdown-item" href="../articles/gen_dist_vignette.html">Genetic distances</a></li>
    <li><a class="dropdown-item" href="../articles/TESS_vignette.html">TESS</a></li>
    <li><a class="dropdown-item" href="../articles/MMRR_vignette.html">MMRR</a></li>
    <li><a class="dropdown-item" href="../articles/GDM_vignette.html">GDM</a></li>
    <li><a class="dropdown-item" href="../articles/RDA_vignette.html">RDA</a></li>
    <li><a class="dropdown-item" href="../articles/LFMM_vignette.html">LFMM</a></li>
    <li><a class="dropdown-item" href="../articles/wingen_vignette.html">Wingen</a></li>
    <li><a class="dropdown-item" href="../articles/alazygatr_vignette.html">Alazygatr</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Calculate genetic distances</h1>

      <div class="d-none name"><code>gen_dist.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Calculate genetic distances</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gen_dist</span><span class="op">(</span></span>
<span>  gen <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dist_type <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>  plink_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plink_id_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  npc_selection <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  criticalpoint <span class="op">=</span> <span class="fl">2.0234</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gen">gen<a class="anchor" aria-label="anchor" href="#arg-gen"></a></dt>
<dd><p>path to vcf file, a <code>vcfR</code> type object, or a dosage matrix</p></dd>


<dt id="arg-dist-type">dist_type<a class="anchor" aria-label="anchor" href="#arg-dist-type"></a></dt>
<dd><p>the type of genetic distance to calculate (options: <code>"euclidean"</code> (default), <code>"bray_curtis"</code>, <code>"dps"</code> for proportion of shared alleles (requires vcf), <code>"plink"</code>, or <code>"pc"</code> for PC-based)</p></dd>


<dt id="arg-plink-file">plink_file<a class="anchor" aria-label="anchor" href="#arg-plink-file"></a></dt>
<dd><p>if <code>"plink"</code> dist_type is used, path to plink distance file (typically ".dist"; required only for calculating plink distance). File must be a <strong>square</strong> distance matrix.</p></dd>


<dt id="arg-plink-id-file">plink_id_file<a class="anchor" aria-label="anchor" href="#arg-plink-id-file"></a></dt>
<dd><p>if <code>"plink"</code> dist_type is used, path to plink id file (typically ".dist.id"; required only for calculating plink distance)</p></dd>


<dt id="arg-npc-selection">npc_selection<a class="anchor" aria-label="anchor" href="#arg-npc-selection"></a></dt>
<dd><p>if <code>dist_type = "pc"</code>, how to perform K selection (options: <code>"auto"</code> for automatic selection based on significant eigenvalues from Tracy-Widom test (default), or <code>"manual"</code> to examine PC screeplot and enter no. PCs into console)</p></dd>


<dt id="arg-criticalpoint">criticalpoint<a class="anchor" aria-label="anchor" href="#arg-criticalpoint"></a></dt>
<dd><p>if <code>dist_type = "pc"</code> used with <code>npc_selection = "auto"</code>, the critical point for the significance threshold for the Tracy-Widom test within the PCA (defaults to 2.0234 which corresponds to an alpha of 0.01)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>pairwise distance matrix for given distance metric</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Euclidean and Bray-Curtis distances calculated using the ecodist package: Goslee, S.C. and Urban, D.L. 2007. The ecodist package for dissimilarity-based analysis of ecological data. Journal of Statistical Software 22(7):1-19. DOI:10.18637/jss.v022.i07.
Proportions of shared alleles calculated using the adegenet package: Jombart T. and Ahmed I. (2011) adegenet 1.3-1: new tools for the analysis of genome-wide SNP data. Bioinformatics. doi:10.1093/bioinformatics/btr521.
For calculating proportions of shared alleles, missing values are ignored (i.e., prop shared alleles calculated from present values; no scaling performed)</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anusha Bishop, Anne Chambers, Ian Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

