<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • algatr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://thewanglab.github.io/algatr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">algatr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference Manual</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/data_processing_vignette.html">Genetic data processing</a></li>
    <li><a class="dropdown-item" href="../articles/enviro_data_vignette.html">Environmental data processing</a></li>
    <li><a class="dropdown-item" href="../articles/masking_vignette.html">Masking</a></li>
    <li><a class="dropdown-item" href="../articles/gen_dist_vignette.html">Genetic distances</a></li>
    <li><a class="dropdown-item" href="../articles/TESS_vignette.html">TESS</a></li>
    <li><a class="dropdown-item" href="../articles/MMRR_vignette.html">MMRR</a></li>
    <li><a class="dropdown-item" href="../articles/GDM_vignette.html">GDM</a></li>
    <li><a class="dropdown-item" href="../articles/RDA_vignette.html">RDA</a></li>
    <li><a class="dropdown-item" href="../articles/LFMM_vignette.html">LFMM</a></li>
    <li><a class="dropdown-item" href="../articles/wingen_vignette.html">Wingen</a></li>
    <li><a class="dropdown-item" href="../articles/alazygatr_vignette.html">Alazygatr</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="genetic-data-processing">Genetic data processing<a class="anchor" aria-label="anchor" href="#genetic-data-processing"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="data_processing_packages.html">data_processing_packages()</a></code>

        </dt>
        <dd>Install data processing packages</dd>

        <dt>

          <code><a href="gen_to_geno.html">gen_to_geno()</a></code>

        </dt>
        <dd>Convert dosage matrix or vcf to geno type object (N.B.: this only works for diploids!)</dd>

        <dt>

          <code><a href="geno_to_dosage.html">geno_to_dosage()</a></code>

        </dt>
        <dd>Convert lfmm/geno matrix to dosage matrix (N.B.: this only works for diploids!)</dd>

        <dt>

          <code><a href="ld_prune.html">ld_prune()</a></code>

        </dt>
        <dd>ld_prune prunes SNPs based on linkage disequilibrium using <code>SNPRelate</code> and <code>SeqArray</code> packages</dd>

        <dt>

          <code><a href="simple_impute.html">simple_impute()</a></code>

        </dt>
        <dd>Impute NA values NOTE: use extreme caution when using this form of simplistic imputation. We mainly provide this code for creating test datasets and highly discourage its use in analyses.</dd>

        <dt>

          <code><a href="str_impute.html">str_impute()</a></code>

        </dt>
        <dd>Imputation of missing values using population structure inferred with <code><a href="https://rdrr.io/pkg/LEA/man/main_sNMF.html" class="external-link">LEA::snmf</a></code></dd>

        <dt>

          <code><a href="vcf_to_dosage.html">vcf_to_dosage()</a></code>

        </dt>
        <dd>Convert a vcf to a dosage matrix</dd>
      </dl></div><div class="section level2">
      <h2 id="environmental-data-processing">Environmental data processing<a class="anchor" aria-label="anchor" href="#environmental-data-processing"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="envirodata_packages.html">envirodata_packages()</a></code>

        </dt>
        <dd>Install environmental and geographic data processing packages</dd>

        <dt>

          <code><a href="check_dists.html">check_dists()</a></code>

        </dt>
        <dd>Check geographic and environmental distances for collinearity</dd>

        <dt>

          <code><a href="check_env.html">check_env()</a></code>

        </dt>
        <dd>Check environmental layers for collinearity</dd>

        <dt>

          <code><a href="check_vals.html">check_vals()</a></code>

        </dt>
        <dd>Check extracted values for collinearity</dd>

        <dt>

          <code><a href="env_dist.html">env_dist()</a></code>

        </dt>
        <dd>Calculate distance between environmental vars</dd>

        <dt>

          <code><a href="geo_dist.html">geo_dist()</a></code>

        </dt>
        <dd>Calculate geographic distance between coordinates</dd>

        <dt>

          <code><a href="get_worldclim.html">get_worldclim()</a></code>

        </dt>
        <dd>Download and merge WorldClim data for study area</dd>

        <dt>

          <code><a href="rm_islands.html">rm_islands()</a></code>

        </dt>
        <dd>Remove islands from mapping</dd>
      </dl></div><div class="section level2">
      <h2 id="masking">Masking<a class="anchor" aria-label="anchor" href="#masking"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="extrap_mask.html">extrap_mask()</a></code> <code><a href="extrap_mask.html">range_mask()</a></code> <code><a href="extrap_mask.html">sd_mask()</a></code> <code><a href="extrap_mask.html">buffer_mask()</a></code> <code><a href="extrap_mask.html">chull_mask()</a></code>

        </dt>
        <dd>Create raster mask based on coordinates</dd>

        <dt>

          <code><a href="masking_packages.html">masking_packages()</a></code>

        </dt>
        <dd>Install masking packages</dd>
      </dl></div><div class="section level2">
      <h2 id="genetic-distances">Genetic distances<a class="anchor" aria-label="anchor" href="#genetic-distances"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="gen_dist.html">gen_dist()</a></code>

        </dt>
        <dd>Calculate genetic distances</dd>

        <dt>

          <code><a href="gen_dist_corr.html">gen_dist_corr()</a></code>

        </dt>
        <dd>Plot the relationship between two distance metrics</dd>

        <dt>

          <code><a href="gen_dist_hm.html">gen_dist_hm()</a></code>

        </dt>
        <dd>Make heatmap of genetic distances</dd>

        <dt>

          <code><a href="gen_dist_packages.html">gen_dist_packages()</a></code>

        </dt>
        <dd>Install genetic distance packages</dd>
      </dl></div><div class="section level2">
      <h2 id="tess">TESS<a class="anchor" aria-label="anchor" href="#tess"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="tess_barplot.html">tess_barplot()</a></code>

        </dt>
        <dd>Create TESS barplot</dd>

        <dt>

          <code><a href="tess_col_default.html">tess_col_default()</a></code>

        </dt>
        <dd>Create default TESS color palette</dd>

        <dt>

          <code><a href="tess_do_everything.html">tess_do_everything()</a></code>

        </dt>
        <dd>TESS function to do everything</dd>

        <dt>

          <code><a href="tess_ggbarplot.html">tess_ggbarplot()</a></code>

        </dt>
        <dd>Create TESS barplot using ggplot2</dd>

        <dt>

          <code><a href="tess_ggplot.html">tess_ggplot()</a></code>

        </dt>
        <dd>ggplot of TESS results</dd>

        <dt>

          <code><a href="tess_krig.html">tess_krig()</a></code>

        </dt>
        <dd>Krige admixture values</dd>

        <dt>

          <code><a href="tess_ktest.html">tess_ktest()</a></code>

        </dt>
        <dd>Test multiple K values</dd>

        <dt>

          <code><a href="tess_legend.html">tess_legend()</a></code>

        </dt>
        <dd>Create a custom legend for TESS maps</dd>

        <dt>

          <code><a href="tess_packages.html">tess_packages()</a></code>

        </dt>
        <dd>Install TESS packages</dd>

        <dt>

          <code><a href="tess_plot_allK.html">tess_plot_allK()</a></code>

        </dt>
        <dd>Plot all kriged Q values for each K</dd>

        <dt>

          <code><a href="bestK.html">bestK()</a></code>

        </dt>
        <dd>Best K Selection based on cross entropy</dd>

        <dt>

          <code><a href="geom_tess.html">geom_tess()</a></code>

        </dt>
        <dd>Create geom of TESS results that can be added to a ggplot object</dd>
      </dl></div><div class="section level2">
      <h2 id="mmrr">MMRR<a class="anchor" aria-label="anchor" href="#mmrr"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="mmrr_df.html">mmrr_df()</a></code>

        </dt>
        <dd>Make nice dataframe from MMRR results</dd>

        <dt>

          <code><a href="mmrr_do_everything.html">mmrr_do_everything()</a></code>

        </dt>
        <dd>MMRR function to do everything</dd>

        <dt>

          <code><a href="mmrr_packages.html">mmrr_packages()</a></code>

        </dt>
        <dd>Install MMRR packages</dd>

        <dt>

          <code><a href="mmrr_plot.html">mmrr_plot()</a></code>

        </dt>
        <dd>Plot MMRR results</dd>

        <dt>

          <code><a href="mmrr_run.html">mmrr_run()</a></code>

        </dt>
        <dd>Run MMRR and return model object</dd>

        <dt>

          <code><a href="mmrr_table.html">mmrr_table()</a></code>

        </dt>
        <dd>Create <code>gt</code> table of MMRR results</dd>

        <dt>

          <code><a href="mmrr_var_sel.html">mmrr_var_sel()</a></code>

        </dt>
        <dd>mmrr_var_sel performs MMRR with backward elimination variable selection</dd>

        <dt>

          <code><a href="MMRR.html">MMRR()</a></code>

        </dt>
        <dd>MMRR performs Multiple Matrix Regression with Randomization analysis</dd>

        <dt>

          <code><a href="unfold.html">unfold()</a></code>

        </dt>
        <dd>unfold converts the lower diagonal elements of a matrix into a vector</dd>
      </dl></div><div class="section level2">
      <h2 id="gdm">GDM<a class="anchor" aria-label="anchor" href="#gdm"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="gdm_coeffs.html">gdm_coeffs()</a></code>

        </dt>
        <dd>Get coefficients for each predictor</dd>

        <dt>

          <code><a href="gdm_df.html">gdm_df()</a></code>

        </dt>
        <dd>Create dataframe of GDM results</dd>

        <dt>

          <code><a href="gdm_do_everything.html">gdm_do_everything()</a></code>

        </dt>
        <dd>GDM function to do everything (fit model, get coefficients, make and save raster)</dd>

        <dt>

          <code><a href="gdm_format.html">gdm_format()</a></code>

        </dt>
        <dd>Format Data for Generalized Dissimilarity Modeling (GDM)</dd>

        <dt>

          <code><a href="gdm_map.html">gdm_map()</a></code>

        </dt>
        <dd>Make map from model</dd>

        <dt>

          <code><a href="gdm_packages.html">gdm_packages()</a></code>

        </dt>
        <dd>Install GDM packages</dd>

        <dt>

          <code><a href="gdm_plot_diss.html">gdm_plot_diss()</a></code>

        </dt>
        <dd>Plot compositional dissimilarity spline plots</dd>

        <dt>

          <code><a href="gdm_plot_isplines.html">gdm_plot_isplines()</a></code>

        </dt>
        <dd>Plot I-splines for each variable</dd>

        <dt>

          <code><a href="gdm_plot_vars.html">gdm_plot_vars()</a></code>

        </dt>
        <dd>Create a PCA plot for GDM</dd>

        <dt>

          <code><a href="gdm_run.html">gdm_run()</a></code>

        </dt>
        <dd>Run GDM and return model object</dd>

        <dt>

          <code><a href="gdm_table.html">gdm_table()</a></code>

        </dt>
        <dd>Create <code>gt</code> table of GDM results</dd>

        <dt>

          <code><a href="gdm_var_sel.html">gdm_var_sel()</a></code>

        </dt>
        <dd>Get best set of variables from a GDM model</dd>

        <dt>

          <code><a href="gdm_varimp_table.html">gdm_varimp_table()</a></code>

        </dt>
        <dd>Generate a Variable Importance Table for GDM Models</dd>

        <dt>

          <code><a href="scale01.html">scale01()</a></code>

        </dt>
        <dd>Scale genetic distances from 0 to 1</dd>

        <dt>

          <code><a href="scaleRGB.html">scaleRGB()</a></code>

        </dt>
        <dd>Scale three layers of environmental data to R, G, and B for mapping</dd>
      </dl></div><div class="section level2">
      <h2 id="rda">RDA<a class="anchor" aria-label="anchor" href="#rda"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="rda_cor.html">rda_cor()</a></code>

        </dt>
        <dd>Genotype-environment correlation test</dd>

        <dt>

          <code><a href="rda_do_everything.html">rda_do_everything()</a></code>

        </dt>
        <dd>RDA function to do everything</dd>

        <dt>

          <code><a href="rda_getoutliers.html">rda_getoutliers()</a></code>

        </dt>
        <dd>Get significant outliers from RDA model</dd>

        <dt>

          <code><a href="rda_packages.html">rda_packages()</a></code>

        </dt>
        <dd>Install RDA packages</dd>

        <dt>

          <code><a href="rda_plot.html">rda_plot()</a></code>

        </dt>
        <dd>Plot RDA results</dd>

        <dt>

          <code><a href="rda_run.html">rda_run()</a></code>

        </dt>
        <dd>Run RDA</dd>

        <dt>

          <code><a href="rda_table.html">rda_table()</a></code>

        </dt>
        <dd>Create <code>gt</code> table of RDA results</dd>

        <dt>

          <code><a href="rda_varpart.html">rda_varpart()</a></code>

        </dt>
        <dd>Partial RDA variance partitioning</dd>

        <dt>

          <code><a href="rda_varpart_table.html">rda_varpart_table()</a></code>

        </dt>
        <dd>Create <code>gt</code> table with RDA variance partitioning results</dd>
      </dl></div><div class="section level2">
      <h2 id="lfmm">LFMM<a class="anchor" aria-label="anchor" href="#lfmm"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="lfmm_df.html">lfmm_df()</a></code>

        </dt>
        <dd>Convert LFMM results into a tidy dataframe for downstream processing</dd>

        <dt>

          <code><a href="lfmm_do_everything.html">lfmm_do_everything()</a></code>

        </dt>
        <dd>LFMM function to do everything</dd>

        <dt>

          <code><a href="lfmm_manhattanplot.html">lfmm_manhattanplot()</a></code>

        </dt>
        <dd>LFMM Manhattan Plot</dd>

        <dt>

          <code><a href="lfmm_packages.html">lfmm_packages()</a></code>

        </dt>
        <dd>Install LFMM packages</dd>

        <dt>

          <code><a href="lfmm_qqplot.html">lfmm_qqplot()</a></code>

        </dt>
        <dd>LFMM QQplot</dd>

        <dt>

          <code><a href="lfmm_run.html">lfmm_run()</a></code>

        </dt>
        <dd>Run LFMM</dd>

        <dt>

          <code><a href="lfmm_table.html">lfmm_table()</a></code>

        </dt>
        <dd>Create <code>gt</code> table of LFMM results</dd>

        <dt>

          <code><a href="select_K.html">select_K()</a></code> <code><a href="select_K.html">select_K_tw()</a></code> <code><a href="select_K.html">select_K_elbow()</a></code> <code><a href="select_K.html">select_K_tess()</a></code> <code><a href="select_K.html">select_K_fc()</a></code>

        </dt>
        <dd>K selection</dd>

        <dt>

          <code><a href="quick_elbow.html">quick_elbow()</a></code>

        </dt>
        <dd>Quickly choose an elbow for a PC</dd>

        <dt>

          <code><a href="tw.html">tw()</a></code>

        </dt>
        <dd>Tracy–Widom test</dd>
      </dl></div><div class="section level2">
      <h2 id="wingen">Wingen<a class="anchor" aria-label="anchor" href="#wingen"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="wingen_do_everything.html">wingen_do_everything()</a></code>

        </dt>
        <dd>wingen function to do everything (preview and generate moving window maps, krige, and mask)</dd>

        <dt>

          <code><a href="wingen_packages.html">wingen_packages()</a></code>

        </dt>
        <dd>Install wingen packages</dd>
      </dl></div><div class="section level2">
      <h2 id="alazygatr">Alazygatr<a class="anchor" aria-label="anchor" href="#alazygatr"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="alazygatr_packages.html">alazygatr_packages()</a></code>

        </dt>
        <dd>Install alazygatr packages</dd>

        <dt>

          <code><a href="do_everything_for_me.html">do_everything_for_me()</a></code>

        </dt>
        <dd>Lazy run of all landscape genomic analyses contained within <code>algatr</code></dd>
      </dl></div><div class="section level2">
      <h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="load_algatr_example.html">load_algatr_example()</a></code>

        </dt>
        <dd>Load example data</dd>

        <dt>

          <code><a href="CA_env.html">CA_env</a></code>

        </dt>
        <dd>Example environmental data, calculated by performing a raster PCA on 18 bioclimatic variables for state of California</dd>

        <dt>

          <code><a href="liz_coords.html">liz_coords</a></code>

        </dt>
        <dd>Example coordinates from Bouzid et al. 2022</dd>

        <dt>

          <code><a href="liz_gendist.html">liz_gendist</a></code>

        </dt>
        <dd>Example genetic distance matrix, calculated with Plink using data from Bouzid et al. 2022</dd>

        <dt>

          <code><a href="liz_vcf.html">liz_vcf</a></code>

        </dt>
        <dd>Example VCF from Bouzid et al. 2022</dd>
      </dl></div><div class="section level2">
      <h2 id="other">Other<a class="anchor" aria-label="anchor" href="#other"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="coords_to_sf.html">coords_to_sf()</a></code>

        </dt>
        <dd>Convert from matrix, data frame, or sf to sf (sf is a pass through)</dd>

        <dt>

          <code><a href="coords_to_sp.html">coords_to_sp()</a></code>

        </dt>
        <dd>Convert from matrix, data frame, or sf to formatted sp</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anusha Bishop, Anne Chambers, Ian Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

