<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="algatr">
<title>Genetic data processing • algatr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genetic data processing">
<meta property="og:description" content="algatr">
<meta property="og:image" content="https://thewanglab.github.io/algatr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">algatr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference Manual</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <a class="dropdown-item" href="../articles/data_processing_vignette.html">Genetic data processing</a>
    <a class="dropdown-item" href="../articles/enviro_data_vignette.html">Environmental data processing</a>
    <a class="dropdown-item" href="../articles/masking_vignette.html">Masking</a>
    <a class="dropdown-item" href="../articles/gen_dist_vignette.html">Genetic distances</a>
    <a class="dropdown-item" href="../articles/TESS_vignette.html">TESS</a>
    <a class="dropdown-item" href="../articles/MMRR_vignette.html">MMRR</a>
    <a class="dropdown-item" href="../articles/GDM_vignette.html">GDM</a>
    <a class="dropdown-item" href="../articles/RDA_vignette.html">RDA</a>
    <a class="dropdown-item" href="../articles/LFMM_vignette.html">LFMM</a>
    <a class="dropdown-item" href="../articles/wingen_vignette.html">Wingen</a>
    <a class="dropdown-item" href="../articles/alazygatr_vignette.html">Alazygatr</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Genetic data processing</h1>
            
      
      
      <div class="d-none name"><code>data_processing_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install packages </span></span>
<span><span class="fu"><a href="../reference/data_processing_packages.html">data_processing_packages</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="processing-genomic-data-files">Processing genomic data files<a class="anchor" aria-label="anchor" href="#processing-genomic-data-files"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thewanglab.github.io/algatr/">algatr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span></span></code></pre></div>
<p>For landscape genomic analyses, we first need to process our input
data files in various ways. It is important to understand what this data
processing is doing, so in this vignette, we’ll do the following:</p>
<ul>
<li><p><code><a href="../reference/vcf_to_dosage.html">vcf_to_dosage()</a></code> to convert a vcf file to a dosage
matrix</p></li>
<li><p><code><a href="../reference/str_impute.html">str_impute()</a></code> and <code><a href="../reference/simple_impute.html">simple_impute()</a></code> to
impute missing data in two different ways</p></li>
<li><p><code><a href="../reference/ld_prune.html">ld_prune()</a></code> to prune variants that are in linkage
disequilibrium with one another; this primarily uses the
<code>snpgdsLDpruning()</code> function within the SNPRelate package (<a href="https://academic.oup.com/bioinformatics/article/28/24/3326/245844" class="external-link">Zheng
et al. 2012</a>; see <a href="https://www.bioconductor.org/packages/release/bioc/manuals/SNPRelate/man/SNPRelate.pdf" class="external-link">here</a>
for further information)</p></li>
</ul>
<div class="section level3">
<h3 id="converting-a-vcf-to-a-dosage-matrix-with-vcf_to_dosage">Converting a vcf to a dosage matrix with
<code>vcf_to_dosage()</code><a class="anchor" aria-label="anchor" href="#converting-a-vcf-to-a-dosage-matrix-with-vcf_to_dosage"></a>
</h3>
<hr>
<p>Recall the way genotypes are encoded within a vcf: <code>"0/0"</code>
denotes individual is homozygous for the reference allele,
<code>"0/1"</code> for heterozygote, and <code>"1/1"</code> for a
homozygote for the alternate allele; <code>NA</code> is missing
data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the example data</span></span>
<span><span class="fu"><a href="../reference/load_algatr_example.html">load_algatr_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; ---------------- example dataset ----------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span> </span></span>
<span><span class="co"><span style="color: #0000BB;">#&gt; Objects loaded:</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_vcf*</span> <span style="color: #BBBB00;">vcfR object (1000 loci x 53 samples)</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_gendist*</span> <span style="color: #BBBB00;">genetic distance matrix (Plink Distance)</span> </span></span>
<span><span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; *liz_coords*</span> <span style="color: #00BB00;">dataframe with x and y coordinates</span> </span></span>
<span><span class="co"><span style="color: #BB00BB; font-weight: bold;">#&gt; *CA_env*</span> <span style="color: #BB00BB;">RasterStack with example environmental layers</span> </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; -------------------------------------------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co">#&gt; </span></span>
<span></span>
<span><span class="co"># Look at genotypes for five individuals at eight sites:</span></span>
<span><span class="va">liz_vcf</span><span class="op">@</span><span class="va">gt</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="co">#&gt;      ALT3  BAR360 BLL5  BNT5  BOF1 </span></span>
<span><span class="co">#&gt; [1,] "0/0" "0/0"  "0/0" "0/0" "0/0"</span></span>
<span><span class="co">#&gt; [2,] "0/0" "0/0"  NA    "0/0" NA   </span></span>
<span><span class="co">#&gt; [3,] "0/0" "0/0"  "0/0" NA    NA   </span></span>
<span><span class="co">#&gt; [4,] "0/0" NA     "0/0" NA    NA   </span></span>
<span><span class="co">#&gt; [5,] "0/0" "0/0"  NA    "0/0" NA   </span></span>
<span><span class="co">#&gt; [6,] NA    "0/0"  "0/0" "0/0" NA   </span></span>
<span><span class="co">#&gt; [7,] NA    "0/0"  "0/0" NA    "0/0"</span></span>
<span><span class="co">#&gt; [8,] "1/1" "1/1"  "0/1" "0/0" "0/0"</span></span></code></pre></div>
<p>A dosage matrix contains each row as an individual (unlike a vcf) and
each column as a site. Importantly, genotypes are now encoded with a
single number. For a diploid organism, there are three possibilities for
coding genotypes in a dosage matrix: 0, 1, or 2 (corresponding to 0/0,
1/1, and 0/1 from the vcf, respectively). <code>NA</code> still
represents missing data. We can convert our vcf to a dosage matrix using
the <code><a href="../reference/vcf_to_dosage.html">vcf_to_dosage()</a></code> function:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dosage</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_to_dosage.html">vcf_to_dosage</a></span><span class="op">(</span><span class="va">liz_vcf</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required namespace: adegenet</span></span>
<span></span>
<span><span class="co"># Look at genotypes for five individuals at five sites:</span></span>
<span><span class="va">dosage</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span>
<span><span class="co">#&gt;        Locus_10 Locus_15 Locus_22 Locus_28 Locus_32 Locus_35 Locus_37 Locus_61</span></span>
<span><span class="co">#&gt; ALT3          0        0        0        0        0       NA       NA        2</span></span>
<span><span class="co">#&gt; BAR360        0        0        0       NA        0        0        0        2</span></span>
<span><span class="co">#&gt; BLL5          0       NA        0        0       NA        0        0        1</span></span>
<span><span class="co">#&gt; BNT5          0        0       NA       NA        0        0       NA        0</span></span>
<span><span class="co">#&gt; BOF1          0       NA       NA       NA       NA       NA        0        0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="imputation-of-missing-values">Imputation of missing values<a class="anchor" aria-label="anchor" href="#imputation-of-missing-values"></a>
</h3>
<hr>
<p>Because of their underlying statistical framework, several landscape
genomics methods cannot accept any missing data as input. Within algatr,
the two GEA methods (RDA and LFMM) are examples of such methods. To deal
with inevitable missing data, researchers typically perform imputation,
wherein values are assigned to missing data based on some criteria. For
example, the most simplistic imputation could be one in which missing
values are imputed based on the median value across all individuals; one
can imagine that this is extremely simplistic and assumes a lot about
the data and samples. There have been quite a few developments in
missing data imputation in recent years; we suggest you take a look at
<a href="https://doi.org/10.1534/g3.115.021667" class="external-link">Money et al. 2015</a>
and <a href="https://doi.org/10.1159/000489758" class="external-link">Shi et al. 2018</a> as a
starting point for examples of how this can be done.</p>
<p>algatr has the functionality to impute missing values in two ways:
the simplistic median-based approach described above using
<code><a href="../reference/simple_impute.html">simple_impute()</a></code> and a more sophisticated imputation method,
<code><a href="../reference/str_impute.html">str_impute()</a></code> that uses underlying population structure to
assign missing values. Let’s first take a look at
<code><a href="../reference/simple_impute.html">simple_impute()</a></code> using the dosage matrix we just generated
in the step prior to this one:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Are there NAs in the data?</span></span>
<span><span class="va">dosage</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;        Locus_10 Locus_15 Locus_22 Locus_28 Locus_32</span></span>
<span><span class="co">#&gt; ALT3          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BAR360        0        0        0       NA        0</span></span>
<span><span class="co">#&gt; BLL5          0       NA        0        0       NA</span></span>
<span><span class="co">#&gt; BNT5          0        0       NA       NA        0</span></span>
<span><span class="co">#&gt; BOF1          0       NA       NA       NA       NA</span></span>
<span><span class="va">simple_dos</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simple_impute.html">simple_impute</a></span><span class="op">(</span><span class="va">dosage</span><span class="op">)</span></span>
<span><span class="co"># Check that NAs are gone</span></span>
<span><span class="va">simple_dos</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;        Locus_10 Locus_15 Locus_22 Locus_28 Locus_32</span></span>
<span><span class="co">#&gt; ALT3          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BAR360        0        0        0        0        0</span></span>
<span><span class="co">#&gt; BLL5          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BNT5          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BOF1          0        0        0        0        0</span></span></code></pre></div>
<p>Now, let’s use <code><a href="../reference/str_impute.html">str_impute()</a></code> to impute missing values
based on population structure. This function does two main things: (1)
it runs SNMF to generate ancestry coefficients based on a single or
range of K-values; and (2) it uses the <code><a href="https://rdrr.io/pkg/LEA/man/impute.html" class="external-link">LEA::impute()</a></code>
function to impute missing values based on the SNMF results. In this
way, this imputation method is taking into consideration population
structure and assuming that individuals sharing the same ancestral
clusters are more similar to one another; certainly, this is a better
assumption than imputing based on the median across all individuals.</p>
<p>There are a number of arguments in the <code><a href="../reference/str_impute.html">str_impute()</a></code>
function, most of which inherit from <code><a href="https://rdrr.io/pkg/LEA/man/main_sNMF.html" class="external-link">LEA::snmf()</a></code>:</p>
<ul>
<li><p><code>gen</code> specifies the genotype data and can be a dosage
matrix, vcfR object, path to vcf file, or an sNMF project (i.e., object
of type snmfProject) if you’ve already run SNMF and want to use that for
imputation</p></li>
<li><p><code>K</code> specifies a single value or range of K-values for
sNMF to be run</p></li>
<li><p><code>entropy</code> whether to calculate cross-entropy criteria
(defaults to TRUE)</p></li>
<li><p><code>repetitions</code> specifies the number of repetitions (or
runs) to do for each K-value (defaults to 10)</p></li>
<li><p><code>project</code> specifies whether you want to start a new
sNMF project (<code>"new"</code>; default), whether you want to continue
an existing SNMF project (<code>"continue"</code>) or overwrite existing
results (<code>"force"</code>)</p></li>
<li><p><code>quiet</code> produces a plot of the cross-entropy scores
for each K value from sNMF, if set to FALSE. The default is for nothing
to be printed (quiet = TRUE). The best K value from the sNMF results is
based on minimizing cross-entropy scores</p></li>
<li><p><code>save_output</code> specifies whether you would like output
files to be saved to file (a string) or not (left NULL; this is the
default). If a string is provided, output files will be saved with this
argument’s string as a prefix. Files to be saved include SNMF results,
the original dosage matrix as .geno and .lfmm files, and an imputed
.lfmm file</p></li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">str_dos</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/str_impute.html">str_impute</a></span><span class="op">(</span>gen <span class="op">=</span> <span class="va">dosage</span>, K <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, entropy <span class="op">=</span> <span class="cn">TRUE</span>, repetitions <span class="op">=</span> <span class="fl">3</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span>, save_output <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_processing_vignette_files/figure-html/str%20impute-1.png" width="700" style="display: block; margin: auto;"><img src="data_processing_vignette_files/figure-html/str%20impute-2.png" width="700" style="display: block; margin: auto;"><img src="data_processing_vignette_files/figure-html/str%20impute-3.png" width="700" style="display: block; margin: auto;"><img src="data_processing_vignette_files/figure-html/str%20impute-4.png" width="700" style="display: block; margin: auto;"></p>
<p>Let’s take a look at how the two imputation methods compare by
calculating genetic distances for each and then using the
<code><a href="../reference/gen_dist_corr.html">gen_dist_corr()</a></code> function which plots the relationship
between two distance metrics (please see the genetic distances vignette
for further information).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">str_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/coerce.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu">ecodist</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ecodist/man/distance.html" class="external-link">distance</a></span><span class="op">(</span><span class="va">str_dos</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">simple_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/coerce.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu">ecodist</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ecodist/man/distance.html" class="external-link">distance</a></span><span class="op">(</span><span class="va">simple_dos</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gen_dist_corr.html">gen_dist_corr</a></span><span class="op">(</span>dist_x <span class="op">=</span> <span class="va">simple_dist</span>, dist_y <span class="op">=</span> <span class="va">str_dist</span>, metric_name_x <span class="op">=</span> <span class="st">"simple"</span>, metric_name_y <span class="op">=</span> <span class="st">"SNMF"</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_processing_vignette_files/figure-html/compare%20impute-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="removing-sites-that-are-in-linkage-disequilibrium-using-ld_prune">Removing sites that are in linkage disequilibrium using
<code>ld_prune()</code><a class="anchor" aria-label="anchor" href="#removing-sites-that-are-in-linkage-disequilibrium-using-ld_prune"></a>
</h3>
<hr>
<p>For some analyses, we may want to prune out any linked sites from our
data to not bias results. This is particularly the case with methods
that estimate population structure (e.g., TESS), as population structure
can be overinflated as more SNPs appear to independently support the
same pattern. For other analyses - such as genotype-environment
association methods like RDA and LFMM - SNPs in LD are sometimes removed
prior to running such analyses, but we do not recommend this because
SNPs most strongly associated with environmental variables may have been
pruned out. However, failing to account for linkage may result in
overinflated conclusions about environmental associations. A solution to
this is to perform LD-pruning after having run GEA methods; one approach
to do so is to retain SNPs with the strongest association (i.e., lowest
p-value) and removing all others that are in linkage with that
particular SNP (see the approach described in Capblancq &amp; Forester
2021 for more information).</p>
<p>To perform LD-pruning, we can use the <code><a href="../reference/ld_prune.html">ld_prune()</a></code>
function (which largely uses the <code>snpgdsLDpruning()</code> function
within the SNPRelate package). The function works by first removing any
variants that occur below a given minor allele frequency, and then by
pruning any variants that have a given correlation using a sliding
window approach.</p>
<p>The main arguments within this function are as follows:</p>
<ul>
<li><p><code>ld.threshold</code> specifies the LD pruning threshold (the
default is the correlation coefficient value; thus, the higher the
<code>ld.threshold</code> value, fewer variants will be pruned)</p></li>
<li><p><code>slide.max.n</code> specifies the maximum number of SNPs in
the sliding window (defaults to 100)</p></li>
<li><p><code>maf</code> specifies the minor allele frequency (defaults
to 0.05)</p></li>
<li><p><code>vcf</code> specifies the path to the vcf file that you
would like to be pruned</p></li>
<li><p><code>out_name</code> specifies the prefix of output file
names</p></li>
<li><p><code>out_format</code> specifies the file type (options are
<code>"vcf"</code> for vcf and gds files [default], or
<code>"plink"</code> for ped and map files)</p></li>
<li><p><code>save_output</code> specifies whether you would like a new
directory made containing all output (and intermediate) files; default
is <code>TRUE.</code> If set to <code>FALSE</code>, the function will
simply return a vcf object.</p></li>
</ul>
<p>The way to call this function is as follows:</p>
<p><code>vcf_ldpruned &lt;- ld_prune(vcf = "test.vcf", out_name = "liz", out_format = "vcf", ld.threshold = 0.9, slide.max.n = 10)</code></p>
<p>We will not be running this function within the vignette to avoid
generating intermediate files, but the example line above creates a new
directory called liz_LDpruned, and it contains five files: a LOGFILE,
with information on how many sites were pruned and how many remain,
three gds files (pre- and post-pruned SNP GDS files and a GDS SeqArray
file), and the pruned vcf. All of these output files are named such that
the <code>ld.threshold</code> and <code>slide.max.n</code> parameter
settings are contained in file names.</p>
</div>
<div class="section level3">
<h3 id="additional-options-for-data-visualization">Additional options for data visualization<a class="anchor" aria-label="anchor" href="#additional-options-for-data-visualization"></a>
</h3>
<hr>
<p>An alternate way to visualize matrices that can be helpful when you
have many overlapping data points is using the <code><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html" class="external-link">geom_hex()</a></code>
function in the ggplot2 package. This plotting style bins data points
into hexagons that are colorized based on the number of data points
contained within the hexagon.</p>
<p>Let’s take a look at this style of plot using the comparison between
imputed genetic distances we generated earlier in this vignette. To
makes things easy, let’s re-run the <code><a href="../reference/gen_dist_corr.html">gen_dist_corr()</a></code>
function and extract the data from the output. This way, it’s already
been processed for us.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/edzer/hexbin" class="external-link">"hexbin"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"hexbin"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/edzer/hexbin" class="external-link">hexbin</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Gather data</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_dist_corr.html">gen_dist_corr</a></span><span class="op">(</span>dist_x <span class="op">=</span> <span class="va">simple_dist</span>, dist_y <span class="op">=</span> <span class="va">str_dist</span>, metric_name_x <span class="op">=</span> <span class="st">"simple"</span>, metric_name_y <span class="op">=</span> <span class="st">"SNMF"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now, build plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot</span><span class="op">$</span><span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">simple</span>, y <span class="op">=</span> <span class="va">SNMF</span><span class="op">)</span> <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html" class="external-link">geom_hex</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_fill_continuous</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_processing_vignette_files/figure-html/geom%20hex-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="additional-documentation-and-citations">Additional documentation and citations<a class="anchor" aria-label="anchor" href="#additional-documentation-and-citations"></a>
</h3>
<hr>
<table class="table">
<colgroup>
<col width="26%">
<col width="39%">
<col width="34%">
</colgroup>
<thead><tr class="header">
<th></th>
<th>Citation/URL</th>
<th>Details</th>
</tr></thead>
<tbody><tr class="odd">
<td>Associated code</td>
<td>
<a href="https://academic.oup.com/bioinformatics/article/28/24/3326/245844" class="external-link">Zheng
et al. 2012</a>; further details <a href="https://www.bioconductor.org/packages/release/bioc/manuals/SNPRelate/man/SNPRelate.pdf" class="external-link">here</a>
</td>
<td>algatr uses the <code>snpgdsLDpruning()</code> function within the
SNPRelate package for LD-pruning</td>
</tr></tbody>
</table>
<p>Retrieve SNPRelate’s vignette:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"SNPRelate"</span><span class="op">)</span></span>
<span><span class="co">#&gt; starting httpd help server ... done</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Anusha Bishop, Anne Chambers, Ian Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
