<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="algatr">
<title>GDM • algatr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GDM">
<meta property="og:description" content="algatr">
<meta property="og:image" content="https://thewanglab.github.io/algatr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">algatr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/data_processing_vignette.html">Genetic data processing</a>
    <a class="dropdown-item" href="../articles/enviro_data_vignette.html">Environmental data processing</a>
    <a class="dropdown-item" href="../articles/Masking_vignette.html">Masking</a>
    <a class="dropdown-item" href="../articles/gen_dist_vignette.html">Genetic distances</a>
    <a class="dropdown-item" href="../articles/TESS_vignette.html">TESS</a>
    <a class="dropdown-item" href="../articles/MMRR_vignette.html">MMRR</a>
    <a class="dropdown-item" href="../articles/GDM_vignette.html">GDM</a>
    <a class="dropdown-item" href="../articles/RDA_vignette.html">RDA</a>
    <a class="dropdown-item" href="../articles/LFMM_vignette.html">LFMM</a>
    <a class="dropdown-item" href="../articles/wingen_vignette.html">Wingen</a>
    <a class="dropdown-item" href="../articles/alazygatr_vignette.html">Alazygatr</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>GDM</h1>
            
      
      
      <div class="d-none name"><code>GDM_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install packages</span></span>
<span><span class="fu"><a href="../reference/gdm_packages.html">gdm_packages</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="generalized-dissimilarity-modeling-gdm">Generalized dissimilarity modeling (GDM)<a class="anchor" aria-label="anchor" href="#generalized-dissimilarity-modeling-gdm"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thewanglab.github.io/algatr/">algatr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/raster" class="external-link">raster</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span></code></pre></div>
<p><strong>If using GDM, please cite the following: for the GDM method,
Ferrier S., Manion G., Elith J., Richardson K. (2007) Using generalized
dissimilarity modelling to analyse and predict patterns of beta
diversity in regional biodiversity assessment, and for the
<code>gdm</code> package, Fitzpatrick M., Mokany K., Manion G.,
Nieto-Lugilde D., Ferrier S. (2022). gdm: Generalized Dissimilarity
Modeling. R package version 1.5.0-9.1.</strong></p>
<p>Generalized dissimilarity modeling (GDM) is a matrix regression
method in which explanatory variables (in our case, genetic data, in the
form of a distance matrix) is regressed against a response matrix (e.g.,
environmental variables for locations from which samples were obtained
and geographic distances between those locations). GDM calculates the
compositional dissimilarity between pairs of sites, and importantly
allows for nonlinear relationships to be modeled.</p>
<p>For additional information on GDM, please see <a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1472-4642.2007.00341.x" class="external-link">Ferrier
et al. 2007</a> for a description of its basic use in estimating
patterns of beta diversity, <a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1365-294X.2010.04684.x" class="external-link">Freedman
et al. 2010</a> for a classic example of its use, and <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/ele.12376#:~:text=Community%2Dlevel%20modelling%20of%20genomic,assessments%20of%20climate%20change%20vulnerability." class="external-link">Fitzpatrick
&amp; Keller 2015</a> and <a href="https://onlinelibrary.wiley.com/doi/full/10.1111/geb.13459" class="external-link">Mokany
et al. 2022</a> for perspectives on its application. Finally, our code
primarily uses the gdm package which has excellent documentation (see <a href="https://cran.r-project.org/web/packages/gdm/gdm.pdf" class="external-link">here</a>)
including a thorough <a href="https://cran.microsoft.com/snapshot/2018-01-13/web/packages/gdm/vignettes/gdmVignette.pdf" class="external-link">vignette</a>
that describes some of the theory behind GDM.</p>
<p>Within algatr, there is one main function to perform a GDM analysis:
<code><a href="../reference/gdm_do_everything.html">gdm_do_everything()</a></code>. This function runs the GDM (using the
<code>gdm()</code> function within the gdm package) and allows a user to
run a GDM with all variables, or with model selection to choose the
best-supported variables.</p>
<p>Contained within <code><a href="../reference/gdm_do_everything.html">gdm_do_everything()</a></code> are the following
functions, which we’ll walk through here:</p>
<ul>
<li><p><code><a href="../reference/gdm_run.html">gdm_run()</a></code> runs the GDM itself</p></li>
<li><p><code><a href="../reference/gdm_df.html">gdm_df()</a></code> generates statistics and coefficients for
predictor variables</p></li>
<li><p><code><a href="../reference/gdm_plot_isplines.html">gdm_plot_isplines()</a></code> plots fitted I-splines for each
variable</p></li>
<li><p><code><a href="../reference/gdm_map.html">gdm_map()</a></code> generates a PCA map of compositional
dissimilarity based on GDM results</p></li>
</ul>
<div class="section level4">
<h4 id="assumptions">
<em>Assumptions</em><a class="anchor" aria-label="anchor" href="#assumptions"></a>
</h4>
<p>There are a few assumptions built within this function that the user
must be aware of: (1) the coords and gendist files <em>must</em> have
the same ordering of individuals; there isn’t a check for this, and (2)
this function assumes that each individual has its own sampling
coordinates (even if population-based sampling was performed).</p>
<p>Running a GDM analysis in algatr will work for both individual-based
(i.e., one individual per sampling locality or population) or
population-based sampling. The test dataset was performed using
individual-based sampling, and so this vignette will walk users through
running GDM with that sampling scheme. For population-based sampling,
users should provide allele frequency or site-based genetic distances
with site locations rather than dosage matrices with individual
locations. Alternatively, users can select a single representative
sample from each population to mimic individual-based sampling, if so
desired; the most important consideration is to ensure that genetic data
match sampling locations. The remainder of the analysis is
identical.</p>
</div>
<div class="section level4">
<h4 id="read-in-and-process-input-data">Read in and process input data<a class="anchor" aria-label="anchor" href="#read-in-and-process-input-data"></a>
</h4>
<p>Let’s first load the objects within the example dataset.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/load_algatr_example.html">load_algatr_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; ---------------- example dataset ----------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span> </span></span>
<span><span class="co"><span style="color: #0000BB;">#&gt; Objects loaded:</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_vcf*</span> <span style="color: #BBBB00;">vcfR object (1000 loci x 53 samples)</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_gendist*</span> <span style="color: #BBBB00;">genetic distance matrix (Plink Distance)</span> </span></span>
<span><span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; *liz_coords*</span> <span style="color: #00BB00;">dataframe with x and y coordinates</span> </span></span>
<span><span class="co"><span style="color: #BB00BB; font-weight: bold;">#&gt; *CA_env*</span> <span style="color: #BB00BB;">RasterStack with PC environmental layers</span> </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; -------------------------------------------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co">#&gt; </span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extract-environmental-variables">Extract environmental variables<a class="anchor" aria-label="anchor" href="#extract-environmental-variables"></a>
</h4>
<p>To run GDM, we need to generate a site-pair table in which extracted
you have extracted variable values for each site. Because we only have
environmental layers for our example dataset, we need to extract the
environmental values from these layers for each sampling locality in
which lizards were collected:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">env</span> <span class="op">&lt;-</span> <span class="fu">raster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/extract.html" class="external-link">extract</a></span><span class="op">(</span><span class="va">CA_env</span>, <span class="va">liz_coords</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-gdm-with-gdm_run">Run GDM with <code>gdm_run()</code><a class="anchor" aria-label="anchor" href="#run-gdm-with-gdm_run"></a>
</h3>
<hr>
<div class="section level4">
<h4 id="gdm-with-all-variables">GDM with all variables<a class="anchor" aria-label="anchor" href="#gdm-with-all-variables"></a>
</h4>
<p>Given that GDM is a regression method, the “full” model (i.e.,
including all predictor variables) will include all environmental layers
in addition to geographic distance, which is also considered a
predictor. Thus, in this example, the maximum number of variables you
can end up with that are significant is four (three enviro PCs +
geographic distance). We can specify running the full model using the
<code>model</code> argument. Extracted environmental values for each
sampling coordinate are specified using the <code>env</code> argument,
and if genetic distances are not bounded by 0-1, they must be scaled
using the <code>scale_gendist</code> argument. Keep in mind that several
arguments are only for model selection and so will not be used.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdm_full</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdm_run.html">gdm_run</a></span><span class="op">(</span></span>
<span>  gendist <span class="op">=</span> <span class="va">liz_gendist</span>,</span>
<span>  coords <span class="op">=</span> <span class="va">liz_coords</span>,</span>
<span>  env <span class="op">=</span> <span class="va">env</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"full"</span>,</span>
<span>  scale_gendist <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="gdm-with-model-selection">GDM with model selection<a class="anchor" aria-label="anchor" href="#gdm-with-model-selection"></a>
</h4>
<p>We can also run a GDM using only significant variables (i.e., a
“best” model). In this case, a variable selection process is done using
a permutation test to assess significance: predictor variables are
randomly permuted between sites and the deviance explained from the
fitted model is compared to that obtained with random permutation. It is
important to be aware that GDM (as implemented in the gdm package)
always considers geographic distance as a variable and does not, by
default, allow a user to perform variable selection without geographic
distances considered. algatr has adjusted this functionality and will do
variable selection on all predictor variables, including geographic
distances.</p>
<p>There are a few additional considerations to make if running GDM with
model selection. The first is the number of permutations performed (the
<code>nperm</code> argument); these permutations represent the number of
times site-pair tables are permuted to perform a backwards elimination
procedure for variable selection. The next argument is <code>sig</code>,
specifying the significance threshold (alpha value).</p>
<p><em>N.B.: Sometimes, the <code><a href="../reference/gdm_run.html">gdm_run()</a></code> function will return
NULL, implying that none of the predictor variables are significantly
associated with genetic distances. Be sure to run adequate numbers of
permutations; occasionally NULL will be returned when insufficient
permutations were specified.</em></p>
<p>See below for an example of how model selection would be run; this
code is not run within the vignette because there is no significant GDM
model for the algatr test dataset.</p>
<pre><code><span><span class="va">gdm_best</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdm_run.html">gdm_run</a></span><span class="op">(</span>gendist <span class="op">=</span> <span class="va">liz_gendist</span>, </span>
<span>                    coords <span class="op">=</span> <span class="va">liz_coords</span>, </span>
<span>                    env <span class="op">=</span> <span class="va">env</span>, </span>
<span>                    model <span class="op">=</span> <span class="st">"best"</span>, </span>
<span>                    scale_gendist <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    nperm <span class="op">=</span> <span class="fl">500</span>, </span>
<span>                    sig <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Look at p-values</span></span>
<span><span class="va">gdm_best</span><span class="op">$</span><span class="va">pvalues</span></span>
<span><span class="va">gdm_best</span><span class="op">$</span><span class="va">varimp</span></span></code></pre>
<p>Within the resulting object(<code>gdm_best</code>), the
<code>pvalues</code> and <code>varimp</code> elements are no longer NULL
(as they were with the full model); this is because model selection was
performed. <code>pvalues</code> indicate the significance of each
environmental variable with genetic distance (given the user-defined
threshold), and <code>varimp</code> contains information about variable
importance and relevant statistics about the model.</p>
</div>
</div>
<div class="section level3">
<h3 id="interpreting-gdm-results">Interpreting GDM results<a class="anchor" aria-label="anchor" href="#interpreting-gdm-results"></a>
</h3>
<hr>
<p>The resulting object from <code><a href="../reference/gdm_run.html">gdm_run()</a></code> contains three
elements: <code>model</code>, containing information on the model that
was run, <code>pvalues</code>, and <code>varimp</code>. For the full
model, these latter two elements are empty because no model selection
was performed.</p>
<p>Let’s take a closer look at the <code>model</code> element. The
relevant items of this list are <code>predictors</code> (the predictor
variables that were considered in the model), <code>splines</code>
(three I-splines for each variable; see below for more information), and
<code>coefficients</code> (coefficients for each of the predictor
I-splines; i.e., 12 total). Predictor variables that have the highest
summed coefficient values are those that have most influence on
predicted dissimilarity. The predicted ecological distances are given
within the <code>ecological</code> element, predicted dissimilarity
within the <code>predicted</code> element, and observed compositional
dissimilarity within the <code>observed</code> element.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gdm_full</span><span class="op">$</span><span class="va">model</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] GDM Modelling Summary</span></span>
<span><span class="co">#&gt; [1] Creation Date:  Wed Nov 15 16:45:03 2023</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] Name:  $        Name:  gdm_full Name:  model   </span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] Data:  gdmData</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] Samples:  1378</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] Geographical distance used in model fitting?  TRUE</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] NULL Deviance:  131.531</span></span>
<span><span class="co">#&gt; [1] GDM Deviance:  83.008</span></span>
<span><span class="co">#&gt; [1] Percent Deviance Explained:  36.89</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] Intercept:  0.363</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] PREDICTOR ORDER BY SUM OF I-SPLINE COEFFICIENTS:</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] Predictor 1: Geographic</span></span>
<span><span class="co">#&gt; [1] Splines: 3</span></span>
<span><span class="co">#&gt; [1] Min Knot: 0</span></span>
<span><span class="co">#&gt; [1] 50% Knot: 3.203</span></span>
<span><span class="co">#&gt; [1] Max Knot: 11.074</span></span>
<span><span class="co">#&gt; [1] Coefficient[1]: 0.44</span></span>
<span><span class="co">#&gt; [1] Coefficient[2]: 0.464</span></span>
<span><span class="co">#&gt; [1] Coefficient[3]: 0.447</span></span>
<span><span class="co">#&gt; [1] Sum of coefficients for Geographic: 1.351</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] Predictor 2: CA_rPCA2</span></span>
<span><span class="co">#&gt; [1] Splines: 3</span></span>
<span><span class="co">#&gt; [1] Min Knot: -3.939</span></span>
<span><span class="co">#&gt; [1] 50% Knot: 0.413</span></span>
<span><span class="co">#&gt; [1] Max Knot: 5.107</span></span>
<span><span class="co">#&gt; [1] Coefficient[1]: 0.059</span></span>
<span><span class="co">#&gt; [1] Coefficient[2]: 0</span></span>
<span><span class="co">#&gt; [1] Coefficient[3]: 0</span></span>
<span><span class="co">#&gt; [1] Sum of coefficients for CA_rPCA2: 0.059</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] Predictor 3: CA_rPCA3</span></span>
<span><span class="co">#&gt; [1] Splines: 3</span></span>
<span><span class="co">#&gt; [1] Min Knot: -4.138</span></span>
<span><span class="co">#&gt; [1] 50% Knot: -0.457</span></span>
<span><span class="co">#&gt; [1] Max Knot: 2.475</span></span>
<span><span class="co">#&gt; [1] Coefficient[1]: 0</span></span>
<span><span class="co">#&gt; [1] Coefficient[2]: 0.021</span></span>
<span><span class="co">#&gt; [1] Coefficient[3]: 0</span></span>
<span><span class="co">#&gt; [1] Sum of coefficients for CA_rPCA3: 0.021</span></span>
<span><span class="co">#&gt; [1] </span></span>
<span><span class="co">#&gt; [1] Predictor 4: CA_rPCA1</span></span>
<span><span class="co">#&gt; [1] Splines: 3</span></span>
<span><span class="co">#&gt; [1] Min Knot: -5.466</span></span>
<span><span class="co">#&gt; [1] 50% Knot: -0.914</span></span>
<span><span class="co">#&gt; [1] Max Knot: 3.849</span></span>
<span><span class="co">#&gt; [1] Coefficient[1]: 0</span></span>
<span><span class="co">#&gt; [1] Coefficient[2]: 0</span></span>
<span><span class="co">#&gt; [1] Coefficient[3]: 0</span></span>
<span><span class="co">#&gt; [1] Sum of coefficients for CA_rPCA1: 0</span></span></code></pre></div>
<p>Let’s assume we want to move forward with examining the results from
the full model. We can look at the relationships between both predicted
ecological distance (the raw predictor) and predicted compositional
dissimilarity against observed compositional dissimilarity using the
<code><a href="../reference/gdm_plot_diss.html">gdm_plot_diss()</a></code> function. Within these plots, each data
point corresponds to a site-pair comparison, and the lines are obtained
by fitting (applying a link function) to the linear predictor.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdm_plot_diss.html">gdm_plot_diss</a></span><span class="op">(</span><span class="va">gdm_full</span><span class="op">$</span><span class="va">model</span><span class="op">)</span></span></code></pre></div>
<p><img src="GDM_vignette_files/figure-html/exp%20and%20obs%20dists-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="section level4">
<h4 id="plotting-fitted-i-splines-for-variables">Plotting fitted I-splines for variables<a class="anchor" aria-label="anchor" href="#plotting-fitted-i-splines-for-variables"></a>
</h4>
<p>Let’s plot the GDM-fitted functions for our full model. These
functions are fitted I-splines that relate each predictor variable to
the genetic distance data. Briefly, the maximum height of the curve
indicates the contribution of that predictor variable to changes
(dissimilarity) in genetic distances, and the shape of the curve
provides information on how genetic distances change across an
environment gradient for that predictor variable. The y-axis (“partial
regression distance”) relates the genetic distances to the variable when
all other variables are held constant. We can look at these I-splines
using the <code><a href="../reference/gdm_plot_isplines.html">gdm_plot_isplines()</a></code> function. In general, three
I-splines have been found to be sufficient to capture dissimilarity
across a gradient and avoids over-fitting (see <a href="https://onlinelibrary.wiley.com/doi/full/10.1111/geb.13459" class="external-link">Mokany
et al. 2022</a> for further information). As we can see from the below
plots, PCA1 has no relationship with genetic distance, which is also why
the coefficient for this variable is 0.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gdm_plot_isplines.html">gdm_plot_isplines</a></span><span class="op">(</span><span class="va">gdm_full</span><span class="op">$</span><span class="va">model</span><span class="op">)</span></span></code></pre></div>
<p><img src="GDM_vignette_files/figure-html/isplines-1.png" width="288"><img src="GDM_vignette_files/figure-html/isplines-2.png" width="288"><img src="GDM_vignette_files/figure-html/isplines-3.png" width="288"><img src="GDM_vignette_files/figure-html/isplines-4.png" width="288"></p>
</div>
<div class="section level4">
<h4 id="table-of-gdm-statistics">Table of GDM statistics<a class="anchor" aria-label="anchor" href="#table-of-gdm-statistics"></a>
</h4>
<p>We can generate a nice table with relevant GDM statistics to report
using the <code><a href="../reference/gdm_table.html">gdm_table()</a></code> function. To understand the relative
contributions explained by IBE (each of the environmental predictor
variables) and IBD (geographic distance), we sum the three I-spline
coefficients for each variable; non-zero sums are variables that are
significantly associated with genetic dissimilarity. The coefficients
contained in this table are the sum of the three I-spline coefficients
we saw above when we looked at the <code>coefficients</code> element
within the GDM model.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdm_table.html">gdm_table</a></span><span class="op">(</span><span class="va">gdm_full</span><span class="op">)</span></span></code></pre></div>
<div id="pkevpibsuj" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#pkevpibsuj table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#pkevpibsuj thead, #pkevpibsuj tbody, #pkevpibsuj tfoot, #pkevpibsuj tr, #pkevpibsuj td, #pkevpibsuj th {
  border-style: none;
}

#pkevpibsuj p {
  margin: 0;
  padding: 0;
}

#pkevpibsuj .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#pkevpibsuj .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#pkevpibsuj .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#pkevpibsuj .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#pkevpibsuj .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#pkevpibsuj .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#pkevpibsuj .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#pkevpibsuj .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#pkevpibsuj .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#pkevpibsuj .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#pkevpibsuj .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#pkevpibsuj .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#pkevpibsuj .gt_spanner_row {
  border-bottom-style: hidden;
}

#pkevpibsuj .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#pkevpibsuj .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#pkevpibsuj .gt_from_md > :first-child {
  margin-top: 0;
}

#pkevpibsuj .gt_from_md > :last-child {
  margin-bottom: 0;
}

#pkevpibsuj .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#pkevpibsuj .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#pkevpibsuj .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#pkevpibsuj .gt_row_group_first td {
  border-top-width: 2px;
}

#pkevpibsuj .gt_row_group_first th {
  border-top-width: 2px;
}

#pkevpibsuj .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#pkevpibsuj .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#pkevpibsuj .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#pkevpibsuj .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#pkevpibsuj .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#pkevpibsuj .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#pkevpibsuj .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#pkevpibsuj .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#pkevpibsuj .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#pkevpibsuj .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#pkevpibsuj .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#pkevpibsuj .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#pkevpibsuj .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#pkevpibsuj .gt_left {
  text-align: left;
}

#pkevpibsuj .gt_center {
  text-align: center;
}

#pkevpibsuj .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#pkevpibsuj .gt_font_normal {
  font-weight: normal;
}

#pkevpibsuj .gt_font_bold {
  font-weight: bold;
}

#pkevpibsuj .gt_font_italic {
  font-style: italic;
}

#pkevpibsuj .gt_super {
  font-size: 65%;
}

#pkevpibsuj .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#pkevpibsuj .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#pkevpibsuj .gt_indent_1 {
  text-indent: 5px;
}

#pkevpibsuj .gt_indent_2 {
  text-indent: 10px;
}

#pkevpibsuj .gt_indent_3 {
  text-indent: 15px;
}

#pkevpibsuj .gt_indent_4 {
  text-indent: 20px;
}

#pkevpibsuj .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="predictor">predictor</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="coefficient">coefficient</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr class="gt_group_heading_row">
<th colspan="2" class="gt_empty_group_heading" scope="colgroup" id=""></th>
    </tr>
<tr class="gt_row_group_first">
<td headers="model  predictor" class="gt_row gt_left">Geographic</td>
<td headers="model  coefficient" class="gt_row gt_right" style="background-color: #7FBF7B; color: #000000;">1.35</td>
</tr>
<tr>
<td headers="model  predictor" class="gt_row gt_left">CA_rPCA1</td>
<td headers="model  coefficient" class="gt_row gt_right" style="background-color: #F7F7F7; color: #000000;">0.00</td>
</tr>
<tr>
<td headers="model  predictor" class="gt_row gt_left">CA_rPCA2</td>
<td headers="model  coefficient" class="gt_row gt_right" style="background-color: #F4F6F4; color: #000000;">0.06</td>
</tr>
<tr>
<td headers="model  predictor" class="gt_row gt_left">CA_rPCA3</td>
<td headers="model  coefficient" class="gt_row gt_right" style="background-color: #F6F7F6; color: #000000;">0.02</td>
</tr>
<tr class="gt_group_heading_row">
<th colspan="2" class="gt_empty_group_heading" scope="colgroup" id=""></th>
    </tr>
<tr class="gt_row_group_first">
<td headers="NA  predictor" class="gt_row gt_left" style="background-color: rgba(255,255,255,0.8); color: #000000; font-weight: bold;">% Explained:</td>
<td headers="NA  coefficient" class="gt_row gt_right" style="background-color: rgba(255,255,255,0.8); color: #000000; font-weight: bold;">36.89<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>1</sup></span>
</td>
</tr>
</tbody>
<tfoot class="gt_footnotes"><tr>
<td class="gt_footnote" colspan="2">
<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;"><sup>1</sup></span> The percentage of null deviance explained by the fitted GDM model.</td>
    </tr></tfoot>
</table>
</div>
</div>
</div>
<div class="section level3">
<h3 id="visualizing-gdm-results">Visualizing GDM results<a class="anchor" aria-label="anchor" href="#visualizing-gdm-results"></a>
</h3>
<hr>
<p>There are two ways we can visualize the results from GDM within
algatr, both using the <code><a href="../reference/gdm_map.html">gdm_map()</a></code> function. First, we can
transform the original environmental layers based on biological
importance (i.e., based on the I-splines), run a raster PCA on these
environmental layers, and visualize the first three PC axes by assigning
each axis to a red, green, or blue color scales to create an RGB plot.
In this way, more similar colors on the map correspond to more similar
genetic composition of our samples. Secondly, we can visualize how
genetic composition is associated with each of these variables by
plotting each as a vector and coloring the individual points in the same
PC RGB space.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gdm_map.html">gdm_map</a></span><span class="op">(</span><span class="va">gdm_full</span><span class="op">$</span><span class="va">model</span>, <span class="va">CA_env</span>, <span class="va">liz_coords</span><span class="op">)</span></span></code></pre></div>
<p><img src="GDM_vignette_files/figure-html/GDM%20plots-1.png" width="700"><img src="GDM_vignette_files/figure-html/GDM%20plots-2.png" width="700"></p>
<div class="section level4">
<h4 id="masking-out-irrelevant-areas">Masking out irrelevant areas<a class="anchor" aria-label="anchor" href="#masking-out-irrelevant-areas"></a>
</h4>
<p>An important consideration for a researcher to be aware of is to not
overinterpret GDM maps. In particular, areas with limited (or no)
samples are still assigned values due to model extrapolation; thus, it
may be more helpful to mask out any areas with no sampling so as to not
overinterpret these results. In looking at the map and PCA plot above,
we can see that there are two regions in California that are lacking
samples: the southeastern and north-central parts of the state. Using
algatr’s <code><a href="../reference/extrap_mask.html">extrap_mask()</a></code> function (see the masking vignette
for more information), we can mask out areas from the GDM map in several
ways. Below, we mask out areas (in white) that fall outside a
user-defined buffer around our sampling coordinates.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the GDM map from the GDM model object</span></span>
<span><span class="va">map</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdm_map.html">gdm_map</a></span><span class="op">(</span><span class="va">gdm_full</span><span class="op">$</span><span class="va">model</span>, <span class="va">CA_env</span>, <span class="va">liz_coords</span>, plot_vars <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="GDM_vignette_files/figure-html/gdm%20masked-1.png" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maprgb</span> <span class="op">&lt;-</span> <span class="va">map</span><span class="op">$</span><span class="va">pcaRastRGB</span></span>
<span></span>
<span><span class="co"># Now, use `extrap_mask()` to do buffer-based masking</span></span>
<span><span class="va">map_mask</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extrap_mask.html">extrap_mask</a></span><span class="op">(</span><span class="va">liz_coords</span>, <span class="va">maprgb</span>, method <span class="op">=</span> <span class="st">"buffer"</span>, buffer_width <span class="op">=</span> <span class="fl">1.25</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the resulting masked map</span></span>
<span><span class="fu"><a href="../reference/plot_extrap_mask.html">plot_extrap_mask</a></span><span class="op">(</span><span class="va">maprgb</span>, <span class="va">map_mask</span>, RGB <span class="op">=</span> <span class="cn">TRUE</span>, mask_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/rgb.html" class="external-link">rgb</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, alpha <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GDM_vignette_files/figure-html/gdm%20masked-2.png" width="480" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="running-gdm-with-gdm_do_everything">Running GDM with <code>gdm_do_everything()</code><a class="anchor" aria-label="anchor" href="#running-gdm-with-gdm_do_everything"></a>
</h3>
<hr>
<p>The algatr package also has an option to run all of the above
functionality in a single function, <code><a href="../reference/gdm_do_everything.html">gdm_do_everything()</a></code>.
This function will output the fitted I-splines, table, map, and PCA. The
resulting object looks identical to the output object from
<code><a href="../reference/gdm_run.html">gdm_run()</a></code>. <strong>Please be aware that the
<code>do_everything()</code> functions are meant to be exploratory. We
do not recommend their use for final analyses unless certain they are
properly parameterized.</strong></p>
<p>Running a GDM with <code><a href="../reference/gdm_do_everything.html">gdm_do_everything()</a></code> requires three
data files for input: a genetic distance matrix (the
<code>gendist</code> argument), coordinates for samples (the
<code>coords</code> argument), and environmental layers (the
<code>envlayers</code> argument).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdm_full_everything</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gdm_do_everything.html">gdm_do_everything</a></span><span class="op">(</span><span class="va">liz_gendist</span>,</span>
<span>  <span class="va">liz_coords</span>,</span>
<span>  envlayers <span class="op">=</span> <span class="va">CA_env</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"full"</span>,</span>
<span>  scale_gendist <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Please be aware: the do_everything functions are meant to be exploratory. We do not recommend their use for final analyses unless certain they are properly parameterized.</span></span>
<span><span class="co">#&gt; Warning in crs_check(coords, envlayers): No CRS found for the provided</span></span>
<span><span class="co">#&gt; coordinates. Make sure the coordinates and the raster have the same projection</span></span>
<span><span class="co">#&gt; (see function details or vignette)</span></span></code></pre></div>
<p><img src="GDM_vignette_files/figure-html/do%20everything-1.png" width="480" style="display: block; margin: auto;"><img src="GDM_vignette_files/figure-html/do%20everything-2.png" width="480" style="display: block; margin: auto;"><img src="GDM_vignette_files/figure-html/do%20everything-3.png" width="480" style="display: block; margin: auto;"><img src="GDM_vignette_files/figure-html/do%20everything-4.png" width="480" style="display: block; margin: auto;"><img src="GDM_vignette_files/figure-html/do%20everything-5.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="additional-documentation-and-citations">Additional documentation and citations<a class="anchor" aria-label="anchor" href="#additional-documentation-and-citations"></a>
</h3>
<hr>
<table class="table">
<colgroup>
<col width="25%">
<col width="40%">
<col width="34%">
</colgroup>
<thead><tr class="header">
<th></th>
<th>Citation/URL</th>
<th>Details</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Associated code</td>
<td><a href="https://cran.r-project.org/web/packages/gdm/gdm.pdf" class="external-link">Fitzpatrick
et al. 2022</a></td>
<td>algatr uses the <code>gdm()</code> package; manual contains
walkthroughs</td>
</tr>
<tr class="even">
<td>Associated literature</td>
<td><a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1472-4642.2007.00341.x" class="external-link">Ferrier
et al. 2007</a></td>
<td>Paper describing basic use of GDM</td>
</tr>
<tr class="odd">
<td>Associated literature</td>
<td><a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1365-294X.2010.04684.x" class="external-link">Freedman
et al. 2010</a></td>
<td>Classic example of using GDM</td>
</tr>
<tr class="even">
<td>Associated literature</td>
<td><a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/ele.12376#:~:text=Community%2Dlevel%20modelling%20of%20genomic,assessments%20of%20climate%20change%20vulnerability." class="external-link">Fitzpatrick
&amp; Keller 2015</a></td>
<td>Perspective on using GDM</td>
</tr>
<tr class="odd">
<td>Associated literature</td>
<td><a href="https://onlinelibrary.wiley.com/doi/full/10.1111/geb.13459" class="external-link">Mokany
et al. 2022</a></td>
<td>Perspective on using GDM</td>
</tr>
</tbody>
</table>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Anusha Bishop, Anne Chambers, Ian Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
