<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Genetic distances • algatr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genetic distances">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">algatr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference Manual</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/data_processing_vignette.html">Genetic data processing</a></li>
    <li><a class="dropdown-item" href="../articles/enviro_data_vignette.html">Environmental data processing</a></li>
    <li><a class="dropdown-item" href="../articles/masking_vignette.html">Masking</a></li>
    <li><a class="dropdown-item" href="../articles/gen_dist_vignette.html">Genetic distances</a></li>
    <li><a class="dropdown-item" href="../articles/TESS_vignette.html">TESS</a></li>
    <li><a class="dropdown-item" href="../articles/MMRR_vignette.html">MMRR</a></li>
    <li><a class="dropdown-item" href="../articles/GDM_vignette.html">GDM</a></li>
    <li><a class="dropdown-item" href="../articles/RDA_vignette.html">RDA</a></li>
    <li><a class="dropdown-item" href="../articles/LFMM_vignette.html">LFMM</a></li>
    <li><a class="dropdown-item" href="../articles/wingen_vignette.html">Wingen</a></li>
    <li><a class="dropdown-item" href="../articles/alazygatr_vignette.html">Alazygatr</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Genetic distances</h1>
            
      

      <div class="d-none name"><code>gen_dist_vignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="calculating-genetic-distances">Calculating genetic distances<a class="anchor" aria-label="anchor" href="#calculating-genetic-distances"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thewanglab.github.io/algatr/">algatr</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install required packages</span></span>
<span><span class="fu"><a href="../reference/gen_dist_packages.html">gen_dist_packages</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span></code></pre></div>
<p>For several landscape genomics analyses, a pairwise genetic distance
matrix is required as input. There are a variety of ways to calculate
genetic distance, and the main function in this package,
<code><a href="../reference/gen_dist.html">gen_dist()</a></code>, allows a user to choose from five different
genetic distances and saves square distance matrices as csv files. The
five genetic distance metrics are specified using the
<code>dist_type</code> argument and are as follows:</p>
<ol style="list-style-type: decimal">
<li><p>Euclidean distance (the <code>"euclidean"</code> argument), which
uses the <code><a href="https://rspatial.github.io/terra/reference/distance.html" class="external-link">distance()</a></code> function within the <a href="https://cran.r-project.org/web/packages/ecodist/index.html" class="external-link">ecodist</a>
package</p></li>
<li><p>Bray-Curtis distance (the <code>"bray_curtis"</code> argument),
which uses the <code><a href="https://rspatial.github.io/terra/reference/distance.html" class="external-link">distance()</a></code> function within the <a href="https://cran.r-project.org/web/packages/ecodist/index.html" class="external-link">ecodist</a>
package</p></li>
<li><p>Proportion of shared alleles (the <code>"dps"</code> argument),
which uses the <code>propShared()</code> function within the <a href="https://cran.r-project.org/web/packages/adegenet/index.html" class="external-link">adegenet</a>
package</p></li>
<li><p>PC-based distance (the <code>"pc"</code> argument), which uses
the <code><a href="https://rspatial.github.io/terra/reference/prcomp.html" class="external-link">prcomp()</a></code> function within the <a href="https://stat.ethz.ch/R-manual/R-devel/library/stats/html/00Index.html" class="external-link">stats</a>
package and selects the number of PCs using a Tracy-Widom test (if
<code>npc_selection = "auto"</code>; also requires specifying
<code>criticalpoint</code> [see below for details]) or user-inputted PCs
based on examining a screeplot and manually entering a value into the
console (if <code>npc_selection = "manual"</code>)</p></li>
<li><p>Processing distances generated using Plink (the
<code>"plink"</code> argument). This type of distance requires providing
paths to two plink output files (the distance file,
<code>plink_file</code>, and the ID file, <code>plink_id_file</code>);
the distance file <strong>must</strong> be a square (i.e., symmetric)
distance matrix</p></li>
</ol>
<p>A good place to start to understand these metrics is <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12684" class="external-link">Shirk
et al., 2017</a>, who tested a number of different genetic distance
metrics (including most of the above) and compared results for use in
landscape genetics analyses.</p>
<p>This package also allows a user to look at the relationship between
distance metrics using the <code><a href="../reference/gen_dist_corr.html">gen_dist_corr()</a></code> function, and to
visualize genetic distance matrices with a heatmap using the
<code><a href="../reference/gen_dist_hm.html">gen_dist_hm()</a></code> function.</p>
<div class="section level4">
<h4 id="load-example-dataset">Load example dataset<a class="anchor" aria-label="anchor" href="#load-example-dataset"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/load_algatr_example.html">load_algatr_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; ---------------- example dataset ----------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span> </span></span>
<span><span class="co"><span style="color: #0000BB;">#&gt; Objects loaded:</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_vcf*</span> <span style="color: #BBBB00;">vcfR object (1000 loci x 53 samples)</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_gendist*</span> <span style="color: #BBBB00;">genetic distance matrix (Plink Distance)</span> </span></span>
<span><span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; *liz_coords*</span> <span style="color: #00BB00;">dataframe with x and y coordinates</span> </span></span>
<span><span class="co"><span style="color: #BB00BB; font-weight: bold;">#&gt; *CA_env*</span> <span style="color: #BB00BB;">RasterStack with example environmental layers</span> </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; -------------------------------------------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co">#&gt; </span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculate-genetic-distances-with-gen_dist">Calculate genetic distances with <code>gen_dist()</code><a class="anchor" aria-label="anchor" href="#calculate-genetic-distances-with-gen_dist"></a>
</h3>
<hr>
<p>Euclidean, Bray-Curtis, and PC-based genetic distance metrics
<strong><em>require no missing data</em></strong>. A simple imputation
(to the median or mean, depending on the distance metric) is built into
the <code><a href="../reference/gen_dist.html">gen_dist()</a></code> function mostly for creating test datasets;
we <strong><em>highly discourage</em></strong> using this form of
imputation in any of your analyses. You can take a look at the
<code><a href="../reference/str_impute.html">str_impute()</a></code> function description in the data processing
vignette for information on an alternate imputation method provided in
algatr.</p>
<p>To calculate Euclidean distances between our samples:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate Euclidean distance matrix</span></span>
<span><span class="va">euc_dists</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_dist.html">gen_dist</a></span><span class="op">(</span><span class="va">liz_vcf</span>, dist_type <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required namespace: vcfR</span></span>
<span><span class="co">#&gt; Loading required namespace: adegenet</span></span>
<span><span class="co">#&gt; Warning in gen_dist(liz_vcf, dist_type = "euclidean"): NAs found in genetic</span></span>
<span><span class="co">#&gt; data, imputing to the median (NOTE: this simplified imputation approach is</span></span>
<span><span class="co">#&gt; strongly discouraged. Consider using another method of removing missing data)</span></span>
<span></span>
<span><span class="va">euc_dists</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;            ALT3   BAR360     BLL5     BNT5     BOF1</span></span>
<span><span class="co">#&gt; ALT3    0.00000 22.92379 17.26992 21.08910 18.77498</span></span>
<span><span class="co">#&gt; BAR360 22.92379  0.00000 17.85357 19.50000 17.56417</span></span>
<span><span class="co">#&gt; BLL5   17.26992 17.85357  0.00000 15.76388 12.63922</span></span>
<span><span class="co">#&gt; BNT5   21.08910 19.50000 15.76388  0.00000 15.99219</span></span>
<span><span class="co">#&gt; BOF1   18.77498 17.56417 12.63922 15.99219  0.00000</span></span></code></pre></div>
<p>Now, let’s process Plink distances, keeping in mind that the provided
Plink distance matrix must be square (i.e., symmetric). Refer the Plink
documentation for how to generate a square distance matrix.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plink_dists</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_dist.html">gen_dist</a></span><span class="op">(</span>plink_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"liz_test.dist"</span>, package <span class="op">=</span> <span class="st">"algatr"</span><span class="op">)</span>, plink_id_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"liz_test.dist.id"</span>, package <span class="op">=</span> <span class="st">"algatr"</span><span class="op">)</span>, dist_type <span class="op">=</span> <span class="st">"plink"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">53</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">53</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (53): X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">53</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">2</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (2): X1, X2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre></div>
<p>Finally, let’s calculate PC-based distances. There are two options
for selecting the “best” number of PCs, specified using the
<code>npc_selection</code> argument: <code>"auto"</code>, which runs a
Tracy-Widom test and selects the number of significant eigenvalues of
the PC, or <code>"manual"</code>, in which a screeplot is printed to the
screen and a user can manually enter the number of PCs to choose into
the console. For the automatic option, a significance threshold must be
specified using the <code>criticalpoint</code> argument; significance
levels of 0.05, 0.01, 0.005, or 0.001 correspond to the
<code>criticalpoint</code> argument being 0.9793, 2.0234, 2.4224, or
3.2724, respectively).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pc_dists</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_dist.html">gen_dist</a></span><span class="op">(</span><span class="va">liz_vcf</span>, dist_type <span class="op">=</span> <span class="st">"pc"</span>, npc_selection <span class="op">=</span> <span class="st">"auto"</span>, criticalpoint <span class="op">=</span> <span class="fl">2.0234</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in gen_dist(liz_vcf, dist_type = "pc", npc_selection = "auto",</span></span>
<span><span class="co">#&gt; criticalpoint = 2.0234): NAs found in genetic data, imputing to the median</span></span>
<span><span class="co">#&gt; (NOTE: this simplified imputation approach is strongly discouraged. Consider</span></span>
<span><span class="co">#&gt; using another method of removing missing data)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="compare-genetic-distance-results">Compare genetic distance results<a class="anchor" aria-label="anchor" href="#compare-genetic-distance-results"></a>
</h3>
<hr>
<div class="section level4">
<h4 id="plot-comparisons-between-two-genetic-distances-with-gen_dist_corr">Plot comparisons between two genetic distances with
<code>gen_dist_corr()</code><a class="anchor" aria-label="anchor" href="#plot-comparisons-between-two-genetic-distances-with-gen_dist_corr"></a>
</h4>
<p>To compare between the genetic distances you just produced, we can
plot them:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot some pairwise comparisons, providing names for the metrics</span></span>
<span><span class="va">p_euc_plink</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_dist_corr.html">gen_dist_corr</a></span><span class="op">(</span><span class="va">euc_dists</span>, <span class="va">plink_dists</span>, <span class="st">"Euclidean"</span>, <span class="st">"Plink"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(comparison, name)`</span></span>
<span><span class="co">#&gt; Warning: `aes_string()` was deprecated in ggplot2 3.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use tidy evaluation idioms with `aes()`.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> See also `vignette("ggplot2-in-packages")` for more information.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">algatr</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue to the authors.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="va">p_pc_plink</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_dist_corr.html">gen_dist_corr</a></span><span class="op">(</span><span class="va">pc_dists</span>, <span class="va">plink_dists</span>, <span class="st">"PC_based"</span>, <span class="st">"Plink"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(comparison, name)`</span></span>
<span></span>
<span><span class="co"># Show all plots as panels on a single plot</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p_euc_plink</span>, <span class="va">p_pc_plink</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 1378 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_point()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 1378 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_point()`).</span></span></code></pre></div>
<p><img src="gen_dist_vignette_files/figure-html/gendist%20corr-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="plot-heatmap-of-one-of-the-genetic-distances-with-gen_dist_hm">Plot heatmap of one of the genetic distances with
<code>gen_dist_hm()</code><a class="anchor" aria-label="anchor" href="#plot-heatmap-of-one-of-the-genetic-distances-with-gen_dist_hm"></a>
</h4>
<p>Let’s look at a heatmap of Euclidean distance for our example
dataset.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gen_dist_hm.html">gen_dist_hm</a></span><span class="op">(</span><span class="va">euc_dists</span><span class="op">)</span></span></code></pre></div>
<p><img src="gen_dist_vignette_files/figure-html/euc%20hm-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<p>Let’s look at a heatmap of the DPS measure of genetic distance which
is calculated as <code>1 - proportion of shared alleles</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dps_dists</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_dist.html">gen_dist</a></span><span class="op">(</span><span class="va">liz_vcf</span>, dist_type <span class="op">=</span> <span class="st">"dps"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Prior to May 2025, this function incorrectly returned the proportion</span></span>
<span><span class="co">#&gt; of shared alleles (PS) instead of the genetic distance measure: DPS = 1 - PS.</span></span>
<span><span class="co">#&gt; Please review results from prior versions accordingly. This warning will appear</span></span>
<span><span class="co">#&gt; once per session. To suppress, set options(wingen.quiet_dps_warning = TRUE).</span></span>
<span><span class="fu"><a href="../reference/gen_dist_hm.html">gen_dist_hm</a></span><span class="op">(</span><span class="va">dps_dists</span><span class="op">)</span></span></code></pre></div>
<p><img src="gen_dist_vignette_files/figure-html/dps%20hm-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="additional-documentation-and-citations">Additional documentation and citations<a class="anchor" aria-label="anchor" href="#additional-documentation-and-citations"></a>
</h3>
<hr>
<table class="table">
<colgroup>
<col width="26%">
<col width="38%">
<col width="34%">
</colgroup>
<thead><tr class="header">
<th></th>
<th>Citation/URL</th>
<th>Details</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Associated code</td>
<td><a href="https://cran.r-project.org/web/packages/ecodist/index.html" class="external-link">Goslee
&amp; Urban 2007</a></td>
<td>algatr’s calculation of Euclidean and Bray-Curtis distances using
the <code><a href="https://rspatial.github.io/terra/reference/distance.html" class="external-link">distance()</a></code> function within the ecodist package</td>
</tr>
<tr class="even">
<td>Associated code</td>
<td>
<a href="https://academic.oup.com/bioinformatics/article/24/11/1403/191127" class="external-link">Jombart
2008</a>; <a href="https://academic.oup.com/bioinformatics/article/27/21/3070/218892" class="external-link">Jombart
&amp; Ahmed 2011</a>
</td>
<td>algatr’s calculation of proportion of shared alleles using the
<code>propShared()</code> function within the adegenet package</td>
</tr>
<tr class="odd">
<td>Associated literature</td>
<td><a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12684" class="external-link">Shirk
et al., 2017</a></td>
<td>Paper testing a number of different genetic distance metrics
(including those used by algatr) for landscape genomics</td>
</tr>
</tbody>
</table>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anusha Bishop, Anne Chambers, Ian Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
