<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>LFMM • algatr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LFMM">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">algatr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference Manual</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/data_processing_vignette.html">Genetic data processing</a></li>
    <li><a class="dropdown-item" href="../articles/enviro_data_vignette.html">Environmental data processing</a></li>
    <li><a class="dropdown-item" href="../articles/masking_vignette.html">Masking</a></li>
    <li><a class="dropdown-item" href="../articles/gen_dist_vignette.html">Genetic distances</a></li>
    <li><a class="dropdown-item" href="../articles/TESS_vignette.html">TESS</a></li>
    <li><a class="dropdown-item" href="../articles/MMRR_vignette.html">MMRR</a></li>
    <li><a class="dropdown-item" href="../articles/GDM_vignette.html">GDM</a></li>
    <li><a class="dropdown-item" href="../articles/RDA_vignette.html">RDA</a></li>
    <li><a class="dropdown-item" href="../articles/LFMM_vignette.html">LFMM</a></li>
    <li><a class="dropdown-item" href="../articles/wingen_vignette.html">Wingen</a></li>
    <li><a class="dropdown-item" href="../articles/alazygatr_vignette.html">Alazygatr</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>LFMM</h1>
            
      

      <div class="d-none name"><code>LFMM_vignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="latent-factor-mixed-models-lfmm">Latent factor mixed models (LFMM)<a class="anchor" aria-label="anchor" href="#latent-factor-mixed-models-lfmm"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thewanglab.github.io/algatr/">algatr</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install required packages</span></span>
<span><span class="fu"><a href="../reference/lfmm_packages.html">lfmm_packages</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/raster" class="external-link">raster</a></span><span class="op">)</span></span></code></pre></div>
<p><strong>If using LFMM, please cite the following: for the LFMM
method, Caye et al. 2019. For the <code>lfmm</code> package, please cite
Jumentier B. (2021). lfmm: Latent Factor Mixed Models. R package version
1.1.</strong></p>
<p>LFMM (<a href="https://academic.oup.com/mbe/article/30/7/1687/972098" class="external-link">Frichot et
al. 2013</a>) is a genotype-environment association (GEA) method in
which mixed models are used to determine loci that are significantly
associated with environmental variables. This method is similar to
performing RDA, except that LFMM takes into account – and corrects for –
unobserved variables that may confound results (known as latent
factors). A good example of a latent factor that we may want to correct
for prior to examining environment-associated loci is population
structure.</p>
<p>There are a variety of ways to determine the best number of latent
factors in LFMM (which are called K values, although these are
<em>not</em> the same as TESS K values). algatr provides the option to
test using four such methods: (1) Tracy-Widom test; (2) quick elbow test
(i.e., scree test); (3) using the clustering algorithm from TESS (i.e.,
determining latent factors corresponding to some measure of population
structure); and (4) K-means clustering. We will explore each of these,
and compare results, in this vignette.</p>
<p>Importantly, as with RDA, LFMM cannot take in missing values.
Imputation based on the per-site median is commonly performed, but there
are several other ways researchers can deal with missing values. For
example, algatr contains the <code><a href="../reference/str_impute.html">str_impute()</a></code> function to
impute missing values based on population structure using the
<code><a href="https://rdrr.io/pkg/LEA/man/impute.html" class="external-link">LEA::impute()</a></code> function. However, here, we’ll impute on the
median, but strongly urge researchers to use extreme caution when using
this form of simplistic imputation. We mainly provide code to impute on
the median for testing datasets and highly discourage its use in further
analyses (please use <code><a href="../reference/str_impute.html">str_impute()</a></code> instead!).</p>
<p>To perform an LFMM analysis (<a href="https://academic.oup.com/mbe/article/36/4/852/5290100" class="external-link">Caye et
al. 2019</a>), algatr uses the lfmm package (see <a href="https://bcm-uga.github.io/lfmm/index.html" class="external-link">here</a> for details).
Despite its name, this package implements the LFMM2 method, which is
similar to the original LFMM algorithm, but is computationally faster.
The LEA package (<a href="https://rdrr.io/bioc/LEA/" class="external-link">Frichot &amp;
Francois 2015</a>) also provides a wrapper for lfmm2 using the
<code>lfmm2()</code> function.</p>
<div class="section level4">
<h4 id="read-in-and-process-data-files">Read in and process data files<a class="anchor" aria-label="anchor" href="#read-in-and-process-data-files"></a>
</h4>
<p>To run LFMM, we will need a genotype dosage matrix and environmental
values extracted at sampling coordinates. Let’s load the example
dataset, convert the vcf to a dosage matrix using the
<code><a href="../reference/vcf_to_dosage.html">vcf_to_dosage()</a></code> function, and extract environmental data
for our sampling coordinates from our environmental raster stack
(CA_env).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/load_algatr_example.html">load_algatr_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; ---------------- example dataset ----------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span> </span></span>
<span><span class="co"><span style="color: #0000BB;">#&gt; Objects loaded:</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_vcf*</span> <span style="color: #BBBB00;">vcfR object (1000 loci x 53 samples)</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_gendist*</span> <span style="color: #BBBB00;">genetic distance matrix (Plink Distance)</span> </span></span>
<span><span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; *liz_coords*</span> <span style="color: #00BB00;">dataframe with x and y coordinates</span> </span></span>
<span><span class="co"><span style="color: #BB00BB; font-weight: bold;">#&gt; *CA_env*</span> <span style="color: #BB00BB;">RasterStack with example environmental layers</span> </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; -------------------------------------------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co"># Convert vcf to dosage matrix</span></span>
<span><span class="va">gen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_to_dosage.html">vcf_to_dosage</a></span><span class="op">(</span><span class="va">liz_vcf</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required namespace: vcfR</span></span>
<span><span class="co">#&gt; Loading required namespace: adegenet</span></span>
<span><span class="co"># Also, our code assumes that sample IDs from gendist and coords are the same order; be sure to check this before moving forward!</span></span>
<span><span class="va">env</span> <span class="op">&lt;-</span> <span class="fu">raster</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/extract.html" class="external-link">extract</a></span><span class="op">(</span><span class="va">CA_env</span>, <span class="va">liz_coords</span><span class="op">)</span></span></code></pre></div>
<p>As mentioned above, LFMM requires that your genotype matrix contains
no missing values. Let’s impute missing values based on the per-site
median. <em>N.B.: this type of simplistic imputation is strongly not
recommended for downstream analyses and is used here for example’s
sake!</em></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Are there NAs in the data?</span></span>
<span><span class="va">gen</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;        Locus_10 Locus_15 Locus_22 Locus_28 Locus_32</span></span>
<span><span class="co">#&gt; ALT3          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BAR360        0        0        0       NA        0</span></span>
<span><span class="co">#&gt; BLL5          0       NA        0        0       NA</span></span>
<span><span class="co">#&gt; BNT5          0        0       NA       NA        0</span></span>
<span><span class="co">#&gt; BOF1          0       NA       NA       NA       NA</span></span>
<span><span class="va">gen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simple_impute.html">simple_impute</a></span><span class="op">(</span><span class="va">gen</span><span class="op">)</span></span>
<span><span class="co"># Check that NAs are gone</span></span>
<span><span class="va">gen</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;        Locus_10 Locus_15 Locus_22 Locus_28 Locus_32</span></span>
<span><span class="co">#&gt; ALT3          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BAR360        0        0        0        0        0</span></span>
<span><span class="co">#&gt; BLL5          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BNT5          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BOF1          0        0        0        0        0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="determining-the-number-of-latent-factors-using-select_k">Determining the number of latent factors using
<code>select_K()</code><a class="anchor" aria-label="anchor" href="#determining-the-number-of-latent-factors-using-select_k"></a>
</h3>
<hr>
<p>LFMM requires a user to define the number of latent factors (K
values). algatr provides options to perform four types of K selection
using the <code>K_selection</code> parameter within the
<code><a href="../reference/select_K.html">select_K()</a></code> function:</p>
<ul>
<li><p><code>K_selection = "tracy_widom"</code>: A Tracy-Widom test is
performed. Briefly, this method finds significant eigenvalues of a
matrix after performing a PCA. This argument uses the <code>tw()</code>
function within the <a href="https://cran.r-project.org/web/packages/AssocTests/index.html" class="external-link">AssocTests</a>
package, and also requires a user to provide the
<code>criticalpoint</code>, which corresponds to the significance level.
Significance levels of alpha=0.05, 0.01, 0.005, or 0.001 correspond to
critical point values of 0.9793, 2.0234, 2.4224, or 3.2724,
respectively. The default significance level is 0.01, so the default for
the <code>criticalpoint</code> parameter is 2.0234. The Tracy-Widom test
is also the default method for the <code><a href="../reference/select_K.html">select_K()</a></code>
function.</p></li>
<li><p><code>K_selection = "quick_elbow"</code>: A “quick elbow” is
found automatically, which refers to a large percent drop in eigenvalues
given a variance below 5% for each principal component (essentially the
same as a scree test). The code to determine a quick elbow is adapted
from that developed by Nicholas Cooper <a href="https://rdrr.io/cran/bigpca/man/quick.elbow.html" class="external-link">here</a>. This
argument requires users to set the threshold that defines whether a
principal component explains ‘much’ of the variance using the
<code>low</code> argument, and also the maximum percentage of the
variance to capture before the elbow (i.e., the cumulative sum to PC
‘n’) using the <code>max.pc</code> argument. The default for
<code>low</code> is 0.08, and the default for <code>max.pc</code> is
0.90.</p></li>
<li><p><code>K_selection = "tess"</code>: The automatic K selection is
performed after running TESS. This method minimizes the slope of the
line that connects cross-entropy scores between K values. This approach
is described <a href="https://chazhyseni.github.io/NALgen/post/determining_bestk/" class="external-link">here</a>.
This argument requires users to provide coordinates
(<code>coords</code>) and set the range of K values to test
(<code>Kvals</code>; defaults to 1:10).</p></li>
<li>
<p><code>K_selection = "find_clusters"</code>: Uses K-means
clustering to detect the number of clusters that best describe the data;
this uses the <code>find.clusters()</code> function in the <a href="https://cran.r-project.org/web/packages/adegenet/index.html" class="external-link">adegenet</a>
package. For this method, the data are first transformed using a PCA,
after which successive K-means clustering is run with an increasing
number of K values. The “best” K is selected using the
<code>"diffNgroup</code>” criterion in adegenet’s
<code>find.clusters</code> function. algatr’s
<code>"find_clusters"</code> argument requires users to set several
parameters for both the PCA and the K-means clustering steps:</p>
<ul>
<li><p>The user must specify the minimum amount of the total variance to
be preserved by the retained axes using the <code>perc.pca</code>
argument. The default for this argument is 90%.</p></li>
<li><p>The maximum number of clusters to be tested using the
<code>max.n.clust</code> parameter; the default for this is 10.</p></li>
</ul>
</li>
</ul>
<p>There are more options within the <code>find.clusters</code> function
in adegenet, but we’ve chosen to include only some of them in algatr
(e.g., the <code>pca.select</code> argument is set to
<code>"percVar"</code>). Please refer to the adegenet documentation for
further information.</p>
<p>Let’s see how the above methods compare to one another in terms of
how K values compare. Regardless of the K selection that’s done, a user
must provide a dosage matrix (<code>gen</code>) and the K selection
procedure they would like to perform (<code>K_selection</code>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Keep relevant params but retain default values for them</span></span>
<span><span class="fu"><a href="../reference/select_K.html">select_K</a></span><span class="op">(</span><span class="va">gen</span>, K_selection <span class="op">=</span> <span class="st">"tracy_widom"</span>, criticalpoint <span class="op">=</span> <span class="fl">2.0234</span><span class="op">)</span> <span class="co"># 6</span></span>
<span><span class="co">#&gt; [1] 6</span></span>
<span></span>
<span><span class="fu"><a href="../reference/select_K.html">select_K</a></span><span class="op">(</span><span class="va">gen</span>, K_selection <span class="op">=</span> <span class="st">"quick_elbow"</span>, low <span class="op">=</span> <span class="fl">0.08</span>, max.pc <span class="op">=</span> <span class="fl">0.90</span><span class="op">)</span> <span class="co"># 3</span></span></code></pre></div>
<p><img src="LFMM_vignette_files/figure-html/K%20selection-1.png" width="288" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; [1] 3</span></span>
<span></span>
<span><span class="fu"><a href="../reference/select_K.html">select_K</a></span><span class="op">(</span><span class="va">gen</span>, K_selection <span class="op">=</span> <span class="st">"tess"</span>, coords <span class="op">=</span> <span class="va">liz_coords</span>, Kvals <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="co"># 3</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span></code></pre>
<p><img src="LFMM_vignette_files/figure-html/K%20selection-2.png" width="288" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; [1] 3</span></span>
<span></span>
<span><span class="fu"><a href="../reference/select_K.html">select_K</a></span><span class="op">(</span><span class="va">gen</span>, K_selection <span class="op">=</span> <span class="st">"find_clusters"</span>, perc.pca <span class="op">=</span> <span class="fl">90</span>, max.n.clust <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="co"># 4</span></span>
<span><span class="co">#&gt; [1] 3</span></span></code></pre>
<p>As you can see, the K selection method has quite an effect on the
best number of latent factors discovered in the data. For this vignette,
let’s move forward with the best K from the default method (Tracy-Widom
test), which means <strong>K = 6</strong>.</p>
</div>
<div class="section level3">
<h3 id="comparing-different-lfmm-methods-with-lfmm_run">Comparing different LFMM methods with <code>lfmm_run()</code><a class="anchor" aria-label="anchor" href="#comparing-different-lfmm-methods-with-lfmm_run"></a>
</h3>
<hr>
<p>The main function in algatr that runs LFMM is
<code><a href="../reference/lfmm_run.html">lfmm_run()</a></code>. This function requires that users provide the
dosage matrix (<code>gen</code>), the extracted environmental variables
(<code>env</code>), the number of latent factors (<code>K</code>), and
the LFMM method (<code>lfmm_method</code>).</p>
<p>Importantly, LFMM can be performed using two main methods
(<code>"lasso"</code> or <code>"ridge"</code>), which primarily differ
in the penalized loss functions that are incorporated into the
least-squares minimization and are specified using the
<code>lfmm_method</code> argument. The least-squares minimization is how
statistical significance is determined for SNPs being associated with
environmental variables. The primary difference between the two methods
is that the ridge method minimizes the problem with an L^2 penalty,
whereas the lasso method minimizes with an L^1 penalty. According to <a href="https://bcm-uga.github.io/lfmm/articles/lfmm#starting-with-lfmm" class="external-link">Caye
&amp; François 2017</a>, a user would want to select the lasso method
(i.e., L^1-norm) to induce “sparsity on the fixed effects, and
corresponds to the prior information that not all response variables may
be associated with the variables of interest. More specifically, the
prior implies that a limited number of rows of the effect size matrix B
are effectively non-zero.” The ridge penalty is further described in <a href="https://academic.oup.com/mbe/article/36/4/852/5290100" class="external-link">Caye et
al. 2019</a>.</p>
<p>One other argument to be aware of within <code><a href="../reference/lfmm_run.html">lfmm_run()</a></code> is
the adjustment of p-values, specified using <code>p_adj</code>. There
are two options for <code>p_adj</code>: <code>"fdr"</code>, which
adjusts calibrated p-values based on the false discovery rate (the
default). Other options for p-value adjustment can be found within the
<code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust()</a></code> function in the stats package.</p>
<p>Finally, the last argument to be aware of within
<code><a href="../reference/lfmm_run.html">lfmm_run()</a></code> is <code>calibrate</code>, which calibrates
p-values based on resulting z-scores. There are two options for this
argument: <code>"gif"</code> (the default), and
<code>"median+MAD"</code>. From the LFMM documentation: “If the”gif”
option is set (default), significance values are calibrated by using the
genomic control method. Genomic control uses a robust estimate of the
variance of z-scores called “genomic inflation factor”. If the
“median+MAD” option is set, the pvalues are calibrated by computing the
median and MAD of the zscores. If NULL, the pvalues are not
calibrated.”</p>
<p>Now that we’ve identified the number of latent factors to provide in
our model, let’s run LFMM using both ridge and lasso penalties with
default settings (i.e., FDR p-value adjustment and the “gif”
calibration).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ridge_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lfmm_run.html">lfmm_run</a></span><span class="op">(</span><span class="va">gen</span>, <span class="va">env</span>, K <span class="op">=</span> <span class="fl">6</span>, lfmm_method <span class="op">=</span> <span class="st">"ridge"</span><span class="op">)</span></span>
<span><span class="va">lasso_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lfmm_run.html">lfmm_run</a></span><span class="op">(</span><span class="va">gen</span>, <span class="va">env</span>, K <span class="op">=</span> <span class="fl">6</span>, lfmm_method <span class="op">=</span> <span class="st">"lasso"</span><span class="op">)</span></span></code></pre></div>
<p>The results object from <code><a href="../reference/lfmm_run.html">lfmm_run()</a></code> contains 5
elements:</p>
<ul>
<li><p><code>lfmm_snps</code>: a list of candidate SNPs that are
statistically significantly associated with environmental variables,
including all relevant information pertaining to the LFMM analysis
(e.g., p-values, fixed effects, etc.)</p></li>
<li><p><code>df</code>: all of the SNPs and their associations,
including all relevant information pertaining to the LFMM
analysis</p></li>
<li><p><code>model</code>: parameters of the LFMM model, including
matrices <strong><em>B</em></strong>, <strong><em>U</em></strong>, and
<strong><em>V</em></strong></p></li>
<li><p><code>lfmm_test_result</code>: results from association
testing</p></li>
<li><p><code>K</code>: the number of latent factors that LFMM was run
using (in our case, 6)</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="visualizing-lfmm-results">Visualizing LFMM results<a class="anchor" aria-label="anchor" href="#visualizing-lfmm-results"></a>
</h3>
<hr>
<p>There are three main ways we can visualize the results from our LFMM
analysis. Each of these is automatically produced by
<code><a href="../reference/lfmm_run.html">lfmm_run()</a></code>, but you can produce them individually as well
with the functions below.</p>
<ul>
<li>
<p><code><a href="../reference/lfmm_table.html">lfmm_table()</a></code>: Build a table with each SNP and its
association (and p-value) with each environmental variable. This
function allows for quite a bit of customization in terms of what is
displayed:</p>
<ul>
<li><p><code>sig</code>: the significance threshold (alpha
value)</p></li>
<li><p><code>sig_only</code>: whether to only display SNPs above the
significance threshold</p></li>
<li><p><code>top</code>: if there are SNPs that are significantly
associated with multiple environmental variables, only display the top
association (i.e., variable with the maximum B value)</p></li>
<li><p><code>order</code>: whether to order the SNPs in descending order
based on their B value, otherwise they will be ordered based on
environmental variable</p></li>
<li><p><code>var</code>: only display SNPs associated with one
environmental variable</p></li>
<li><p><code>nrow</code>: the number of rows to display</p></li>
<li><p><code>digits</code>: the number of decimal places to
include</p></li>
<li><p><code>footnotes</code>: whether to include footnotes describing
the variables displayed</p></li>
<li><p><code>p_adj</code>: whether p-values were adjusted in
<code><a href="../reference/lfmm_run.html">lfmm_run()</a></code></p></li>
</ul>
</li>
<li><p><code><a href="../reference/lfmm_qqplot.html">lfmm_qqplot()</a></code>: Build a QQ-plot (quantile-quantile
plot) which plots expected quantiles (percentiles) against
p-values.</p></li>
<li><p><code><a href="../reference/lfmm_manhattanplot.html">lfmm_manhattanplot()</a></code>: Build a Manhattan plot with
defined significance threshold.</p></li>
</ul>
<div class="section level4">
<h4 id="building-a-table-of-snp-associations-with-lfmm_table">Building a table of SNP associations with
<code>lfmm_table()</code><a class="anchor" aria-label="anchor" href="#building-a-table-of-snp-associations-with-lfmm_table"></a>
</h4>
<p>Let’s see what the top candidate SNPs were for each of our LFMM
runs:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Build tables for each of our LFMM runs, displaying only significant SNPs and ordering according to effect size (B)</span></span>
<span><span class="fu"><a href="../reference/lfmm_table.html">lfmm_table</a></span><span class="op">(</span><span class="va">lasso_results</span><span class="op">$</span><span class="va">df</span>, order <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div id="kcoblywemr" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#kcoblywemr table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#kcoblywemr thead, #kcoblywemr tbody, #kcoblywemr tfoot, #kcoblywemr tr, #kcoblywemr td, #kcoblywemr th {
  border-style: none;
}

#kcoblywemr p {
  margin: 0;
  padding: 0;
}

#kcoblywemr .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#kcoblywemr .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#kcoblywemr .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#kcoblywemr .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#kcoblywemr .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#kcoblywemr .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#kcoblywemr .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#kcoblywemr .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#kcoblywemr .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#kcoblywemr .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#kcoblywemr .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#kcoblywemr .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#kcoblywemr .gt_spanner_row {
  border-bottom-style: hidden;
}

#kcoblywemr .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#kcoblywemr .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#kcoblywemr .gt_from_md > :first-child {
  margin-top: 0;
}

#kcoblywemr .gt_from_md > :last-child {
  margin-bottom: 0;
}

#kcoblywemr .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#kcoblywemr .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#kcoblywemr .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#kcoblywemr .gt_row_group_first td {
  border-top-width: 2px;
}

#kcoblywemr .gt_row_group_first th {
  border-top-width: 2px;
}

#kcoblywemr .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#kcoblywemr .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#kcoblywemr .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#kcoblywemr .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#kcoblywemr .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#kcoblywemr .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#kcoblywemr .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#kcoblywemr .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#kcoblywemr .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#kcoblywemr .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#kcoblywemr .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#kcoblywemr .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#kcoblywemr .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#kcoblywemr .gt_left {
  text-align: left;
}

#kcoblywemr .gt_center {
  text-align: center;
}

#kcoblywemr .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#kcoblywemr .gt_font_normal {
  font-weight: normal;
}

#kcoblywemr .gt_font_bold {
  font-weight: bold;
}

#kcoblywemr .gt_font_italic {
  font-style: italic;
}

#kcoblywemr .gt_super {
  font-size: 65%;
}

#kcoblywemr .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#kcoblywemr .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#kcoblywemr .gt_indent_1 {
  text-indent: 5px;
}

#kcoblywemr .gt_indent_2 {
  text-indent: 10px;
}

#kcoblywemr .gt_indent_3 {
  text-indent: 15px;
}

#kcoblywemr .gt_indent_4 {
  text-indent: 20px;
}

#kcoblywemr .gt_indent_5 {
  text-indent: 25px;
}

#kcoblywemr .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#kcoblywemr div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="snp">snp</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="variable">variable</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="B">B<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="z-score">z-score</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p-value">p-value</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="calibrated-z-score">calibrated z-score</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="calibrated-p-value">calibrated p-value</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="adjusted-p-value">adjusted p-value</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2338</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA3</td>
<td headers="B" class="gt_row gt_right" style="background-color: #AF8DC3; color: #FFFFFF;">-0.26</td>
<td headers="z-score" class="gt_row gt_right">-4.44</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">19.19</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.01</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_125</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA3</td>
<td headers="B" class="gt_row gt_right" style="background-color: #B7DDB1; color: #000000;">0.18</td>
<td headers="z-score" class="gt_row gt_right">3.80</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">14.09</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.04</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_859</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA3</td>
<td headers="B" class="gt_row gt_right" style="background-color: #B7DDB1; color: #000000;">0.18</td>
<td headers="z-score" class="gt_row gt_right">3.80</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">14.09</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.04</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_249</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #D2ECCC; color: #000000;">0.14</td>
<td headers="z-score" class="gt_row gt_right">4.95</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">16.78</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.04</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_1051</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA3</td>
<td headers="B" class="gt_row gt_right" style="background-color: #E2F2DE; color: #000000;">0.09</td>
<td headers="z-score" class="gt_row gt_right">3.80</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">14.09</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.04</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_562</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F1E9F1; color: #000000;">-0.05</td>
<td headers="z-score" class="gt_row gt_right">-4.74</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">15.42</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.04</td>
</tr>
</tbody>
<tfoot class="gt_footnotes"><tr>
<td class="gt_footnote" colspan="8">
<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span> LFMM effect size</td>
    </tr></tfoot>
</table>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/lfmm_table.html">lfmm_table</a></span><span class="op">(</span><span class="va">ridge_results</span><span class="op">$</span><span class="va">df</span>, order <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div id="kcdzdgmoqo" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#kcdzdgmoqo table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#kcdzdgmoqo thead, #kcdzdgmoqo tbody, #kcdzdgmoqo tfoot, #kcdzdgmoqo tr, #kcdzdgmoqo td, #kcdzdgmoqo th {
  border-style: none;
}

#kcdzdgmoqo p {
  margin: 0;
  padding: 0;
}

#kcdzdgmoqo .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#kcdzdgmoqo .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#kcdzdgmoqo .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#kcdzdgmoqo .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#kcdzdgmoqo .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#kcdzdgmoqo .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#kcdzdgmoqo .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#kcdzdgmoqo .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#kcdzdgmoqo .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#kcdzdgmoqo .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#kcdzdgmoqo .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#kcdzdgmoqo .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#kcdzdgmoqo .gt_spanner_row {
  border-bottom-style: hidden;
}

#kcdzdgmoqo .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#kcdzdgmoqo .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#kcdzdgmoqo .gt_from_md > :first-child {
  margin-top: 0;
}

#kcdzdgmoqo .gt_from_md > :last-child {
  margin-bottom: 0;
}

#kcdzdgmoqo .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#kcdzdgmoqo .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#kcdzdgmoqo .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#kcdzdgmoqo .gt_row_group_first td {
  border-top-width: 2px;
}

#kcdzdgmoqo .gt_row_group_first th {
  border-top-width: 2px;
}

#kcdzdgmoqo .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#kcdzdgmoqo .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#kcdzdgmoqo .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#kcdzdgmoqo .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#kcdzdgmoqo .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#kcdzdgmoqo .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#kcdzdgmoqo .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#kcdzdgmoqo .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#kcdzdgmoqo .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#kcdzdgmoqo .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#kcdzdgmoqo .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#kcdzdgmoqo .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#kcdzdgmoqo .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#kcdzdgmoqo .gt_left {
  text-align: left;
}

#kcdzdgmoqo .gt_center {
  text-align: center;
}

#kcdzdgmoqo .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#kcdzdgmoqo .gt_font_normal {
  font-weight: normal;
}

#kcdzdgmoqo .gt_font_bold {
  font-weight: bold;
}

#kcdzdgmoqo .gt_font_italic {
  font-style: italic;
}

#kcdzdgmoqo .gt_super {
  font-size: 65%;
}

#kcdzdgmoqo .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#kcdzdgmoqo .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#kcdzdgmoqo .gt_indent_1 {
  text-indent: 5px;
}

#kcdzdgmoqo .gt_indent_2 {
  text-indent: 10px;
}

#kcdzdgmoqo .gt_indent_3 {
  text-indent: 15px;
}

#kcdzdgmoqo .gt_indent_4 {
  text-indent: 20px;
}

#kcdzdgmoqo .gt_indent_5 {
  text-indent: 25px;
}

#kcdzdgmoqo .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#kcdzdgmoqo div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="snp">snp</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="variable">variable</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="B">B<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="z-score">z-score</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p-value">p-value</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="calibrated-z-score">calibrated z-score</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="calibrated-p-value">calibrated p-value</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="adjusted-p-value">adjusted p-value</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2338</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA3</td>
<td headers="B" class="gt_row gt_right" style="background-color: #AF8DC3; color: #FFFFFF;">-0.29</td>
<td headers="z-score" class="gt_row gt_right">-5.25</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">25.50</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_125</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA3</td>
<td headers="B" class="gt_row gt_right" style="background-color: #CAE8C3; color: #000000;">0.17</td>
<td headers="z-score" class="gt_row gt_right">4.08</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">15.43</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.02</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_859</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA3</td>
<td headers="B" class="gt_row gt_right" style="background-color: #CAE8C3; color: #000000;">0.17</td>
<td headers="z-score" class="gt_row gt_right">4.08</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">15.43</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.02</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2947</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #E9D8EA; color: #000000;">-0.13</td>
<td headers="z-score" class="gt_row gt_right">-7.49</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">38.03</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_512</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA3</td>
<td headers="B" class="gt_row gt_right" style="background-color: #DCF1D7; color: #000000;">0.13</td>
<td headers="z-score" class="gt_row gt_right">3.85</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">13.71</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.03</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2014</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA3</td>
<td headers="B" class="gt_row gt_right" style="background-color: #DCF1D7; color: #000000;">0.13</td>
<td headers="z-score" class="gt_row gt_right">3.85</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">13.71</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.03</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_1524</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EADAEB; color: #000000;">-0.12</td>
<td headers="z-score" class="gt_row gt_right">-4.84</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">15.89</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_249</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #DEF1D9; color: #000000;">0.12</td>
<td headers="z-score" class="gt_row gt_right">5.92</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">23.78</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2338</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EBDCEC; color: #000000;">-0.11</td>
<td headers="z-score" class="gt_row gt_right">-4.00</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">10.84</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.03</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2133</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #ECDFED; color: #000000;">-0.10</td>
<td headers="z-score" class="gt_row gt_right">-6.45</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">28.25</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_831</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #ECDFED; color: #000000;">-0.10</td>
<td headers="z-score" class="gt_row gt_right">-5.07</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">17.46</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2669</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EDE1EE; color: #000000;">-0.09</td>
<td headers="z-score" class="gt_row gt_right">-5.29</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">18.98</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2703</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #E4F3E1; color: #000000;">0.09</td>
<td headers="z-score" class="gt_row gt_right">4.02</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">10.97</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.03</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_1278</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EDE1EE; color: #000000;">-0.09</td>
<td headers="z-score" class="gt_row gt_right">-5.48</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">20.33</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_786</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EDE1EE; color: #000000;">-0.09</td>
<td headers="z-score" class="gt_row gt_right">-4.31</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">12.61</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.01</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_1318</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EDE1EE; color: #000000;">-0.09</td>
<td headers="z-score" class="gt_row gt_right">-6.51</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">28.74</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_1051</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA3</td>
<td headers="B" class="gt_row gt_right" style="background-color: #E4F3E1; color: #000000;">0.09</td>
<td headers="z-score" class="gt_row gt_right">4.08</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">15.43</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.02</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2045</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EEE4EF; color: #000000;">-0.08</td>
<td headers="z-score" class="gt_row gt_right">-5.72</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">22.21</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_1319</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EEE4EF; color: #000000;">-0.08</td>
<td headers="z-score" class="gt_row gt_right">-5.59</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">21.15</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_1745</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EEE4EF; color: #000000;">-0.08</td>
<td headers="z-score" class="gt_row gt_right">-5.59</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">21.15</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2670</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EEE4EF; color: #000000;">-0.08</td>
<td headers="z-score" class="gt_row gt_right">-6.91</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">32.38</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_138</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #E7F3E3; color: #000000;">0.08</td>
<td headers="z-score" class="gt_row gt_right">4.66</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">14.73</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.01</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_903</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EFE6F0; color: #000000;">-0.07</td>
<td headers="z-score" class="gt_row gt_right">-5.06</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">17.39</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_870</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F0E8F1; color: #000000;">-0.06</td>
<td headers="z-score" class="gt_row gt_right">-5.75</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">22.40</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2426</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F0E8F1; color: #000000;">-0.06</td>
<td headers="z-score" class="gt_row gt_right">-5.75</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">22.40</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2699</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F0E8F1; color: #000000;">-0.06</td>
<td headers="z-score" class="gt_row gt_right">-5.75</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">22.40</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2145</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F0E8F1; color: #000000;">-0.06</td>
<td headers="z-score" class="gt_row gt_right">-3.95</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">10.55</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.04</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2770</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F0E8F1; color: #000000;">-0.06</td>
<td headers="z-score" class="gt_row gt_right">-4.11</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">11.47</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.02</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_562</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F0E8F1; color: #000000;">-0.06</td>
<td headers="z-score" class="gt_row gt_right">-7.05</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">33.70</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_1414</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EBF4E8; color: #000000;">0.06</td>
<td headers="z-score" class="gt_row gt_right">3.90</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">10.30</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.04</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_974</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F3EDF3; color: #000000;">-0.04</td>
<td headers="z-score" class="gt_row gt_right">-5.59</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">21.15</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_1254</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F3EDF3; color: #000000;">-0.04</td>
<td headers="z-score" class="gt_row gt_right">-5.59</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">21.15</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_3116</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F3EDF3; color: #000000;">-0.04</td>
<td headers="z-score" class="gt_row gt_right">-5.59</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">21.15</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.00</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_2421</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F3EDF3; color: #000000;">-0.04</td>
<td headers="z-score" class="gt_row gt_right">-4.38</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">13.04</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.01</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_637</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EFF5ED; color: #000000;">0.04</td>
<td headers="z-score" class="gt_row gt_right">4.22</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">12.06</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.02</td>
</tr>
</tbody>
<tfoot class="gt_footnotes"><tr>
<td class="gt_footnote" colspan="8">
<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span> LFMM effect size</td>
    </tr></tfoot>
</table>
</div>
<p>As you can see from the above, the penalty chosen makes quite a bit
of difference in terms of how many significant SNPs are found (only 6
with the lasso method, but 35 with the ridge method). However, the three
SNPs with the largest effect size are the same between the two methods
(loci 2338, 125, and 859), and the effect sizes and p-values are
comparable. The tables also contain columns for calibrated z-score,
calibrated p-values, and adjusted p-values. Recall that adjusted
p-values are calculated based on setting the <code>p_adj</code>
argument; in our case, these are adjusted based on the false discovery
rate, which is why they differ from non-adjusted p-values (i.e., values
in the p-value column).</p>
<p>Let’s only look at SNPs significantly associated with the PCA2
environmental variable from the lasso method:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/lfmm_table.html">lfmm_table</a></span><span class="op">(</span><span class="va">lasso_results</span><span class="op">$</span><span class="va">df</span>, order <span class="op">=</span> <span class="cn">TRUE</span>, var <span class="op">=</span> <span class="st">"CA_rPCA2"</span><span class="op">)</span></span></code></pre></div>
<div id="oxthbtreze" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#oxthbtreze table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#oxthbtreze thead, #oxthbtreze tbody, #oxthbtreze tfoot, #oxthbtreze tr, #oxthbtreze td, #oxthbtreze th {
  border-style: none;
}

#oxthbtreze p {
  margin: 0;
  padding: 0;
}

#oxthbtreze .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#oxthbtreze .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#oxthbtreze .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#oxthbtreze .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#oxthbtreze .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#oxthbtreze .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#oxthbtreze .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#oxthbtreze .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#oxthbtreze .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#oxthbtreze .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#oxthbtreze .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#oxthbtreze .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#oxthbtreze .gt_spanner_row {
  border-bottom-style: hidden;
}

#oxthbtreze .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#oxthbtreze .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#oxthbtreze .gt_from_md > :first-child {
  margin-top: 0;
}

#oxthbtreze .gt_from_md > :last-child {
  margin-bottom: 0;
}

#oxthbtreze .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#oxthbtreze .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#oxthbtreze .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#oxthbtreze .gt_row_group_first td {
  border-top-width: 2px;
}

#oxthbtreze .gt_row_group_first th {
  border-top-width: 2px;
}

#oxthbtreze .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#oxthbtreze .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#oxthbtreze .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#oxthbtreze .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#oxthbtreze .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#oxthbtreze .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#oxthbtreze .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#oxthbtreze .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#oxthbtreze .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#oxthbtreze .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#oxthbtreze .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#oxthbtreze .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#oxthbtreze .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#oxthbtreze .gt_left {
  text-align: left;
}

#oxthbtreze .gt_center {
  text-align: center;
}

#oxthbtreze .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#oxthbtreze .gt_font_normal {
  font-weight: normal;
}

#oxthbtreze .gt_font_bold {
  font-weight: bold;
}

#oxthbtreze .gt_font_italic {
  font-style: italic;
}

#oxthbtreze .gt_super {
  font-size: 65%;
}

#oxthbtreze .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#oxthbtreze .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#oxthbtreze .gt_indent_1 {
  text-indent: 5px;
}

#oxthbtreze .gt_indent_2 {
  text-indent: 10px;
}

#oxthbtreze .gt_indent_3 {
  text-indent: 15px;
}

#oxthbtreze .gt_indent_4 {
  text-indent: 20px;
}

#oxthbtreze .gt_indent_5 {
  text-indent: 25px;
}

#oxthbtreze .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#oxthbtreze div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="snp">snp</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="variable">variable</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="B">B<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="z-score">z-score</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p-value">p-value</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="calibrated-z-score">calibrated z-score</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="calibrated-p-value">calibrated p-value</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="adjusted-p-value">adjusted p-value</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="snp" class="gt_row gt_left">Locus_249</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #7FBF7B; color: #000000;">0.14</td>
<td headers="z-score" class="gt_row gt_right">4.95</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">16.78</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.04</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">Locus_562</td>
<td headers="variable" class="gt_row gt_left">CA_rPCA2</td>
<td headers="B" class="gt_row gt_right" style="background-color: #ECDEEC; color: #000000;">-0.05</td>
<td headers="z-score" class="gt_row gt_right">-4.74</td>
<td headers="p-value" class="gt_row gt_right">0</td>
<td headers="calibrated z-score" class="gt_row gt_right">15.42</td>
<td headers="calibrated p-value" class="gt_row gt_right">0</td>
<td headers="adjusted p-value" class="gt_row gt_right">0.04</td>
</tr>
</tbody>
<tfoot class="gt_footnotes"><tr>
<td class="gt_footnote" colspan="8">
<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span> LFMM effect size</td>
    </tr></tfoot>
</table>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Be aware that if significant SNPs &lt; nrow, function will return NULL object</span></span>
<span><span class="co"># lfmm_table(lasso_results$df, sig_only = FALSE, order = FALSE, nrow=10)</span></span>
<span></span>
<span><span class="co"># Similarly, the same will occur if you try to specify a variable that is not significantly associated with a SNP</span></span>
<span><span class="co"># lfmm_table(lasso_results$df, sig_only = TRUE, var = "CA_rPCA1")</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="building-a-qqplot-with-lfmm_qqplot">Building a QQplot with <code>lfmm_qqplot()</code><a class="anchor" aria-label="anchor" href="#building-a-qqplot-with-lfmm_qqplot"></a>
</h4>
<p>A QQplot plots theoretical quantiles against p-value quantiles,
providing a way to visualize the normality of the test significance
values. The line that is plotted runs through the intercept and has a
slope of 1. The points should ideally fall along this reference
line:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/lfmm_qqplot.html">lfmm_qqplot</a></span><span class="op">(</span><span class="va">lasso_results</span><span class="op">$</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 414 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_qq()`).</span></span></code></pre></div>
<p><img src="LFMM_vignette_files/figure-html/qqplot-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="building-a-manhattan-plot-using-lfmm_manhattanplot">Building a Manhattan plot using
<code>lfmm_manhattanplot()</code><a class="anchor" aria-label="anchor" href="#building-a-manhattan-plot-using-lfmm_manhattanplot"></a>
</h4>
<p>Manhattan plots are helpful ways to visualize the position of the
outlier SNPs across the genome, as well as how much they exceed a
user-specified significance threshold for each environmental
variable.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># As displayed in our table from above, only six SNPs are visible on the lasso method plots as outliers:</span></span>
<span><span class="fu"><a href="../reference/lfmm_manhattanplot.html">lfmm_manhattanplot</a></span><span class="op">(</span><span class="va">lasso_results</span><span class="op">$</span><span class="va">df</span>, sig <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 414 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_point()`).</span></span></code></pre></div>
<p><img src="LFMM_vignette_files/figure-html/manhattan%20lasso-1.png" width="480" style="display: block; margin: auto;"></p>
<p>For the ridge method LFMM run, we can see that there are (a) more
outliers than the lasso method above, and (b) the majority of these
outliers are associated with environmental PC2.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/lfmm_manhattanplot.html">lfmm_manhattanplot</a></span><span class="op">(</span><span class="va">ridge_results</span><span class="op">$</span><span class="va">df</span>, sig <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 414 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_point()`).</span></span></code></pre></div>
<p><img src="LFMM_vignette_files/figure-html/manhattan%20ridge-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="running-lfmm-with-lfmm_do_everything">Running LFMM with <code>lfmm_do_everything()</code><a class="anchor" aria-label="anchor" href="#running-lfmm-with-lfmm_do_everything"></a>
</h3>
<hr>
<p>The algatr package also has an option to run all of the above
functionality in a single function, <code><a href="../reference/lfmm_do_everything.html">lfmm_do_everything()</a></code>.
This function will automatically extract relevant values from our
environmental layers given our coordinate data; it will also
automatically convert a vcf object to a dosage matrix for us and impute
missing values if present. The output looks the same as the output from
<code><a href="../reference/lfmm_run.html">lfmm_run()</a></code> above, and plots will be automatically produced.
<strong>Please be aware that the <code>do_everything()</code> functions
are meant to be exploratory. We do not recommend their use for final
analyses unless certain they are properly parameterized.</strong></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lfmm_full_everything</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lfmm_do_everything.html">lfmm_do_everything</a></span><span class="op">(</span><span class="va">liz_vcf</span>, <span class="va">CA_env</span>, <span class="va">liz_coords</span>, </span>
<span>                              impute <span class="op">=</span> <span class="st">"structure"</span>,</span>
<span>                              K_impute <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                              lfmm_method <span class="op">=</span> <span class="st">"lasso"</span>, </span>
<span>                              K_selection <span class="op">=</span> <span class="st">"tracy_widom"</span><span class="op">)</span></span></code></pre></div>
<img src="LFMM_vignette_files/figure-html/do%20everything-1.png" width="576" style="display: block; margin: auto;"><img src="LFMM_vignette_files/figure-html/do%20everything-2.png" width="576" style="display: block; margin: auto;"><div id="cifpsawimt" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#cifpsawimt table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#cifpsawimt thead, #cifpsawimt tbody, #cifpsawimt tfoot, #cifpsawimt tr, #cifpsawimt td, #cifpsawimt th {
  border-style: none;
}

#cifpsawimt p {
  margin: 0;
  padding: 0;
}

#cifpsawimt .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#cifpsawimt .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#cifpsawimt .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#cifpsawimt .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#cifpsawimt .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#cifpsawimt .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#cifpsawimt .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#cifpsawimt .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#cifpsawimt .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#cifpsawimt .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#cifpsawimt .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#cifpsawimt .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#cifpsawimt .gt_spanner_row {
  border-bottom-style: hidden;
}

#cifpsawimt .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#cifpsawimt .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#cifpsawimt .gt_from_md > :first-child {
  margin-top: 0;
}

#cifpsawimt .gt_from_md > :last-child {
  margin-bottom: 0;
}

#cifpsawimt .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#cifpsawimt .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#cifpsawimt .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#cifpsawimt .gt_row_group_first td {
  border-top-width: 2px;
}

#cifpsawimt .gt_row_group_first th {
  border-top-width: 2px;
}

#cifpsawimt .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#cifpsawimt .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#cifpsawimt .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#cifpsawimt .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#cifpsawimt .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#cifpsawimt .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#cifpsawimt .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#cifpsawimt .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#cifpsawimt .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#cifpsawimt .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#cifpsawimt .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#cifpsawimt .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#cifpsawimt .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#cifpsawimt .gt_left {
  text-align: left;
}

#cifpsawimt .gt_center {
  text-align: center;
}

#cifpsawimt .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#cifpsawimt .gt_font_normal {
  font-weight: normal;
}

#cifpsawimt .gt_font_bold {
  font-weight: bold;
}

#cifpsawimt .gt_font_italic {
  font-style: italic;
}

#cifpsawimt .gt_super {
  font-size: 65%;
}

#cifpsawimt .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#cifpsawimt .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#cifpsawimt .gt_indent_1 {
  text-indent: 5px;
}

#cifpsawimt .gt_indent_2 {
  text-indent: 10px;
}

#cifpsawimt .gt_indent_3 {
  text-indent: 15px;
}

#cifpsawimt .gt_indent_4 {
  text-indent: 20px;
}

#cifpsawimt .gt_indent_5 {
  text-indent: 25px;
}

#cifpsawimt .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#cifpsawimt div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="snp">
snp
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="variable">
variable
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="B">
B<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span>
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="z-score">
z-score
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p-value">
p-value
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="calibrated-z-score">
calibrated z-score
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="calibrated-p-value">
calibrated p-value
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="adjusted-p-value">
adjusted p-value
</th>
</tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="snp" class="gt_row gt_left">
Locus_2338
</td>
<td headers="variable" class="gt_row gt_left">
CA_rPCA3
</td>
<td headers="B" class="gt_row gt_right" style="background-color: #AF8DC3; color: #FFFFFF;">
-0.26
</td>
<td headers="z-score" class="gt_row gt_right">
-4.44
</td>
<td headers="p-value" class="gt_row gt_right">
0
</td>
<td headers="calibrated z-score" class="gt_row gt_right">
17.50
</td>
<td headers="calibrated p-value" class="gt_row gt_right">
0
</td>
<td headers="adjusted p-value" class="gt_row gt_right">
0.02
</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">
Locus_249
</td>
<td headers="variable" class="gt_row gt_left">
CA_rPCA2
</td>
<td headers="B" class="gt_row gt_right" style="background-color: #CBE8C5; color: #000000;">
0.15
</td>
<td headers="z-score" class="gt_row gt_right">
5.12
</td>
<td headers="p-value" class="gt_row gt_right">
0
</td>
<td headers="calibrated z-score" class="gt_row gt_right">
17.18
</td>
<td headers="calibrated p-value" class="gt_row gt_right">
0
</td>
<td headers="adjusted p-value" class="gt_row gt_right">
0.02
</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">
Locus_2133
</td>
<td headers="variable" class="gt_row gt_left">
CA_rPCA2
</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EBDCEB; color: #000000;">
-0.10
</td>
<td headers="z-score" class="gt_row gt_right">
-4.65
</td>
<td headers="p-value" class="gt_row gt_right">
0
</td>
<td headers="calibrated z-score" class="gt_row gt_right">
14.18
</td>
<td headers="calibrated p-value" class="gt_row gt_right">
0
</td>
<td headers="adjusted p-value" class="gt_row gt_right">
0.04
</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">
Locus_2670
</td>
<td headers="variable" class="gt_row gt_left">
CA_rPCA2
</td>
<td headers="B" class="gt_row gt_right" style="background-color: #EDE1EE; color: #000000;">
-0.08
</td>
<td headers="z-score" class="gt_row gt_right">
-4.88
</td>
<td headers="p-value" class="gt_row gt_right">
0
</td>
<td headers="calibrated z-score" class="gt_row gt_right">
15.61
</td>
<td headers="calibrated p-value" class="gt_row gt_right">
0
</td>
<td headers="adjusted p-value" class="gt_row gt_right">
0.02
</td>
</tr>
<tr>
<td headers="snp" class="gt_row gt_left">
Locus_562
</td>
<td headers="variable" class="gt_row gt_left">
CA_rPCA2
</td>
<td headers="B" class="gt_row gt_right" style="background-color: #F0E7F0; color: #000000;">
-0.06
</td>
<td headers="z-score" class="gt_row gt_right">
-5.04
</td>
<td headers="p-value" class="gt_row gt_right">
0
</td>
<td headers="calibrated z-score" class="gt_row gt_right">
16.62
</td>
<td headers="calibrated p-value" class="gt_row gt_right">
0
</td>
<td headers="adjusted p-value" class="gt_row gt_right">
0.02
</td>
</tr>
</tbody>
<tfoot class="gt_footnotes"><tr>
<td class="gt_footnote" colspan="8">
<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;"><sup>1</sup></span>
LFMM effect size
</td>
</tr></tfoot>
</table>
</div>
</div>
<div class="section level3">
<h3 id="additional-documentation-and-citations">Additional documentation and citations<a class="anchor" aria-label="anchor" href="#additional-documentation-and-citations"></a>
</h3>
<hr>
<table class="table">
<colgroup>
<col width="25%">
<col width="40%">
<col width="34%">
</colgroup>
<thead><tr class="header">
<th></th>
<th>Citation/URL</th>
<th>Details</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Main manuscript describing method</td>
<td>
<a href="https://academic.oup.com/mbe/article/36/4/852/5290100" class="external-link">Caye
et al. 2019</a>; see <a href="https://bcm-uga.github.io/lfmm/index.html" class="external-link">here</a> for
documentation</td>
<td>Paper and code describing LFMM2</td>
</tr>
<tr class="even">
<td>Associated literature</td>
<td><a href="https://academic.oup.com/mbe/article/30/7/1687/972098" class="external-link">Frichot et
al. 2013</a></td>
<td>Paper describing LFMM</td>
</tr>
<tr class="odd">
<td>Associated code</td>
<td><a href="https://cran.r-project.org/web/packages/AssocTests/index.html" class="external-link">Wang
et al. 2020</a></td>
<td>algatr performs a Tracy-Widom test using the <code>tw()</code>
function in the AssocTests package</td>
</tr>
<tr class="even">
<td>Associated code</td>
<td><a href="https://chazhyseni.github.io/NALgen/post/determining_bestk/" class="external-link">Blog
post on automatic K selection</a></td>
<td>algatr’s automatic K selection code</td>
</tr>
<tr class="odd">
<td>Additional code</td>
<td><a href="https://rdrr.io/bioc/LEA/" class="external-link">Frichot &amp; Francois
2015</a></td>
<td>The LEA package provides wrapper functions for LFMM (using the
<code>lfmm2()</code> function</td>
</tr>
</tbody>
</table>
<p>Retrieve LFMM’s vignette:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"lfmm"</span><span class="op">)</span></span>
<span><span class="co">#&gt; starting httpd help server ... done</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anusha Bishop, Anne Chambers, Ian Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
