<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="algatr">
<title>RDA • algatr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="RDA">
<meta property="og:description" content="algatr">
<meta property="og:image" content="https://thewanglab.github.io/algatr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">algatr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference Manual</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <a class="dropdown-item" href="../articles/data_processing_vignette.html">Genetic data processing</a>
    <a class="dropdown-item" href="../articles/enviro_data_vignette.html">Environmental data processing</a>
    <a class="dropdown-item" href="../articles/masking_vignette.html">Masking</a>
    <a class="dropdown-item" href="../articles/gen_dist_vignette.html">Genetic distances</a>
    <a class="dropdown-item" href="../articles/TESS_vignette.html">TESS</a>
    <a class="dropdown-item" href="../articles/MMRR_vignette.html">MMRR</a>
    <a class="dropdown-item" href="../articles/GDM_vignette.html">GDM</a>
    <a class="dropdown-item" href="../articles/RDA_vignette.html">RDA</a>
    <a class="dropdown-item" href="../articles/LFMM_vignette.html">LFMM</a>
    <a class="dropdown-item" href="../articles/wingen_vignette.html">Wingen</a>
    <a class="dropdown-item" href="../articles/alazygatr_vignette.html">Alazygatr</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>RDA</h1>
            
      
      
      <div class="d-none name"><code>RDA_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install packages</span></span>
<span><span class="fu"><a href="../reference/rda_packages.html">rda_packages</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="redundancy-analysis-rda">Redundancy analysis (RDA)<a class="anchor" aria-label="anchor" href="#redundancy-analysis-rda"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thewanglab.github.io/algatr/">algatr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/raster" class="external-link">raster</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vegandevs/vegan" class="external-link">vegan</a></span><span class="op">)</span></span></code></pre></div>
<p><strong>If using RDA, please cite the following: Capblancq T.,
Forester B.R. (2021) Redundancy analysis: A Swiss Army Knife for
landscape genomics. Methods in Ecology and Evolution 12:2298-2309. <a href="DOI:10.1111/2041-210X.13722" class="uri">DOI:10.1111/2041-210X.13722</a></strong></p>
<p>Redundancy analysis (RDA) is a genotype-environment association (GEA)
method that uses constrained ordination to detect outlier loci that are
significantly associated with environmental variables. It does so by
combining regression (in which response variables are genetic data, in
our case, while explanatory variables are environmental data) with
ordination (a PCA). Importantly, RDA is multivariate (i.e., multiple
loci can be considered at a time), which is appealing in cases where
multilocus selection may be occurring.</p>
<p>Model selection is performed by starting with a null model wherein
the response is only explained by an intercept, and environmental
variables are added until the amount of variance explained by a model
that includes all variables is reached. By doing so, this method
minimizes the <em>redundancy</em> among explanatory variables.
Importantly, we can also perform an RDA that accounts for
<em>covariables</em> (or “conditioning variables”) such as those that
account for neutral population structure. When covariables are included,
this is called a <strong>partial RDA</strong>. In algatr, population
structure is quantified by running a PCA on the genetic data, and
selecting a number of PC axes to best represent this type of neutral
variation.</p>
<p>Some of the earliest examples of uses of RDA for identifying
environment-associated loci are <a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1365-294X.2012.05709.x" class="external-link">Lasky
et al. 2012</a> and <a href="https://onlinelibrary.wiley.com/doi/10.1111/mec.13476" class="external-link">Forester et
al. 2016</a>. Since then, there have been several reviews and
walkthroughs of the method that provide additional information,
including workflows and comparisons with other GEA methods (e.g., <a href="https://onlinelibrary.wiley.com/doi/full/10.1111/mec.14584" class="external-link">Forester
et al. 2018</a>). Much of algatr’s code is adapted from one such paper,
<a href="https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722" class="external-link">Capblancq
&amp; Forester 2021</a>, the code of which is available <a href="https://github.com/Capblancq/RDA-landscape-genomics" class="external-link">here</a>.
algatr’s general workflow roughly follows the framework of Capblancq
&amp; Forester (2021).</p>
<p>RDA cannot take in missing values. Imputation based on the per-site
median is commonly performed, but there are several other ways
researchers can deal with missing values. For example, algatr contains
the <code><a href="../reference/str_impute.html">str_impute()</a></code> function to impute missing values based on
population structure using the <code><a href="https://rdrr.io/pkg/LEA/man/impute.html" class="external-link">LEA::impute()</a></code> function.
However, here, we’ll opt to impute to the median, but strongly urge
researchers to use extreme caution when using this form of simplistic
imputation. We mainly provide code to impute on the median for testing
datasets and highly discourage its use in further analyses (please use
<code><a href="../reference/str_impute.html">str_impute()</a></code> instead!).</p>
<p>The general workflow to perform an RDA with algatr is as follows:</p>
<ul>
<li><p><strong>Simple vs. partial RDAs:</strong> We can run a simple RDA
(no covariables) or partial RDA (conditioning covariables included)
using <code><a href="../reference/rda_run.html">rda_run()</a></code></p></li>
<li><p><strong>Variable selection:</strong> Both simple and partial RDAs
can be performed considering all variables (<code>"full"</code> model)
or by performing variable selection to determine variables that
contribute most to genetic variance (<code>"best"</code> model)</p></li>
<li><p><strong>Variance partitioning:</strong> We can perform variance
partitioning considering only variables that contribute significantly to
genetic variance using <code><a href="../reference/rda_varpart.html">rda_varpart()</a></code></p></li>
<li><p><strong>Detecting outlier loci:</strong> Based on the results
from above, we can detect outlier loci with
<code><a href="../reference/rda_getoutliers.html">rda_getoutliers()</a></code></p></li>
</ul>
<div class="section level4">
<h4 id="read-in-and-process-genetic-data">Read in and process genetic data<a class="anchor" aria-label="anchor" href="#read-in-and-process-genetic-data"></a>
</h4>
<p>Running an RDA requires two data files for input: a genotype dosage
matrix (the <code>gen</code> argument) and the environmental values
extracted at sampling coordinates (the <code>env</code> argument). Let’s
first convert our vcf to a dosage matrix using the
<code><a href="../reference/vcf_to_dosage.html">vcf_to_dosage()</a></code> function. <em>N.B.: our code assumes that
sample IDs from our genetic data and our coords are in the same order;
be sure to check this before moving forward!</em></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/load_algatr_example.html">load_algatr_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; ---------------- example dataset ----------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span> </span></span>
<span><span class="co"><span style="color: #0000BB;">#&gt; Objects loaded:</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_vcf*</span> <span style="color: #BBBB00;">vcfR object (1000 loci x 53 samples)</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_gendist*</span> <span style="color: #BBBB00;">genetic distance matrix (Plink Distance)</span> </span></span>
<span><span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; *liz_coords*</span> <span style="color: #00BB00;">dataframe with x and y coordinates</span> </span></span>
<span><span class="co"><span style="color: #BB00BB; font-weight: bold;">#&gt; *CA_env*</span> <span style="color: #BB00BB;">RasterStack with example environmental layers</span> </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; -------------------------------------------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co"># Convert from vcf to dosage matrix:</span></span>
<span><span class="va">gen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_to_dosage.html">vcf_to_dosage</a></span><span class="op">(</span><span class="va">liz_vcf</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required namespace: adegenet</span></span></code></pre></div>
<p>As mentioned above, running an RDA requires that your genotype matrix
contains no missing values. Let’s impute missing values based on the
per-site median. <em>N.B.: this type of simplistic imputation is
strongly not recommended for downstream analyses and is used here for
example’s sake!</em></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Are there NAs in the data?</span></span>
<span><span class="va">gen</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;        Locus_10 Locus_15 Locus_22 Locus_28 Locus_32</span></span>
<span><span class="co">#&gt; ALT3          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BAR360        0        0        0       NA        0</span></span>
<span><span class="co">#&gt; BLL5          0       NA        0        0       NA</span></span>
<span><span class="co">#&gt; BNT5          0        0       NA       NA        0</span></span>
<span><span class="co">#&gt; BOF1          0       NA       NA       NA       NA</span></span>
<span><span class="va">gen</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simple_impute.html">simple_impute</a></span><span class="op">(</span><span class="va">gen</span><span class="op">)</span></span>
<span><span class="co"># Check that NAs are gone</span></span>
<span><span class="va">gen</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;        Locus_10 Locus_15 Locus_22 Locus_28 Locus_32</span></span>
<span><span class="co">#&gt; ALT3          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BAR360        0        0        0        0        0</span></span>
<span><span class="co">#&gt; BLL5          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BNT5          0        0        0        0        0</span></span>
<span><span class="co">#&gt; BOF1          0        0        0        0        0</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="process-environmental-data">Process environmental data<a class="anchor" aria-label="anchor" href="#process-environmental-data"></a>
</h4>
<p>Let’s extract environmental variables using the
<code><a href="https://rspatial.github.io/terra/reference/extract.html" class="external-link">extract()</a></code> function from the raster package. We also need to
standardize environmental variables. This is particularly important if
we are using (for example) bioclimatic variables as input, as units of
measurement are completely different (e.g., mm for precipitation
vs. degrees Celsius for temperature). To do so, we’ll use the
<code><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale()</a></code> function within the raster package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#  Extract values from our environmental raster (CA_env)</span></span>
<span><span class="va">env</span> <span class="op">&lt;-</span> <span class="fu">raster</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/extract.html" class="external-link">extract</a></span><span class="op">(</span><span class="va">CA_env</span>, <span class="va">liz_coords</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Standardize environmental variables and make into dataframe</span></span>
<span><span class="va">env</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">env</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">env</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">env</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="running-simple-and-partial-rdas-using-rda_run-with-and-without-variable-selection">Running simple and partial RDAs using <code>rda_run()</code>, with
and without variable selection<a class="anchor" aria-label="anchor" href="#running-simple-and-partial-rdas-using-rda_run-with-and-without-variable-selection"></a>
</h3>
<hr>
<p>The main function within algatr to perform an RDA is
<code><a href="../reference/rda_run.html">rda_run()</a></code>, which uses the <code><a href="https://rdrr.io/pkg/vegan/man/cca.html" class="external-link">rda()</a></code> function
within the vegan package.</p>
<div class="section level4">
<h4 id="run-a-simple-rda-with-no-variable-selection">Run a simple RDA with no variable selection<a class="anchor" aria-label="anchor" href="#run-a-simple-rda-with-no-variable-selection"></a>
</h4>
<p>A simple RDA is one in which our model will not account for
covariables in the model, and is the default for <code><a href="../reference/rda_run.html">rda_run()</a></code>.
Let’s first run a simple RDA on all environmental variables (i.e., no
variable selection). This is specified using the
<code>model = "full"</code> argument.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod_full</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_run.html">rda_run</a></span><span class="op">(</span><span class="va">gen</span>, <span class="va">env</span>, model <span class="op">=</span> <span class="st">"full"</span><span class="op">)</span></span></code></pre></div>
<p>The resulting object is large, containing 12 elements relevant to the
RDA. Let’s take a look at what function was called. We can see that all
environmental variables (CA_rPCA1, 2, and 3) were included in the model,
and that geography or structure were not included in the model (and
there were no conditioning variables).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod_full</span><span class="op">$</span><span class="va">call</span></span>
<span><span class="co">#&gt; rda(formula = gen ~ CA_rPCA1 + CA_rPCA2 + CA_rPCA3, data = moddf)</span></span></code></pre></div>
<p>Now, let’s take a look at the summary of this model. One of the most
important parts of this object is the partitioning of variance. Within
an RDA, the term “inertia” can be interpreted as variance, and our
results show us that the amount of variance explained by explanatory
variables alone (constrained inertia) is only 9.768%. Unconstrained
inertia is the residual variance, which is very high (90.232%). The
summary also provides site scores, site constraints, and biplot
scores.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mod_full</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; rda(formula = gen ~ CA_rPCA1 + CA_rPCA2 + CA_rPCA3, data = moddf) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Partitioning of variance:</span></span>
<span><span class="co">#&gt;               Inertia Proportion</span></span>
<span><span class="co">#&gt; Total          143.39    1.00000</span></span>
<span><span class="co">#&gt; Constrained     14.01    0.09768</span></span>
<span><span class="co">#&gt; Unconstrained  129.38    0.90232</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvalues, and their contribution to the variance </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Importance of components:</span></span>
<span><span class="co">#&gt;                          RDA1    RDA2    RDA3     PC1     PC2     PC3     PC4</span></span>
<span><span class="co">#&gt; Eigenvalue            7.28806 4.98205 1.73568 18.6608 9.31326 7.51572 5.53312</span></span>
<span><span class="co">#&gt; Proportion Explained  0.05083 0.03474 0.01210  0.1301 0.06495 0.05241 0.03859</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.05083 0.08557 0.09768  0.2278 0.29277 0.34518 0.38377</span></span>
<span><span class="co">#&gt;                           PC5    PC6     PC7     PC8     PC9    PC10    PC11</span></span>
<span><span class="co">#&gt; Eigenvalue            4.51893 4.0443 3.69111 3.30413 3.09940 3.04539 3.00412</span></span>
<span><span class="co">#&gt; Proportion Explained  0.03152 0.0282 0.02574 0.02304 0.02162 0.02124 0.02095</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.41529 0.4435 0.46923 0.49228 0.51389 0.53513 0.55608</span></span>
<span><span class="co">#&gt;                          PC12    PC13    PC14    PC15    PC16   PC17    PC18</span></span>
<span><span class="co">#&gt; Eigenvalue            2.83164 2.80969 2.74342 2.58682 2.47393 2.4664 2.32471</span></span>
<span><span class="co">#&gt; Proportion Explained  0.01975 0.01959 0.01913 0.01804 0.01725 0.0172 0.01621</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.57583 0.59542 0.61456 0.63260 0.64985 0.6671 0.68326</span></span>
<span><span class="co">#&gt;                          PC19    PC20    PC21    PC22    PC23    PC24    PC25</span></span>
<span><span class="co">#&gt; Eigenvalue            2.29159 2.22955 2.15249 2.10481 2.01172 1.94388 1.90538</span></span>
<span><span class="co">#&gt; Proportion Explained  0.01598 0.01555 0.01501 0.01468 0.01403 0.01356 0.01329</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.69925 0.71480 0.72981 0.74449 0.75852 0.77207 0.78536</span></span>
<span><span class="co">#&gt;                          PC26    PC27    PC28    PC29    PC30    PC31    PC32</span></span>
<span><span class="co">#&gt; Eigenvalue            1.82439 1.79687 1.68505 1.64293 1.59963 1.56602 1.50690</span></span>
<span><span class="co">#&gt; Proportion Explained  0.01272 0.01253 0.01175 0.01146 0.01116 0.01092 0.01051</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.79808 0.81062 0.82237 0.83382 0.84498 0.85590 0.86641</span></span>
<span><span class="co">#&gt;                          PC33     PC34    PC35     PC36    PC37     PC38</span></span>
<span><span class="co">#&gt; Eigenvalue            1.46174 1.433419 1.34638 1.331355 1.28477 1.253763</span></span>
<span><span class="co">#&gt; Proportion Explained  0.01019 0.009997 0.00939 0.009285 0.00896 0.008744</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.87661 0.886602 0.89599 0.905277 0.91424 0.922981</span></span>
<span><span class="co">#&gt;                          PC39     PC40    PC41     PC42     PC43     PC44</span></span>
<span><span class="co">#&gt; Eigenvalue            1.23738 1.175289 1.13416 1.072082 1.067608 1.017457</span></span>
<span><span class="co">#&gt; Proportion Explained  0.00863 0.008197 0.00791 0.007477 0.007446 0.007096</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.93161 0.939807 0.94772 0.955193 0.962639 0.969734</span></span>
<span><span class="co">#&gt;                           PC45     PC46     PC47     PC48     PC49</span></span>
<span><span class="co">#&gt; Eigenvalue            0.998927 0.910062 0.866439 0.836444 0.727872</span></span>
<span><span class="co">#&gt; Proportion Explained  0.006967 0.006347 0.006043 0.005833 0.005076</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.976701 0.983048 0.989090 0.994924 1.000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Accumulated constrained eigenvalues</span></span>
<span><span class="co">#&gt; Importance of components:</span></span>
<span><span class="co">#&gt;                         RDA1   RDA2   RDA3</span></span>
<span><span class="co">#&gt; Eigenvalue            7.2881 4.9820 1.7357</span></span>
<span><span class="co">#&gt; Proportion Explained  0.5204 0.3557 0.1239</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.5204 0.8761 1.0000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Scaling 2 for species and site scores</span></span>
<span><span class="co">#&gt; * Species are scaled proportional to eigenvalues</span></span>
<span><span class="co">#&gt; * Sites are unscaled: weighted dispersion equal on all dimensions</span></span>
<span><span class="co">#&gt; * General scaling constant of scores:  9.29244 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Species scores</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                RDA1       RDA2       RDA3        PC1        PC2        PC3</span></span>
<span><span class="co">#&gt; Locus_10  3.073e-02 -1.130e-02  2.569e-02 -1.786e-02  1.512e-02  3.392e-02</span></span>
<span><span class="co">#&gt; Locus_15  8.232e-02 -5.560e-02 -9.949e-02 -9.373e-02  9.032e-03  1.796e-02</span></span>
<span><span class="co">#&gt; Locus_22  9.598e-03  1.021e-03  2.333e-02  1.249e-01 -2.411e-02 -8.721e-03</span></span>
<span><span class="co">#&gt; Locus_28 -7.716e-02 -8.271e-02 -6.968e-02 -2.229e-02  3.464e-02  4.338e-02</span></span>
<span><span class="co">#&gt; Locus_32  1.095e-02 -1.692e-04  3.179e-02 -1.580e-02  7.483e-03  2.773e-02</span></span>
<span><span class="co">#&gt; Locus_35 -4.632e-32 -1.047e-31 -8.290e-32 -6.887e-17 -1.174e-16  4.446e-17</span></span>
<span><span class="co">#&gt; ....                                                                      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Site scores (weighted sums of species scores)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;           RDA1    RDA2    RDA3     PC1      PC2     PC3</span></span>
<span><span class="co">#&gt; ALT3    0.1226 -2.0313 -1.2154 -1.4589 -5.19649 -0.2248</span></span>
<span><span class="co">#&gt; BAR360 -0.1057  0.4007  1.8064  3.0848 -0.65319 -0.1428</span></span>
<span><span class="co">#&gt; BLL5    1.0500  0.2750 -1.8505 -0.3736  0.06138 -0.9738</span></span>
<span><span class="co">#&gt; BNT5   -3.0872 -1.0264 -1.1162 -0.3220  0.95783 -0.5125</span></span>
<span><span class="co">#&gt; BOF1    1.4722 -1.1017  0.2533 -0.2387  0.85134  0.9245</span></span>
<span><span class="co">#&gt; BOT3    0.8133  1.4762 -3.2847 -0.9511 -0.83835 -0.9639</span></span>
<span><span class="co">#&gt; ....                                                   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Site constraints (linear combinations of constraining variables)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;           RDA1    RDA2    RDA3     PC1      PC2     PC3</span></span>
<span><span class="co">#&gt; ALT3   -1.8641 -1.5623 -0.2679 -1.4589 -5.19649 -0.2248</span></span>
<span><span class="co">#&gt; BAR360  0.1641 -0.9647 -0.5288  3.0848 -0.65319 -0.1428</span></span>
<span><span class="co">#&gt; BLL5    1.8430  0.6239 -2.8780 -0.3736  0.06138 -0.9738</span></span>
<span><span class="co">#&gt; BNT5   -2.2991 -1.2711 -0.6227 -0.3220  0.95783 -0.5125</span></span>
<span><span class="co">#&gt; BOF1    1.4722 -1.0680  0.1644 -0.2387  0.85134  0.9245</span></span>
<span><span class="co">#&gt; BOT3    0.9142  1.3548 -2.0063 -0.9511 -0.83835 -0.9639</span></span>
<span><span class="co">#&gt; ....                                                   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Biplot scores for constraining variables</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             RDA1    RDA2    RDA3 PC1 PC2 PC3</span></span>
<span><span class="co">#&gt; CA_rPCA1 -0.2047  0.5350  0.8197   0   0   0</span></span>
<span><span class="co">#&gt; CA_rPCA2  0.5188  0.8393 -0.1625   0   0   0</span></span>
<span><span class="co">#&gt; CA_rPCA3  0.5561 -0.6174  0.5564   0   0   0</span></span></code></pre></div>
<p>One of the most relevant statistics to present from an RDA model is
the adjusted R2 value. The R2 value <em>must</em> be adjusted; if not,
it is biased because it will always increase if independent variables
are added (i.e., there is no penalization for adding independent
variables that aren’t significantly affecting dependent variables within
the model). The adjusted R2 value from the full (global) model can also
help us determine the stopping point when we go on to do variable
selection. As we can see, the unadjusted R2 is 0.098, while the adjusted
value is 0.042. Be sure to always report <em>adjusted</em> R2
values.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/RsquareAdj.html" class="external-link">RsquareAdj</a></span><span class="op">(</span><span class="va">mod_full</span><span class="op">)</span></span>
<span><span class="co">#&gt; $r.squared</span></span>
<span><span class="co">#&gt; [1] 0.0976769</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $adj.r.squared</span></span>
<span><span class="co">#&gt; [1] 0.04243263</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="run-a-simple-rda-with-variable-selection">Run a simple RDA with variable selection<a class="anchor" aria-label="anchor" href="#run-a-simple-rda-with-variable-selection"></a>
</h4>
<p>Now, let’s run an RDA model with variable selection by specifying the
<code>"best"</code> model within <code><a href="../reference/rda_run.html">rda_run()</a></code>. Variable
selection occurs using the <code><a href="https://rdrr.io/pkg/vegan/man/ordistep.html" class="external-link">ordiR2step()</a></code> function in the
vegan package, and is a forward selection method that begins with a null
model and adds explanatory variables one at a time until genetic
variance is maximized. To perform forward selection, we need to specify
stopping criteria (i.e., when to stop adding more explanatory
variables). There are two primary ways to specify stopping criteria:
parameters involved in a permutation-based significance test (two
parameters: the limits of permutation p-values for adding
(<code>Pin</code>) or dropping (<code>Pout</code>) a term to the model,
and the number of permutations used in the estimation of adjusted R2
(<code>R2permutations</code>)), and whether to use the adjusted R2 value
as the stopping criterion (<code>R2scope</code>; only models with
adjusted R2 values lower than specified are accepted).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod_best</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_run.html">rda_run</a></span><span class="op">(</span><span class="va">gen</span>, <span class="va">env</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"best"</span>,</span>
<span>  Pin <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  R2permutations <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  R2scope <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Step: R2.adj= 0 </span></span>
<span><span class="co">#&gt; Call: gen ~ 1 </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;                  R2.adjusted</span></span>
<span><span class="co">#&gt; &lt;All variables&gt; 0.0424326339</span></span>
<span><span class="co">#&gt; + CA_rPCA2      0.0196214014</span></span>
<span><span class="co">#&gt; + CA_rPCA3      0.0137427906</span></span>
<span><span class="co">#&gt; + CA_rPCA1      0.0009953656</span></span>
<span><span class="co">#&gt; &lt;none&gt;          0.0000000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            Df    AIC      F Pr(&gt;F)   </span></span>
<span><span class="co">#&gt; + CA_rPCA2  1 264.09 2.0407  0.004 **</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Step: R2.adj= 0.0196214 </span></span>
<span><span class="co">#&gt; Call: gen ~ CA_rPCA2 </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;                 R2.adjusted</span></span>
<span><span class="co">#&gt; &lt;All variables&gt;  0.04243263</span></span>
<span><span class="co">#&gt; + CA_rPCA3       0.03919420</span></span>
<span><span class="co">#&gt; + CA_rPCA1       0.01991617</span></span>
<span><span class="co">#&gt; &lt;none&gt;           0.01962140</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            Df    AIC      F Pr(&gt;F)   </span></span>
<span><span class="co">#&gt; + CA_rPCA3  1 263.97 2.0389  0.002 **</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Step: R2.adj= 0.0391942 </span></span>
<span><span class="co">#&gt; Call: gen ~ CA_rPCA2 + CA_rPCA3 </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;                 R2.adjusted</span></span>
<span><span class="co">#&gt; &lt;All variables&gt;  0.04243263</span></span>
<span><span class="co">#&gt; + CA_rPCA1       0.04243263</span></span>
<span><span class="co">#&gt; &lt;none&gt;           0.03919420</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            Df    AIC      F Pr(&gt;F)</span></span>
<span><span class="co">#&gt; + CA_rPCA1  1 264.72 1.1691  0.172</span></span></code></pre></div>
<p>Let’s look at our best model, and see how it compares to our full
model. There is also an additional element in the <code>mod_best</code>
object, which are the results from an ANOVA-like permutation test. When
we look at the call, we can now see that only environmental PCs 2 and 3
are considered in the model, meaning that PC1 is not considered to be
significantly associated with genetic variance. Importantly, our
adjusted R2 value is 0.039 (compared to 0.042 for the full model), which
tells us that the model with only two of the environmental variables
explains nearly as much as the RDA with all three.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod_best</span><span class="op">$</span><span class="va">call</span></span>
<span><span class="co">#&gt; rda(formula = gen ~ CA_rPCA2 + CA_rPCA3, data = moddf)</span></span>
<span><span class="va">mod_best</span><span class="op">$</span><span class="va">anova</span></span>
<span><span class="co">#&gt;                   R2.adj Df    AIC      F Pr(&gt;F)   </span></span>
<span><span class="co">#&gt; + CA_rPCA2      0.019621  1 264.09 2.0407  0.004 **</span></span>
<span><span class="co">#&gt; + CA_rPCA3      0.039194  1 263.97 2.0389  0.002 **</span></span>
<span><span class="co">#&gt; &lt;All variables&gt; 0.042433                           </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/RsquareAdj.html" class="external-link">RsquareAdj</a></span><span class="op">(</span><span class="va">mod_best</span><span class="op">)</span></span>
<span><span class="co">#&gt; $r.squared</span></span>
<span><span class="co">#&gt; [1] 0.07614827</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $adj.r.squared</span></span>
<span><span class="co">#&gt; [1] 0.0391942</span></span></code></pre></div>
<p>Following variable selection, we want to retain two environmental
variables (CA_rPCA2 and CA_rPCA3) for all subsequent analyses.</p>
</div>
<div class="section level4">
<h4 id="run-a-partial-rda-with-geography-as-a-covariable">Run a partial RDA with geography as a covariable<a class="anchor" aria-label="anchor" href="#run-a-partial-rda-with-geography-as-a-covariable"></a>
</h4>
<p>A partial RDA, as described at the beginning of the vignette, is when
covariables (or conditioning variables) are incorporated into the RDA.
Within <code><a href="../reference/rda_run.html">rda_run()</a></code>, to correct for geography, users can set
<code>correctGEO = TRUE</code>, in which case sampling coordinates
(<code>coords</code>) must also be provided. Let’s now run a partial
RDA, without any variable selection (the full model) and view the
results.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod_pRDA_geo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_run.html">rda_run</a></span><span class="op">(</span><span class="va">gen</span>, <span class="va">env</span>, <span class="va">liz_coords</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"full"</span>,</span>
<span>  correctGEO <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  correctPC <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Let’s take a look at our partial RDA results. Note that now, the call
includes conditioning variables x and y, which correspond to our
latitude and longitude values. Conditioning variables are specified
within the <code><a href="https://rdrr.io/pkg/vegan/man/cca.html" class="external-link">rda()</a></code> function using the
<code>Condition()</code> function. Finally, when we look at the summary
of the resulting object, we can see that along with total, constrained,
and unconstrained inertia (or variance) as with our simple RDA, there is
an additional row for conditioned inertia because our partial RDA
includes conditioning variables. Interestingly, the conditioned inertia
is higher than constrained inertia, suggesting that geography is more
associated with genetic variance than our environmental variables
are.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html" class="external-link">anova</a></span><span class="op">(</span><span class="va">mod_pRDA_geo</span><span class="op">)</span></span>
<span><span class="co">#&gt; Permutation test for rda under reduced model</span></span>
<span><span class="co">#&gt; Permutation: free</span></span>
<span><span class="co">#&gt; Number of permutations: 999</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Model: rda(formula = gen ~ CA_rPCA1 + CA_rPCA2 + CA_rPCA3 + Condition(x + y), data = moddf)</span></span>
<span><span class="co">#&gt;          Df Variance      F Pr(&gt;F)    </span></span>
<span><span class="co">#&gt; Model     3   11.306 1.5948  0.001 ***</span></span>
<span><span class="co">#&gt; Residual 47  111.068                  </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/RsquareAdj.html" class="external-link">RsquareAdj</a></span><span class="op">(</span><span class="va">mod_pRDA_geo</span><span class="op">)</span> <span class="co"># 0.0305</span></span>
<span><span class="co">#&gt; $r.squared</span></span>
<span><span class="co">#&gt; [1] 0.07884828</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $adj.r.squared</span></span>
<span><span class="co">#&gt; [1] 0.03058243</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mod_pRDA_geo</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; rda(formula = gen ~ CA_rPCA1 + CA_rPCA2 + CA_rPCA3 + Condition(x +      y), data = moddf) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Partitioning of variance:</span></span>
<span><span class="co">#&gt;               Inertia Proportion</span></span>
<span><span class="co">#&gt; Total          143.39    1.00000</span></span>
<span><span class="co">#&gt; Conditioned     21.01    0.14656</span></span>
<span><span class="co">#&gt; Constrained     11.31    0.07885</span></span>
<span><span class="co">#&gt; Unconstrained  111.07    0.77459</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvalues, and their contribution to the variance </span></span>
<span><span class="co">#&gt; after removing the contribution of conditiniong variables</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Importance of components:</span></span>
<span><span class="co">#&gt;                          RDA1    RDA2    RDA3      PC1     PC2     PC3     PC4</span></span>
<span><span class="co">#&gt; Eigenvalue            6.87372 2.80216 1.63009 11.00385 7.49907 5.16346 4.68171</span></span>
<span><span class="co">#&gt; Proportion Explained  0.05617 0.02290 0.01332  0.08992 0.06128 0.04219 0.03826</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.05617 0.07907 0.09239  0.18231 0.24359 0.28578 0.32404</span></span>
<span><span class="co">#&gt;                           PC5     PC6     PC7     PC8     PC9    PC10    PC11</span></span>
<span><span class="co">#&gt; Eigenvalue            4.21148 3.72312 3.45801 3.19495 3.01860 2.98188 2.93280</span></span>
<span><span class="co">#&gt; Proportion Explained  0.03441 0.03042 0.02826 0.02611 0.02467 0.02437 0.02397</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.35845 0.38888 0.41714 0.44324 0.46791 0.49228 0.51624</span></span>
<span><span class="co">#&gt;                          PC12    PC13   PC14    PC15    PC16    PC17    PC18</span></span>
<span><span class="co">#&gt; Eigenvalue            2.79673 2.59175 2.5826 2.49887 2.45425 2.32321 2.29777</span></span>
<span><span class="co">#&gt; Proportion Explained  0.02285 0.02118 0.0211 0.02042 0.02006 0.01898 0.01878</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.53910 0.56028 0.5814 0.60180 0.62186 0.64084 0.65962</span></span>
<span><span class="co">#&gt;                          PC19    PC20    PC21    PC22    PC23   PC24    PC25</span></span>
<span><span class="co">#&gt; Eigenvalue            2.15916 2.14664 2.03834 2.00771 1.91618 1.8479 1.80441</span></span>
<span><span class="co">#&gt; Proportion Explained  0.01764 0.01754 0.01666 0.01641 0.01566 0.0151 0.01475</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.67726 0.69480 0.71146 0.72787 0.74352 0.7586 0.77337</span></span>
<span><span class="co">#&gt;                         PC26    PC27    PC28    PC29    PC30    PC31    PC32</span></span>
<span><span class="co">#&gt; Eigenvalue            1.7740 1.68265 1.64600 1.58626 1.52630 1.47782 1.45805</span></span>
<span><span class="co">#&gt; Proportion Explained  0.0145 0.01375 0.01345 0.01296 0.01247 0.01208 0.01191</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.7879 0.80162 0.81507 0.82803 0.84050 0.85258 0.86449</span></span>
<span><span class="co">#&gt;                          PC33    PC34    PC35    PC36    PC37     PC38     PC39</span></span>
<span><span class="co">#&gt; Eigenvalue            1.42866 1.37519 1.30318 1.27465 1.24443 1.221578 1.169694</span></span>
<span><span class="co">#&gt; Proportion Explained  0.01167 0.01124 0.01065 0.01042 0.01017 0.009982 0.009558</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.87617 0.88741 0.89805 0.90847 0.91864 0.928622 0.938181</span></span>
<span><span class="co">#&gt;                           PC40     PC41     PC42     PC43     PC44     PC45</span></span>
<span><span class="co">#&gt; Eigenvalue            1.073702 1.068009 1.045150 1.006285 0.935246 0.866846</span></span>
<span><span class="co">#&gt; Proportion Explained  0.008774 0.008727 0.008541 0.008223 0.007643 0.007084</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.946955 0.955682 0.964223 0.972446 0.980088 0.987172</span></span>
<span><span class="co">#&gt;                           PC46     PC47</span></span>
<span><span class="co">#&gt; Eigenvalue            0.841414 0.728431</span></span>
<span><span class="co">#&gt; Proportion Explained  0.006876 0.005952</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.994048 1.000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Accumulated constrained eigenvalues</span></span>
<span><span class="co">#&gt; Importance of components:</span></span>
<span><span class="co">#&gt;                        RDA1   RDA2   RDA3</span></span>
<span><span class="co">#&gt; Eigenvalue            6.874 2.8022 1.6301</span></span>
<span><span class="co">#&gt; Proportion Explained  0.608 0.2478 0.1442</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.608 0.8558 1.0000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Scaling 2 for species and site scores</span></span>
<span><span class="co">#&gt; * Species are scaled proportional to eigenvalues</span></span>
<span><span class="co">#&gt; * Sites are unscaled: weighted dispersion equal on all dimensions</span></span>
<span><span class="co">#&gt; * General scaling constant of scores:  9.29244 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Species scores</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                RDA1       RDA2       RDA3        PC1        PC2        PC3</span></span>
<span><span class="co">#&gt; Locus_10  3.455e-02  8.516e-03  2.220e-02  2.631e-02  3.045e-02 -5.347e-03</span></span>
<span><span class="co">#&gt; Locus_15  3.361e-02 -8.041e-02 -7.979e-02  5.156e-02  3.933e-02 -1.483e-01</span></span>
<span><span class="co">#&gt; Locus_22  4.500e-02  3.760e-03  1.079e-02 -8.323e-02 -1.622e-02 -5.073e-02</span></span>
<span><span class="co">#&gt; Locus_28 -6.112e-02 -6.035e-02 -6.888e-02  3.955e-02  3.835e-02  1.034e-02</span></span>
<span><span class="co">#&gt; Locus_32  1.352e-02  1.105e-02  2.971e-02  1.883e-02  2.562e-02  1.820e-02</span></span>
<span><span class="co">#&gt; Locus_35 -5.728e-32 -2.203e-31 -4.550e-32  5.385e-17 -1.621e-16  6.186e-17</span></span>
<span><span class="co">#&gt; ....                                                                      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Site scores (weighted sums of species scores)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;           RDA1    RDA2    RDA3     PC1     PC2      PC3</span></span>
<span><span class="co">#&gt; ALT3   -0.6074  1.5856  0.3496 -3.5502  0.5751  1.36394</span></span>
<span><span class="co">#&gt; BAR360  1.9028 -0.7660 -0.6439 -2.1245 -0.2723  0.32928</span></span>
<span><span class="co">#&gt; BLL5    0.6935  1.9687 -2.1865 -0.8111 -1.6626  0.32307</span></span>
<span><span class="co">#&gt; BNT5   -2.6158 -0.2921 -1.0298  0.5356 -0.7971  0.92231</span></span>
<span><span class="co">#&gt; BOF1    1.5909 -0.2182  0.1445  0.9682  0.7195  0.00126</span></span>
<span><span class="co">#&gt; BOT3   -1.1366 -1.4518 -1.2000 -0.1237 -0.2652 -1.77037</span></span>
<span><span class="co">#&gt; ....                                                   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Site constraints (linear combinations of constraining variables)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;           RDA1     RDA2     RDA3     PC1     PC2      PC3</span></span>
<span><span class="co">#&gt; ALT3   -1.4796  1.42372 -0.56620 -3.5502  0.5751  1.36394</span></span>
<span><span class="co">#&gt; BAR360  0.9953 -1.10207 -0.72739 -2.1245 -0.2723  0.32928</span></span>
<span><span class="co">#&gt; BLL5    0.7741  1.44224 -2.79896 -0.8111 -1.6626  0.32307</span></span>
<span><span class="co">#&gt; BNT5   -2.0626 -0.81597 -0.56826  0.5356 -0.7971  0.92231</span></span>
<span><span class="co">#&gt; BOF1    1.8148 -0.09947 -0.03338  0.9682  0.7195  0.00126</span></span>
<span><span class="co">#&gt; BOT3   -0.4265 -0.95308 -1.45302 -0.1237 -0.2652 -1.77037</span></span>
<span><span class="co">#&gt; ....                                                     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Biplot scores for constraining variables</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             RDA1    RDA2     RDA3 PC1 PC2 PC3</span></span>
<span><span class="co">#&gt; CA_rPCA1 -0.2345  0.4009  0.82275   0   0   0</span></span>
<span><span class="co">#&gt; CA_rPCA2  0.2021  0.3349 -0.08967   0   0   0</span></span>
<span><span class="co">#&gt; CA_rPCA3  0.6824 -0.2889  0.51661   0   0   0</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="run-a-partial-rda-with-population-genetic-structure-and-geography-as-covariables">Run a partial RDA with population genetic structure and geography as
covariables<a class="anchor" aria-label="anchor" href="#run-a-partial-rda-with-population-genetic-structure-and-geography-as-covariables"></a>
</h4>
<p>Lastly, let’s run a partial RDA with both geography
(<code>correctGEO = TRUE</code>) and population structure
(<code>correctPC = TRUE</code>) as conditioning variables. Within
algatr, population structure is quantified using a PCA, and a user can
specify the number of axes to retain in the RDA model. In this case,
we’ll retain three axes, which we can specify using the <code>nPC</code>
argument. If this argument is set to <code>"manual"</code>, a screeplot
will be printed and a user can manually enter the number of PCs that
they feel best describe the data.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod_pRDA_gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_run.html">rda_run</a></span><span class="op">(</span><span class="va">gen</span>, <span class="va">env</span>, <span class="va">liz_coords</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"full"</span>,</span>
<span>  correctGEO <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  correctPC <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  nPC <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="RDA_vignette_files/figure-html/pRDA%20struct+geo-1.png" width="700"></p>
<p>Now, we see that even more inertia/variance is explained by the
conditioning variables.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mod_pRDA_gs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; rda(formula = gen ~ CA_rPCA1 + CA_rPCA2 + CA_rPCA3 + Condition(PC1 +      PC2 + PC3 + x + y), data = moddf) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Partitioning of variance:</span></span>
<span><span class="co">#&gt;               Inertia Proportion</span></span>
<span><span class="co">#&gt; Total         143.389    1.00000</span></span>
<span><span class="co">#&gt; Conditioned    48.063    0.33519</span></span>
<span><span class="co">#&gt; Constrained     6.672    0.04653</span></span>
<span><span class="co">#&gt; Unconstrained  88.654    0.61828</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvalues, and their contribution to the variance </span></span>
<span><span class="co">#&gt; after removing the contribution of conditiniong variables</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Importance of components:</span></span>
<span><span class="co">#&gt;                          RDA1    RDA2    RDA3     PC1     PC2     PC3     PC4</span></span>
<span><span class="co">#&gt; Eigenvalue            2.72234 2.30006 1.64942 5.15035 4.44551 3.82383 3.55107</span></span>
<span><span class="co">#&gt; Proportion Explained  0.02856 0.02413 0.01730 0.05403 0.04663 0.04011 0.03725</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.02856 0.05269 0.06999 0.12402 0.17065 0.21077 0.24802</span></span>
<span><span class="co">#&gt;                           PC5     PC6     PC7     PC8     PC9    PC10    PC11</span></span>
<span><span class="co">#&gt; Eigenvalue            3.24764 3.02810 2.99645 2.95013 2.79913 2.65085 2.59129</span></span>
<span><span class="co">#&gt; Proportion Explained  0.03407 0.03177 0.03143 0.03095 0.02936 0.02781 0.02718</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.28209 0.31385 0.34529 0.37623 0.40560 0.43341 0.46059</span></span>
<span><span class="co">#&gt;                          PC12    PC13    PC14    PC15    PC16    PC17    PC18</span></span>
<span><span class="co">#&gt; Eigenvalue            2.52112 2.46520 2.34331 2.30029 2.16258 2.15304 2.05042</span></span>
<span><span class="co">#&gt; Proportion Explained  0.02645 0.02586 0.02458 0.02413 0.02269 0.02259 0.02151</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.48704 0.51290 0.53748 0.56161 0.58430 0.60688 0.62839</span></span>
<span><span class="co">#&gt;                          PC19    PC20    PC21   PC22    PC23    PC24    PC25</span></span>
<span><span class="co">#&gt; Eigenvalue            2.01366 1.92134 1.85435 1.8112 1.79300 1.68266 1.65184</span></span>
<span><span class="co">#&gt; Proportion Explained  0.02112 0.02016 0.01945 0.0190 0.01881 0.01765 0.01733</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.64952 0.66967 0.68912 0.7081 0.72693 0.74458 0.76191</span></span>
<span><span class="co">#&gt;                          PC26    PC27    PC28    PC29    PC30    PC31    PC32</span></span>
<span><span class="co">#&gt; Eigenvalue            1.58977 1.52980 1.47897 1.46048 1.44680 1.38037 1.30440</span></span>
<span><span class="co">#&gt; Proportion Explained  0.01668 0.01605 0.01551 0.01532 0.01518 0.01448 0.01368</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.77859 0.79464 0.81015 0.82547 0.84065 0.85513 0.86882</span></span>
<span><span class="co">#&gt;                          PC33    PC34    PC35    PC36    PC37    PC38    PC39</span></span>
<span><span class="co">#&gt; Eigenvalue            1.27639 1.24668 1.22365 1.17070 1.07515 1.06819 1.05602</span></span>
<span><span class="co">#&gt; Proportion Explained  0.01339 0.01308 0.01284 0.01228 0.01128 0.01121 0.01108</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.88221 0.89528 0.90812 0.92040 0.93168 0.94289 0.95396</span></span>
<span><span class="co">#&gt;                          PC40     PC41    PC42     PC43     PC44</span></span>
<span><span class="co">#&gt; Eigenvalue            1.00764 0.940164 0.86839 0.842461 0.729857</span></span>
<span><span class="co">#&gt; Proportion Explained  0.01057 0.009863 0.00911 0.008838 0.007656</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.96453 0.974396 0.98351 0.992344 1.000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Accumulated constrained eigenvalues</span></span>
<span><span class="co">#&gt; Importance of components:</span></span>
<span><span class="co">#&gt;                        RDA1   RDA2   RDA3</span></span>
<span><span class="co">#&gt; Eigenvalue            2.722 2.3001 1.6494</span></span>
<span><span class="co">#&gt; Proportion Explained  0.408 0.3447 0.2472</span></span>
<span><span class="co">#&gt; Cumulative Proportion 0.408 0.7528 1.0000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Scaling 2 for species and site scores</span></span>
<span><span class="co">#&gt; * Species are scaled proportional to eigenvalues</span></span>
<span><span class="co">#&gt; * Sites are unscaled: weighted dispersion equal on all dimensions</span></span>
<span><span class="co">#&gt; * General scaling constant of scores:  9.29244 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Species scores</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                RDA1       RDA2       RDA3        PC1        PC2        PC3</span></span>
<span><span class="co">#&gt; Locus_10 -1.265e-02 -1.062e-02  2.431e-02  9.693e-03 -2.614e-03 -1.546e-03</span></span>
<span><span class="co">#&gt; Locus_15  7.156e-02 -4.414e-02 -7.566e-02  1.528e-01 -1.710e-02  9.218e-02</span></span>
<span><span class="co">#&gt; Locus_22  6.219e-03  1.821e-04  1.421e-02  3.812e-02 -1.478e-02 -2.112e-02</span></span>
<span><span class="co">#&gt; Locus_28  7.276e-02  6.051e-02 -7.430e-02 -2.306e-02 -2.088e-02  9.904e-03</span></span>
<span><span class="co">#&gt; Locus_32 -1.466e-02  1.498e-03  2.902e-02 -1.417e-02 -3.597e-03  4.626e-02</span></span>
<span><span class="co">#&gt; Locus_35  5.557e-32 -1.811e-32 -4.247e-32  6.817e-17 -7.435e-17  1.546e-18</span></span>
<span><span class="co">#&gt; ....                                                                      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Site scores (weighted sums of species scores)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            RDA1     RDA2    RDA3      PC1      PC2      PC3</span></span>
<span><span class="co">#&gt; ALT3   -3.04653  0.18212 -1.0379 -0.18125  1.19052 -0.81292</span></span>
<span><span class="co">#&gt; BAR360  1.30674  0.16343 -0.8067 -0.88740 -0.67411  0.08499</span></span>
<span><span class="co">#&gt; BLL5   -0.49562  0.28265 -1.8134 -1.24385  0.01825  0.64437</span></span>
<span><span class="co">#&gt; BNT5    1.40006  2.26681 -0.9036 -1.60174  0.47698  0.20989</span></span>
<span><span class="co">#&gt; BOF1    0.02238 -0.58659  0.3529  0.09616  0.13024 -0.17068</span></span>
<span><span class="co">#&gt; BOT3    1.17354 -0.09847 -1.3900  1.99555 -0.56848  0.21693</span></span>
<span><span class="co">#&gt; ....                                                       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Site constraints (linear combinations of constraining variables)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            RDA1     RDA2    RDA3      PC1      PC2      PC3</span></span>
<span><span class="co">#&gt; ALT3   -2.30233  0.17848 -1.2731 -0.18125  1.19052 -0.81292</span></span>
<span><span class="co">#&gt; BAR360  1.37941 -0.06513 -0.6461 -0.88740 -0.67411  0.08499</span></span>
<span><span class="co">#&gt; BLL5   -0.43214  0.25009 -2.2783 -1.24385  0.01825  0.64437</span></span>
<span><span class="co">#&gt; BNT5    1.44455  1.47169 -0.5322 -1.60174  0.47698  0.20989</span></span>
<span><span class="co">#&gt; BOF1   -0.06742 -1.15996  0.1743  0.09616  0.13024 -0.17068</span></span>
<span><span class="co">#&gt; BOT3    0.55754 -0.97519 -1.5380  1.99555 -0.56848  0.21693</span></span>
<span><span class="co">#&gt; ....                                                       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Biplot scores for constraining variables</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             RDA1     RDA2     RDA3 PC1 PC2 PC3</span></span>
<span><span class="co">#&gt; CA_rPCA1 -0.3709  0.31874  0.78617   0   0   0</span></span>
<span><span class="co">#&gt; CA_rPCA2 -0.3373 -0.09457 -0.06287   0   0   0</span></span>
<span><span class="co">#&gt; CA_rPCA3  0.2301 -0.44499  0.58235   0   0   0</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="variance-partitioning-with-partial-rda">Variance partitioning with partial RDA<a class="anchor" aria-label="anchor" href="#variance-partitioning-with-partial-rda"></a>
</h3>
<hr>
<p>In many cases, we want to understand the independent contributions of
each explanatory variable in combination with our covariables. Also, we
can understand how much confounded variance there is (i.e., variation
that is explained by the combination of multiple explanatory variables).
The best way to do this is to run partial RDAs on each set of
explanatory variables and examine the inertia of each pRDA result as it
compares to the full model. In this case, the full model is an RDA that
considers neutral processes as explanatory variables rather than
conditioning variables (covariables).</p>
<p>The call for this full model looks like this:
<code>rda(gen ~ env + covariable, data)</code>, rather than a true
partial RDA which would take the form:
<code>rda(gen ~ env + Condition(covariable), data)</code>.</p>
<p>In algatr, we can run variance partitioning using the function
<code><a href="../reference/rda_varpart.html">rda_varpart()</a></code>, which will run the full model (as above) and
then will run successive pRDAs which are conditioned on three successive
sets of variables. These variable sets are significant environmental
variables (in our case, environmental PCs 2 and 3), population genetic
structure (once again determined using a PC with a user-specified number
of PC axes), and geography (sampling coordinates). This function will
then provide us with information on how variance is partitioned in the
form of a table containing relevant statistics; this table is nearly
identical to Table 2 from <a href="https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722" class="external-link">Capblancq
&amp; Forester 2021</a>. The <code><a href="../reference/rda_varpart.html">rda_varpart()</a></code> parameters are
identical to those from <code><a href="../reference/rda_run.html">rda_run()</a></code>. As with
<code><a href="../reference/rda_run.html">rda_run()</a></code>, we can provide a number for the number of PCs to
consider for population structure (<code>nPC</code>), or this argument
can be set to <code>"manual"</code> if a user would like to manually
enter a number into the console after viewing the screeplot. We can then
use <code><a href="../reference/rda_varpart_table.html">rda_varpart_table()</a></code> to visualize these results in an
aesthetically pleasing way; users can specify whether to display the
function call in the table (<code>call_col</code>; defaults to
<code>FALSE</code>) and the number of digits to round to
<code>digits</code> (the default is 2).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">varpart</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_varpart.html">rda_varpart</a></span><span class="op">(</span><span class="va">gen</span>, <span class="va">env</span>, <span class="va">liz_coords</span>,</span>
<span>  Pin <span class="op">=</span> <span class="fl">0.05</span>, R2permutations <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  R2scope <span class="op">=</span> <span class="cn">T</span>, nPC <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Step: R2.adj= 0 </span></span>
<span><span class="co">#&gt; Call: gen ~ 1 </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;                  R2.adjusted</span></span>
<span><span class="co">#&gt; &lt;All variables&gt; 0.0424326339</span></span>
<span><span class="co">#&gt; + CA_rPCA2      0.0196214014</span></span>
<span><span class="co">#&gt; + CA_rPCA3      0.0137427906</span></span>
<span><span class="co">#&gt; + CA_rPCA1      0.0009953656</span></span>
<span><span class="co">#&gt; &lt;none&gt;          0.0000000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            Df    AIC      F Pr(&gt;F)   </span></span>
<span><span class="co">#&gt; + CA_rPCA2  1 264.09 2.0407  0.004 **</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Step: R2.adj= 0.0196214 </span></span>
<span><span class="co">#&gt; Call: gen ~ CA_rPCA2 </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;                 R2.adjusted</span></span>
<span><span class="co">#&gt; &lt;All variables&gt;  0.04243263</span></span>
<span><span class="co">#&gt; + CA_rPCA3       0.03919420</span></span>
<span><span class="co">#&gt; + CA_rPCA1       0.01991617</span></span>
<span><span class="co">#&gt; &lt;none&gt;           0.01962140</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            Df    AIC      F Pr(&gt;F)   </span></span>
<span><span class="co">#&gt; + CA_rPCA3  1 263.97 2.0389  0.002 **</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Step: R2.adj= 0.0391942 </span></span>
<span><span class="co">#&gt; Call: gen ~ CA_rPCA2 + CA_rPCA3 </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;                 R2.adjusted</span></span>
<span><span class="co">#&gt; &lt;All variables&gt;  0.04243263</span></span>
<span><span class="co">#&gt; + CA_rPCA1       0.04243263</span></span>
<span><span class="co">#&gt; &lt;none&gt;           0.03919420</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            Df    AIC      F Pr(&gt;F)</span></span>
<span><span class="co">#&gt; + CA_rPCA1  1 264.72 1.1691  0.166</span></span></code></pre></div>
<p><img src="RDA_vignette_files/figure-html/varpart-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rda_varpart_table.html">rda_varpart_table</a></span><span class="op">(</span><span class="va">varpart</span>, call_col <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div id="cxypfadxww" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#cxypfadxww table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#cxypfadxww thead, #cxypfadxww tbody, #cxypfadxww tfoot, #cxypfadxww tr, #cxypfadxww td, #cxypfadxww th {
  border-style: none;
}

#cxypfadxww p {
  margin: 0;
  padding: 0;
}

#cxypfadxww .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#cxypfadxww .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#cxypfadxww .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#cxypfadxww .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#cxypfadxww .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#cxypfadxww .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#cxypfadxww .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#cxypfadxww .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#cxypfadxww .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#cxypfadxww .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#cxypfadxww .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#cxypfadxww .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#cxypfadxww .gt_spanner_row {
  border-bottom-style: hidden;
}

#cxypfadxww .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#cxypfadxww .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#cxypfadxww .gt_from_md > :first-child {
  margin-top: 0;
}

#cxypfadxww .gt_from_md > :last-child {
  margin-bottom: 0;
}

#cxypfadxww .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#cxypfadxww .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#cxypfadxww .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#cxypfadxww .gt_row_group_first td {
  border-top-width: 2px;
}

#cxypfadxww .gt_row_group_first th {
  border-top-width: 2px;
}

#cxypfadxww .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#cxypfadxww .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#cxypfadxww .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#cxypfadxww .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#cxypfadxww .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#cxypfadxww .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#cxypfadxww .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#cxypfadxww .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#cxypfadxww .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#cxypfadxww .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#cxypfadxww .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#cxypfadxww .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#cxypfadxww .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#cxypfadxww .gt_left {
  text-align: left;
}

#cxypfadxww .gt_center {
  text-align: center;
}

#cxypfadxww .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#cxypfadxww .gt_font_normal {
  font-weight: normal;
}

#cxypfadxww .gt_font_bold {
  font-weight: bold;
}

#cxypfadxww .gt_font_italic {
  font-style: italic;
}

#cxypfadxww .gt_super {
  font-size: 65%;
}

#cxypfadxww .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#cxypfadxww .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#cxypfadxww .gt_indent_1 {
  text-indent: 5px;
}

#cxypfadxww .gt_indent_2 {
  text-indent: 10px;
}

#cxypfadxww .gt_indent_3 {
  text-indent: 15px;
}

#cxypfadxww .gt_indent_4 {
  text-indent: 20px;
}

#cxypfadxww .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="gt_heading">
<td colspan="8" class="gt_heading gt_title gt_font_normal gt_bottom_border" style>Variance partitioning</td>
    </tr>
<tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="Model">Model</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="pRDA model call">pRDA model call</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="Inertia">Inertia</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="R2">R2</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="Adjusted R2">Adjusted R2</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p (&amp;gt;F)">p (&gt;F)</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="Prop. of explainable variance">Prop. of explainable variance</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="Prop. of total variance">Prop. of total variance</th>
    </tr>
</thead>
<tbody class="gt_table_body">
<tr>
<td headers="Model" class="gt_row gt_left">Full model</td>
<td headers="pRDA model call" class="gt_row gt_left">gen ~ CA_rPCA2 + CA_rPCA3 + PC1 + PC2 + PC3 + x + y</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #E1F2DD; color: #000000;">52.67</td>
<td headers="R2" class="gt_row gt_right">0.37</td>
<td headers="Adjusted R2" class="gt_row gt_right">0.27</td>
<td headers="p (&gt;F)" class="gt_row gt_right">0.00</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">1.00</td>
<td headers="Prop. of total variance" class="gt_row gt_right">0.37</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">Pure enviro. model</td>
<td headers="pRDA model call" class="gt_row gt_left">gen ~ CA_rPCA2 + CA_rPCA3 + Condition(PC1 + PC2 + PC3 + x + y)</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #F5F7F5; color: #000000;">4.61</td>
<td headers="R2" class="gt_row gt_right">0.03</td>
<td headers="Adjusted R2" class="gt_row gt_right">0.00</td>
<td headers="p (&gt;F)" class="gt_row gt_right">0.06</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">0.09</td>
<td headers="Prop. of total variance" class="gt_row gt_right">0.03</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">Pure pop. structure model</td>
<td headers="pRDA model call" class="gt_row gt_left">gen ~ PC1 + PC2 + PC3 + Condition(CA_rPCA2 + CA_rPCA3 + x + y)</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #EDF5EB; color: #000000;">23.09</td>
<td headers="R2" class="gt_row gt_right">0.16</td>
<td headers="Adjusted R2" class="gt_row gt_right">0.13</td>
<td headers="p (&gt;F)" class="gt_row gt_right">0.00</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">0.44</td>
<td headers="Prop. of total variance" class="gt_row gt_right">0.16</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">Pure geography model</td>
<td headers="pRDA model call" class="gt_row gt_left">gen ~ x + y + Condition(PC1 + PC2 + PC3 + CA_rPCA2 + CA_rPCA3)</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #F5F7F4; color: #000000;">5.32</td>
<td headers="R2" class="gt_row gt_right">0.04</td>
<td headers="Adjusted R2" class="gt_row gt_right">0.01</td>
<td headers="p (&gt;F)" class="gt_row gt_right">0.00</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">0.10</td>
<td headers="Prop. of total variance" class="gt_row gt_right">0.04</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">Confounded variance</td>
<td headers="pRDA model call" class="gt_row gt_left"><br></td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #EFF5ED; color: #000000;">19.66</td>
<td headers="R2" class="gt_row gt_right"><br></td>
<td headers="Adjusted R2" class="gt_row gt_right"><br></td>
<td headers="p (&gt;F)" class="gt_row gt_right"><br></td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">0.37</td>
<td headers="Prop. of total variance" class="gt_row gt_right">0.14</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">Total unexplained variance</td>
<td headers="pRDA model call" class="gt_row gt_left"><br></td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #C1E3BB; color: #000000;">90.72</td>
<td headers="R2" class="gt_row gt_right"><br></td>
<td headers="Adjusted R2" class="gt_row gt_right"><br></td>
<td headers="p (&gt;F)" class="gt_row gt_right"><br></td>
<td headers="Prop. of explainable variance" class="gt_row gt_right"><br></td>
<td headers="Prop. of total variance" class="gt_row gt_right">0.63</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">Total inertia</td>
<td headers="pRDA model call" class="gt_row gt_left"><br></td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #7FBF7B; color: #000000;">143.39</td>
<td headers="R2" class="gt_row gt_right"><br></td>
<td headers="Adjusted R2" class="gt_row gt_right"><br></td>
<td headers="p (&gt;F)" class="gt_row gt_right"><br></td>
<td headers="Prop. of explainable variance" class="gt_row gt_right"><br></td>
<td headers="Prop. of total variance" class="gt_row gt_right">1.00</td>
</tr>
</tbody>
</table>
</div>
<p>From the above table, we can see that the full model only explains
27% of the total variance in our genetic data. Furthermore, our pure
environmental model explains only a small proportion of the variance
(“inertia”) contained within the full model (9%), and that population
structure explains substantially more (44%). There is also a large
amount of variance (37%) within the full model that is confounded,
meaning that variables within the three variable sets explain a
substantial amount of the full model’s variance <em>when in combination
with one another</em>. This is a relatively common occurrence (see
Capblancq &amp; Forester 2021), and users should always keep in mind how
collinearity may affect their results (and account for this, if
possible; see algatr’s environmental data vignette for more information
on how to do so). However, accounting for collinearity may also elevate
false positive rates, while failing to account for it at all can lead to
elevated false negative rates.</p>
<p>In our case, it is clear from our variance partitioning analysis that
much of the variation in our genetic data can explained by population
structure, and thus users may want to further explore more sophisticated
approaches for quantifying demographic history or population structure
before moving forward with outlier loci detection. As with most
analyses, careful consideration of the input data and the biology of the
system are critical to draw robust and reliable conclusions.</p>
</div>
<div class="section level3">
<h3 id="identifying-candidate-snps-with-rda_getoutliers">Identifying candidate SNPs with <code>rda_getoutliers()</code><a class="anchor" aria-label="anchor" href="#identifying-candidate-snps-with-rda_getoutliers"></a>
</h3>
<hr>
<p>Having run variance partitioning on our data, we can now move forward
with detecting outlier loci that are significantly associated with
environmental variables. To do so, we’ll first run a partial RDA on only
those significant environmental variables, and we’ll also want to add
population structure as a covariable (i.e., a conditioning variable).
We’ll only retain two PC axes for our measurement of population
structure.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod_pRDA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_run.html">rda_run</a></span><span class="op">(</span><span class="va">gen</span>, <span class="va">env</span>, model <span class="op">=</span> <span class="st">"best"</span>, correctPC <span class="op">=</span> <span class="cn">TRUE</span>, nPC <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="RDA_vignette_files/figure-html/pRDA%20final-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; Step: R2.adj= 0 </span></span>
<span><span class="co">#&gt; Call: gen ~ 1 </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;                         R2.adjusted</span></span>
<span><span class="co">#&gt; &lt;All variables&gt;        0.0202527037</span></span>
<span><span class="co">#&gt; + CA_rPCA2             0.0196214014</span></span>
<span><span class="co">#&gt; + CA_rPCA3             0.0137427906</span></span>
<span><span class="co">#&gt; + CA_rPCA1             0.0009953656</span></span>
<span><span class="co">#&gt; &lt;none&gt;                 0.0000000000</span></span>
<span><span class="co">#&gt; + Condition(PC1 + PC2) 0.0000000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            Df    AIC      F Pr(&gt;F)   </span></span>
<span><span class="co">#&gt; + CA_rPCA2  1 264.09 2.0407  0.006 **</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Step: R2.adj= 0.0196214 </span></span>
<span><span class="co">#&gt; Call: gen ~ CA_rPCA2 </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;                        R2.adjusted</span></span>
<span><span class="co">#&gt; + CA_rPCA3              0.03919420</span></span>
<span><span class="co">#&gt; &lt;All variables&gt;         0.02025270</span></span>
<span><span class="co">#&gt; + CA_rPCA1              0.01991617</span></span>
<span><span class="co">#&gt; &lt;none&gt;                  0.01962140</span></span>
<span><span class="co">#&gt; + Condition(PC1 + PC2)  0.01834925</span></span></code></pre>
<p>Given this partial RDA, only environmental PC2 is considered
significant.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod_pRDA</span><span class="op">$</span><span class="va">anova</span></span>
<span><span class="co">#&gt;                   R2.adj Df    AIC      F Pr(&gt;F)   </span></span>
<span><span class="co">#&gt; + CA_rPCA2      0.019621  1 264.09 2.0407  0.006 **</span></span>
<span><span class="co">#&gt; &lt;All variables&gt; 0.020253                           </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre></div>
<p>Now that we have the RDA model, we can discover SNPs that are
outliers. In algatr, this can be done using two different methods to
detect outlier loci, specified by the <code>outlier_method</code>
argument. The first uses Z-scores for outlier detection, and is from <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/mec.14584" class="external-link">Forester
et al. 2018</a>; their code can be found <a href="https://popgen.nescent.org/2018-03-27_RDA_GEA.html" class="external-link">here</a>. The
other method transforms RDA loadings into p-values, and was developed by
<a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906" class="external-link">Capblancq
et al. 2018</a> and is further described in <a href="https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722" class="external-link">Capblancq
&amp; Forester 2021</a>; code developed to run this is associated with
Capblancq and Forester 2021 and can be found <a href="https://github.com/Capblancq/RDA-landscape-genomics" class="external-link">here</a>. We
provide brief explanations for each of these methods below.</p>
<p><strong>N.B.: Some researchers may prefer to do MAF filtering and
linkage disequilibrium pruning during outlier detection (e.g., Capblancq
&amp; Forester 2021). In this case, only one SNP per contig is selected
based on the highest loading. The algatr pipeline does this filtering
initially when processing genetic data; see data processing vignette for
more information.</strong></p>
<div class="section level4">
<h4 id="identifying-outliers-using-the-z-scores-method">Identifying outliers using the Z-scores method<a class="anchor" aria-label="anchor" href="#identifying-outliers-using-the-z-scores-method"></a>
</h4>
<p>To understand how the Z-scores method works to detect outliers, let’s
first take a look at the loadings of SNPs on the first two RDA axes
using the <code><a href="../reference/rda_plot.html">rda_plot()</a></code> function. This will show us histograms
for each of the selected RDA axes. SNPs that are not associated with
environmental variables will fall in the center of these distributions,
with outliers falling at the edges. The Z-score method works by setting
a Z-score (number of standard deviations) threshold beyond which SNPs
will be considered outliers (see <a href="https://popgen.nescent.org/2018-03-27_RDA_GEA.html" class="external-link">here</a> for
further details).</p>
<p>If we set this value too stringently, we minimize our false positive
rate but risk detecting false negatives; if it’s set too high, we
increase our false positive rate. Researchers will want to adjust this
parameter based on their tolerance for detecting false positives, but
also the strength of selection under which outliers are subject to
(assuming that SNPs that are more significantly associated with
environmental variables are also under stronger selection).</p>
<p>Because only one environmental variable was associated with our
genetic data, we only have one resulting RDA axis.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rda_plot.html">rda_plot</a></span><span class="op">(</span><span class="va">mod_pRDA</span>, axes <span class="op">=</span> <span class="st">"all"</span>, binwidth <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p>Now, let’s discover outliers with the <code><a href="../reference/rda_getoutliers.html">rda_getoutliers()</a></code>
function. The method is specified using the <code>outlier_method</code>
argument, and the number of standard deviations to detect outliers is
specified by the <code>z</code> argument; it defaults to 3. The
resulting object is a dataframe containing the SNP name, the RDA axis
it’s associated with, and the loading along this axis. A screeplot for
the RDA axes is automatically generated (but can be switched off using
the <code>plot</code> argument).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rda_sig_z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_getoutliers.html">rda_getoutliers</a></span><span class="op">(</span><span class="va">mod_pRDA</span>, naxes <span class="op">=</span> <span class="st">"all"</span>, outlier_method <span class="op">=</span> <span class="st">"z"</span>, z <span class="op">=</span> <span class="fl">3</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># How many outlier SNPs were detected?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">rda_sig_z</span><span class="op">$</span><span class="va">rda_snps</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 19</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="identifying-outliers-using-the-p-value-method">Identifying outliers using the p-value method<a class="anchor" aria-label="anchor" href="#identifying-outliers-using-the-p-value-method"></a>
</h4>
<p>Another method to detect outliers is a p-value method wherein rather
than using the loadings along RDA axes to identify extreme outliers (as
with the Z-scores method above), the loadings are transformed into
p-values. To do so, Mahalanobis distances are calculated for each locus,
scaled based on a genomic inflation factor, from which p-values and
q-values can be computed. We then adjust our p-values (based on a false
discovery rate, for example; this adjustment is done using the
<code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust()</a></code> function) and subsequently identify outliers
based on a user-specified significance threshold. Rather than detecting
outliers based on a p-value significance threshold, one could also
detect outliers based on q-values (which are also provided as output in
algatr), as is done in <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906" class="external-link">Capblancq
et al. 2018</a> and <a href="https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722" class="external-link">Capblancq
&amp; Forester 2021</a>. To do so, we would select loci that fall below
a user-specified threshold based on the proportion of false positives we
allow. For example, if we allow 10% or fewer outliers to be false
positives, outliers could be identified based on loci that have q-values
=&lt; 0.1.</p>
<p>Now, let’s detect outliers using this method. To do so, along with
<code>outlier_method</code>, we need to specify <code>p_adj</code>,
which is the method by which p-values are adjusted, as well as
<code>sig</code>, which is our significance threshold. We will also tell
the function not to produce a screeplot.</p>
<p>One limitation of the p-value outlier method is that you cannot
calculate outliers given fewer than two RDA axes (as is the case with
our partial RDA results). So, let’s move forward with this outlier
detection method but using the results from our simple RDA model with
variable selection, <code>mod_best</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rda_sig_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_getoutliers.html">rda_getoutliers</a></span><span class="op">(</span><span class="va">mod_best</span>, naxes <span class="op">=</span> <span class="st">"all"</span>, outlier_method <span class="op">=</span> <span class="st">"p"</span>, p_adj <span class="op">=</span> <span class="st">"fdr"</span>, sig <span class="op">=</span> <span class="fl">0.01</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># How many outlier SNPs were detected?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">rda_sig_p</span><span class="op">$</span><span class="va">rda_snps</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 101</span></span></code></pre></div>
<p>As mentioned above, we could also identify outliers based on
q-values. To do so, we can extract the q-values from the resulting
object from <code><a href="../reference/rda_getoutliers.html">rda_getoutliers()</a></code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract SNP names; choices is number of axes</span></span>
<span><span class="va">snp_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/scores.html" class="external-link">scores</a></span><span class="op">(</span><span class="va">mod_best</span>, choices <span class="op">=</span> <span class="fl">2</span>, display <span class="op">=</span> <span class="st">"species"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Identify outliers that have q-values &lt; 0.1</span></span>
<span><span class="va">q_sig</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">rda_sig_p</span><span class="op">$</span><span class="va">rdadapt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>snp_names <span class="op">=</span> <span class="va">snp_names</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">q.values</span> <span class="op">&lt;=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># How many outlier SNPs were detected?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">q_sig</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 170</span></span></code></pre></div>
<p>Note the differences between the two methods: using the p-value
method, we found many more outlier SNPs than with the Z-scores method
(101 vs. 19). Even more outliers were found when only using q-values as
a significant threshold (170). We can also see that 19 SNPs were
detected using all three methods.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">intersect</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">q_sig</span><span class="op">$</span><span class="va">snp_names</span>,</span>
<span>  <span class="va">rda_sig_p</span><span class="op">$</span><span class="va">rda_snps</span>,</span>
<span>  <span class="va">rda_sig_z</span><span class="op">$</span><span class="va">rda_snps</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Locus_517"  "Locus_771"  "Locus_786"  "Locus_945"  "Locus_1247"</span></span>
<span><span class="co">#&gt;  [6] "Locus_1263" "Locus_1278" "Locus_1318" "Locus_1374" "Locus_1439"</span></span>
<span><span class="co">#&gt; [11] "Locus_1524" "Locus_1801" "Locus_1949" "Locus_2045" "Locus_2338"</span></span>
<span><span class="co">#&gt; [16] "Locus_2431" "Locus_2559" "Locus_2665" "Locus_2947"</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="visualizing-rda-results-with-rda_plot">Visualizing RDA results with <code>rda_plot()</code><a class="anchor" aria-label="anchor" href="#visualizing-rda-results-with-rda_plot"></a>
</h3>
<hr>
<p>algatr provides options for visualizing RDA results in two ways with
the <code><a href="../reference/rda_plot.html">rda_plot()</a></code> function: a biplot and a Manhattan plot.</p>
<div class="section level4">
<h4 id="rda-biplot">RDA biplot<a class="anchor" aria-label="anchor" href="#rda-biplot"></a>
</h4>
<p>Let’s build a biplot (or some call it a triplot, since it also shows
loadings of environmental variables) with RDA axes 1 and 2, specified
using the <code>rdaplot</code> argument. In this case, we also need to
specify the number of axes to display using the <code>biplot_axes</code>
argument. The default of <code><a href="../reference/rda_plot.html">rda_plot()</a></code> is to plot all
variables, regardless of whether they were considered significant or
not.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rda_plot.html">rda_plot</a></span><span class="op">(</span><span class="va">mod_best</span>, <span class="va">rda_sig_p</span><span class="op">$</span><span class="va">rda_snps</span>, biplot_axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, rdaplot <span class="op">=</span> <span class="cn">TRUE</span>, manhattan <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># In the case of our partial RDA, there was only one RDA axis, so a histogram is generated</span></span>
<span><span class="fu"><a href="../reference/rda_plot.html">rda_plot</a></span><span class="op">(</span><span class="va">mod_pRDA</span>, <span class="va">rda_sig_z</span><span class="op">$</span><span class="va">rda_snps</span>, rdaplot <span class="op">=</span> <span class="cn">TRUE</span>, manhattan <span class="op">=</span> <span class="cn">FALSE</span>, binwidth <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="manhattan-plot">Manhattan plot<a class="anchor" aria-label="anchor" href="#manhattan-plot"></a>
</h4>
<p>Let’s now build a Manhattan plot with a significance level of 0.05
using the <code>manhattan</code> argument; if this is set to TRUE, we
must also provide a significance threshold value using the
<code>sig</code> argument (the default is 0.05). Be aware that outliers
were detected with a different p-value threshold which is why there are
grey points above our threshold line (i.e., data points are colorized
according to the threshold of the model, not the visualization).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rda_plot.html">rda_plot</a></span><span class="op">(</span><span class="va">mod_best</span>, <span class="va">rda_sig_p</span><span class="op">$</span><span class="va">rda_snps</span>, <span class="va">rda_sig_p</span><span class="op">$</span><span class="va">pvalues</span>, rdaplot <span class="op">=</span> <span class="cn">FALSE</span>, manhattan <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="RDA_vignette_files/figure-html/Manhattan-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="interpreting-rda-results">Interpreting RDA results<a class="anchor" aria-label="anchor" href="#interpreting-rda-results"></a>
</h3>
<hr>
<p>As discussed in <a href="https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722" class="external-link">Capblancq
&amp; Forester 2021</a>, users should be aware that
<strong><em>“statistical selection of variables optimizes variance
explained but cannot, by itself, identify the ecological or mechanistic
drivers of genetic variation.”</em></strong></p>
<div class="section level4">
<h4 id="identifying-environmental-associations-with-rda_cor-and-rda_table">Identifying environmental associations with <code>rda_cor()</code>
and <code>rda_table()</code><a class="anchor" aria-label="anchor" href="#identifying-environmental-associations-with-rda_cor-and-rda_table"></a>
</h4>
<p>Now that we’ve identified outlier loci, it is useful to be able to
identify how much they correlate to each environmental variable. To do
so, we can run a correlation test of each SNP using
<code><a href="../reference/rda_cor.html">rda_cor()</a></code> and summarize results in a customizable table
using <code><a href="../reference/rda_table.html">rda_table()</a></code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract genotypes for outlier SNPs</span></span>
<span><span class="va">rda_snps</span> <span class="op">&lt;-</span> <span class="va">rda_sig_p</span><span class="op">$</span><span class="va">rda_snps</span></span>
<span><span class="va">rda_gen</span> <span class="op">&lt;-</span> <span class="va">gen</span><span class="op">[</span>, <span class="va">rda_snps</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Run correlation test</span></span>
<span><span class="va">cor_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_cor.html">rda_cor</a></span><span class="op">(</span><span class="va">rda_gen</span>, <span class="va">env</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make a table from these results (displaying only the first 5 rows):</span></span>
<span><span class="fu"><a href="../reference/rda_table.html">rda_table</a></span><span class="op">(</span><span class="va">cor_df</span>, nrow <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div id="jvnqmnrhrw" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#jvnqmnrhrw table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#jvnqmnrhrw thead, #jvnqmnrhrw tbody, #jvnqmnrhrw tfoot, #jvnqmnrhrw tr, #jvnqmnrhrw td, #jvnqmnrhrw th {
  border-style: none;
}

#jvnqmnrhrw p {
  margin: 0;
  padding: 0;
}

#jvnqmnrhrw .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#jvnqmnrhrw .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#jvnqmnrhrw .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#jvnqmnrhrw .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#jvnqmnrhrw .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#jvnqmnrhrw .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#jvnqmnrhrw .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#jvnqmnrhrw .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#jvnqmnrhrw .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#jvnqmnrhrw .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#jvnqmnrhrw .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#jvnqmnrhrw .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#jvnqmnrhrw .gt_spanner_row {
  border-bottom-style: hidden;
}

#jvnqmnrhrw .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#jvnqmnrhrw .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#jvnqmnrhrw .gt_from_md > :first-child {
  margin-top: 0;
}

#jvnqmnrhrw .gt_from_md > :last-child {
  margin-bottom: 0;
}

#jvnqmnrhrw .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#jvnqmnrhrw .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#jvnqmnrhrw .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#jvnqmnrhrw .gt_row_group_first td {
  border-top-width: 2px;
}

#jvnqmnrhrw .gt_row_group_first th {
  border-top-width: 2px;
}

#jvnqmnrhrw .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#jvnqmnrhrw .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#jvnqmnrhrw .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#jvnqmnrhrw .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#jvnqmnrhrw .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#jvnqmnrhrw .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#jvnqmnrhrw .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#jvnqmnrhrw .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#jvnqmnrhrw .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#jvnqmnrhrw .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#jvnqmnrhrw .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#jvnqmnrhrw .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#jvnqmnrhrw .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#jvnqmnrhrw .gt_left {
  text-align: left;
}

#jvnqmnrhrw .gt_center {
  text-align: center;
}

#jvnqmnrhrw .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#jvnqmnrhrw .gt_font_normal {
  font-weight: normal;
}

#jvnqmnrhrw .gt_font_bold {
  font-weight: bold;
}

#jvnqmnrhrw .gt_font_italic {
  font-style: italic;
}

#jvnqmnrhrw .gt_super {
  font-size: 65%;
}

#jvnqmnrhrw .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#jvnqmnrhrw .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#jvnqmnrhrw .gt_indent_1 {
  text-indent: 5px;
}

#jvnqmnrhrw .gt_indent_2 {
  text-indent: 10px;
}

#jvnqmnrhrw .gt_indent_3 {
  text-indent: 15px;
}

#jvnqmnrhrw .gt_indent_4 {
  text-indent: 20px;
}

#jvnqmnrhrw .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="r">r</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p">p</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="snp">snp</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="var">var</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #B7DDB1; color: #000000;">0.27</td>
<td headers="p" class="gt_row gt_right">0.02</td>
<td headers="snp" class="gt_row gt_left">Locus_125</td>
<td headers="var" class="gt_row gt_left">CA_rPCA3</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #7FBF7B; color: #000000;">0.39</td>
<td headers="p" class="gt_row gt_right">0.00</td>
<td headers="snp" class="gt_row gt_left">Locus_166</td>
<td headers="var" class="gt_row gt_left">CA_rPCA3</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #AED8A7; color: #000000;">0.29</td>
<td headers="p" class="gt_row gt_right">0.01</td>
<td headers="snp" class="gt_row gt_left">Locus_249</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #DAC3DF; color: #000000;">-0.24</td>
<td headers="p" class="gt_row gt_right">0.03</td>
<td headers="snp" class="gt_row gt_left">Locus_263</td>
<td headers="var" class="gt_row gt_left">CA_rPCA1</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #B7DDB1; color: #000000;">0.27</td>
<td headers="p" class="gt_row gt_right">0.02</td>
<td headers="snp" class="gt_row gt_left">Locus_263</td>
<td headers="var" class="gt_row gt_left">CA_rPCA3</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Order by the strength of the correlation</span></span>
<span><span class="fu"><a href="../reference/rda_table.html">rda_table</a></span><span class="op">(</span><span class="va">cor_df</span>, order <span class="op">=</span> <span class="cn">TRUE</span>, nrow <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div id="ptvnbjinoh" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#ptvnbjinoh table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#ptvnbjinoh thead, #ptvnbjinoh tbody, #ptvnbjinoh tfoot, #ptvnbjinoh tr, #ptvnbjinoh td, #ptvnbjinoh th {
  border-style: none;
}

#ptvnbjinoh p {
  margin: 0;
  padding: 0;
}

#ptvnbjinoh .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#ptvnbjinoh .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#ptvnbjinoh .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#ptvnbjinoh .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#ptvnbjinoh .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ptvnbjinoh .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ptvnbjinoh .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ptvnbjinoh .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#ptvnbjinoh .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#ptvnbjinoh .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#ptvnbjinoh .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#ptvnbjinoh .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#ptvnbjinoh .gt_spanner_row {
  border-bottom-style: hidden;
}

#ptvnbjinoh .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#ptvnbjinoh .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#ptvnbjinoh .gt_from_md > :first-child {
  margin-top: 0;
}

#ptvnbjinoh .gt_from_md > :last-child {
  margin-bottom: 0;
}

#ptvnbjinoh .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#ptvnbjinoh .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#ptvnbjinoh .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#ptvnbjinoh .gt_row_group_first td {
  border-top-width: 2px;
}

#ptvnbjinoh .gt_row_group_first th {
  border-top-width: 2px;
}

#ptvnbjinoh .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ptvnbjinoh .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#ptvnbjinoh .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#ptvnbjinoh .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ptvnbjinoh .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ptvnbjinoh .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#ptvnbjinoh .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#ptvnbjinoh .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#ptvnbjinoh .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ptvnbjinoh .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ptvnbjinoh .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ptvnbjinoh .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ptvnbjinoh .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ptvnbjinoh .gt_left {
  text-align: left;
}

#ptvnbjinoh .gt_center {
  text-align: center;
}

#ptvnbjinoh .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#ptvnbjinoh .gt_font_normal {
  font-weight: normal;
}

#ptvnbjinoh .gt_font_bold {
  font-weight: bold;
}

#ptvnbjinoh .gt_font_italic {
  font-style: italic;
}

#ptvnbjinoh .gt_super {
  font-size: 65%;
}

#ptvnbjinoh .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#ptvnbjinoh .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#ptvnbjinoh .gt_indent_1 {
  text-indent: 5px;
}

#ptvnbjinoh .gt_indent_2 {
  text-indent: 10px;
}

#ptvnbjinoh .gt_indent_3 {
  text-indent: 15px;
}

#ptvnbjinoh .gt_indent_4 {
  text-indent: 20px;
}

#ptvnbjinoh .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="r">r</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p">p</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="snp">snp</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="var">var</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #AF8DC3; color: #FFFFFF;">-0.49</td>
<td headers="p" class="gt_row gt_right">0</td>
<td headers="snp" class="gt_row gt_left">Locus_2947</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BD9ECC; color: #000000;">-0.43</td>
<td headers="p" class="gt_row gt_right">0</td>
<td headers="snp" class="gt_row gt_left">Locus_1278</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BFA1CE; color: #000000;">-0.42</td>
<td headers="p" class="gt_row gt_right">0</td>
<td headers="snp" class="gt_row gt_left">Locus_2338</td>
<td headers="var" class="gt_row gt_left">CA_rPCA3</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BFA1CE; color: #000000;">-0.42</td>
<td headers="p" class="gt_row gt_right">0</td>
<td headers="snp" class="gt_row gt_left">Locus_786</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BFA1CE; color: #000000;">-0.42</td>
<td headers="p" class="gt_row gt_right">0</td>
<td headers="snp" class="gt_row gt_left">Locus_1318</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Only retain the top variable for each SNP based on the strength of the correlation</span></span>
<span><span class="fu"><a href="../reference/rda_table.html">rda_table</a></span><span class="op">(</span><span class="va">cor_df</span>, top <span class="op">=</span> <span class="cn">TRUE</span>, nrow <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div id="ooqzbqlnpb" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#ooqzbqlnpb table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#ooqzbqlnpb thead, #ooqzbqlnpb tbody, #ooqzbqlnpb tfoot, #ooqzbqlnpb tr, #ooqzbqlnpb td, #ooqzbqlnpb th {
  border-style: none;
}

#ooqzbqlnpb p {
  margin: 0;
  padding: 0;
}

#ooqzbqlnpb .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#ooqzbqlnpb .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#ooqzbqlnpb .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#ooqzbqlnpb .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#ooqzbqlnpb .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ooqzbqlnpb .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ooqzbqlnpb .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ooqzbqlnpb .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#ooqzbqlnpb .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#ooqzbqlnpb .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#ooqzbqlnpb .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#ooqzbqlnpb .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#ooqzbqlnpb .gt_spanner_row {
  border-bottom-style: hidden;
}

#ooqzbqlnpb .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#ooqzbqlnpb .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#ooqzbqlnpb .gt_from_md > :first-child {
  margin-top: 0;
}

#ooqzbqlnpb .gt_from_md > :last-child {
  margin-bottom: 0;
}

#ooqzbqlnpb .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#ooqzbqlnpb .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#ooqzbqlnpb .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#ooqzbqlnpb .gt_row_group_first td {
  border-top-width: 2px;
}

#ooqzbqlnpb .gt_row_group_first th {
  border-top-width: 2px;
}

#ooqzbqlnpb .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ooqzbqlnpb .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#ooqzbqlnpb .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#ooqzbqlnpb .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ooqzbqlnpb .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ooqzbqlnpb .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#ooqzbqlnpb .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#ooqzbqlnpb .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#ooqzbqlnpb .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ooqzbqlnpb .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ooqzbqlnpb .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ooqzbqlnpb .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ooqzbqlnpb .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ooqzbqlnpb .gt_left {
  text-align: left;
}

#ooqzbqlnpb .gt_center {
  text-align: center;
}

#ooqzbqlnpb .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#ooqzbqlnpb .gt_font_normal {
  font-weight: normal;
}

#ooqzbqlnpb .gt_font_bold {
  font-weight: bold;
}

#ooqzbqlnpb .gt_font_italic {
  font-style: italic;
}

#ooqzbqlnpb .gt_super {
  font-size: 65%;
}

#ooqzbqlnpb .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#ooqzbqlnpb .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#ooqzbqlnpb .gt_indent_1 {
  text-indent: 5px;
}

#ooqzbqlnpb .gt_indent_2 {
  text-indent: 10px;
}

#ooqzbqlnpb .gt_indent_3 {
  text-indent: 15px;
}

#ooqzbqlnpb .gt_indent_4 {
  text-indent: 20px;
}

#ooqzbqlnpb .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="r">r</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p">p</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="snp">snp</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="var">var</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #B7DDB1; color: #000000;">0.27</td>
<td headers="p" class="gt_row gt_right">0.02</td>
<td headers="snp" class="gt_row gt_left">Locus_125</td>
<td headers="var" class="gt_row gt_left">CA_rPCA3</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #7FBF7B; color: #000000;">0.39</td>
<td headers="p" class="gt_row gt_right">0.00</td>
<td headers="snp" class="gt_row gt_left">Locus_166</td>
<td headers="var" class="gt_row gt_left">CA_rPCA3</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #AED8A7; color: #000000;">0.29</td>
<td headers="p" class="gt_row gt_right">0.01</td>
<td headers="snp" class="gt_row gt_left">Locus_249</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #B7DDB1; color: #000000;">0.27</td>
<td headers="p" class="gt_row gt_right">0.02</td>
<td headers="snp" class="gt_row gt_left">Locus_263</td>
<td headers="var" class="gt_row gt_left">CA_rPCA3</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #B2DBAC; color: #000000;">0.28</td>
<td headers="p" class="gt_row gt_right">0.01</td>
<td headers="snp" class="gt_row gt_left">Locus_312</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Display results for only one environmental variable</span></span>
<span><span class="fu"><a href="../reference/rda_table.html">rda_table</a></span><span class="op">(</span><span class="va">cor_df</span>, var <span class="op">=</span> <span class="st">"CA_rPCA2"</span>, nrow <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div id="ovjkcctnqb" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#ovjkcctnqb table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#ovjkcctnqb thead, #ovjkcctnqb tbody, #ovjkcctnqb tfoot, #ovjkcctnqb tr, #ovjkcctnqb td, #ovjkcctnqb th {
  border-style: none;
}

#ovjkcctnqb p {
  margin: 0;
  padding: 0;
}

#ovjkcctnqb .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#ovjkcctnqb .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#ovjkcctnqb .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#ovjkcctnqb .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#ovjkcctnqb .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ovjkcctnqb .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ovjkcctnqb .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ovjkcctnqb .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#ovjkcctnqb .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#ovjkcctnqb .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#ovjkcctnqb .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#ovjkcctnqb .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#ovjkcctnqb .gt_spanner_row {
  border-bottom-style: hidden;
}

#ovjkcctnqb .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#ovjkcctnqb .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#ovjkcctnqb .gt_from_md > :first-child {
  margin-top: 0;
}

#ovjkcctnqb .gt_from_md > :last-child {
  margin-bottom: 0;
}

#ovjkcctnqb .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#ovjkcctnqb .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#ovjkcctnqb .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#ovjkcctnqb .gt_row_group_first td {
  border-top-width: 2px;
}

#ovjkcctnqb .gt_row_group_first th {
  border-top-width: 2px;
}

#ovjkcctnqb .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ovjkcctnqb .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#ovjkcctnqb .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#ovjkcctnqb .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ovjkcctnqb .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ovjkcctnqb .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#ovjkcctnqb .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#ovjkcctnqb .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#ovjkcctnqb .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ovjkcctnqb .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ovjkcctnqb .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ovjkcctnqb .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ovjkcctnqb .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ovjkcctnqb .gt_left {
  text-align: left;
}

#ovjkcctnqb .gt_center {
  text-align: center;
}

#ovjkcctnqb .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#ovjkcctnqb .gt_font_normal {
  font-weight: normal;
}

#ovjkcctnqb .gt_font_bold {
  font-weight: bold;
}

#ovjkcctnqb .gt_font_italic {
  font-style: italic;
}

#ovjkcctnqb .gt_super {
  font-size: 65%;
}

#ovjkcctnqb .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#ovjkcctnqb .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#ovjkcctnqb .gt_indent_1 {
  text-indent: 5px;
}

#ovjkcctnqb .gt_indent_2 {
  text-indent: 10px;
}

#ovjkcctnqb .gt_indent_3 {
  text-indent: 15px;
}

#ovjkcctnqb .gt_indent_4 {
  text-indent: 20px;
}

#ovjkcctnqb .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="r">r</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p">p</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="snp">snp</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="var">var</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #9ACD94; color: #000000;">0.29</td>
<td headers="p" class="gt_row gt_right">0.01</td>
<td headers="snp" class="gt_row gt_left">Locus_249</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #9FD099; color: #000000;">0.28</td>
<td headers="p" class="gt_row gt_right">0.01</td>
<td headers="snp" class="gt_row gt_left">Locus_312</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #A4D39F; color: #000000;">0.27</td>
<td headers="p" class="gt_row gt_right">0.02</td>
<td headers="snp" class="gt_row gt_left">Locus_404</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #7FBF7B; color: #000000;">0.34</td>
<td headers="p" class="gt_row gt_right">0.00</td>
<td headers="snp" class="gt_row gt_left">Locus_517</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BADFB3; color: #000000;">0.23</td>
<td headers="p" class="gt_row gt_right">0.04</td>
<td headers="snp" class="gt_row gt_left">Locus_612</td>
<td headers="var" class="gt_row gt_left">CA_rPCA2</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="section level3">
<h3 id="running-rda-with-rda_do_everything">Running RDA with <code>rda_do_everything()</code><a class="anchor" aria-label="anchor" href="#running-rda-with-rda_do_everything"></a>
</h3>
<hr>
<p>The algatr package also has an option to run all of the above
functionality in a single function, <code><a href="../reference/rda_do_everything.html">rda_do_everything()</a></code>.
This function is similar in structure to <code><a href="../reference/rda_run.html">rda_run()</a></code>, but
also performs imputation (<code>impute</code> option), can generate
figures, the correlation test table (<code>cortest</code> option;
default is TRUE), and variance partitioning (<code>varpart</code>
option; default is FALSE). <strong>Please be aware that the
<code>do_everything()</code> functions are meant to be exploratory. We
do not recommend their use for final analyses unless certain they are
properly parameterized.</strong></p>
<p>The resulting object from <code><a href="../reference/rda_do_everything.html">rda_do_everything()</a></code>
contains:</p>
<ul>
<li><p>Candidate SNPs: <code>rda_snps</code></p></li>
<li><p>Data frame of environmental associations with outlier SNPs:
<code>cor_df</code></p></li>
<li><p>RDA model specifics: <code>rda_mod</code></p></li>
<li><p>Outlier test results: <code>rda_outlier_test</code></p></li>
<li><p>Relevant R-squared values: <code>rsq</code></p></li>
<li><p>ANOVA results: <code>anova</code></p></li>
<li><p>List of all p-values: <code>pvalues</code></p></li>
<li><p>Variance partitioning results: <code>varpart</code></p></li>
</ul>
<p>Let’s first run a simple RDA, imputing to the median, with variable
selection and the p-value outlier method:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_do_everything.html">rda_do_everything</a></span><span class="op">(</span><span class="va">liz_vcf</span>, <span class="va">CA_env</span>, <span class="va">liz_coords</span>,</span>
<span>                             impute <span class="op">=</span> <span class="st">"simple"</span>,</span>
<span>                             correctGEO <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                             correctPC <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                             outlier_method <span class="op">=</span> <span class="st">"p"</span>,</span>
<span>                             sig <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                             p_adj <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>                             cortest <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                             varpart <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<img src="RDA_vignette_files/figure-html/simple%20RDA%20do%20everything-1.png" width="700" style="display: block; margin: auto;"><div id="pcjcopfmrn" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#pcjcopfmrn table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#pcjcopfmrn thead, #pcjcopfmrn tbody, #pcjcopfmrn tfoot, #pcjcopfmrn tr, #pcjcopfmrn td, #pcjcopfmrn th {
  border-style: none;
}

#pcjcopfmrn p {
  margin: 0;
  padding: 0;
}

#pcjcopfmrn .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#pcjcopfmrn .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#pcjcopfmrn .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#pcjcopfmrn .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#pcjcopfmrn .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#pcjcopfmrn .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#pcjcopfmrn .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#pcjcopfmrn .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#pcjcopfmrn .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#pcjcopfmrn .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#pcjcopfmrn .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#pcjcopfmrn .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#pcjcopfmrn .gt_spanner_row {
  border-bottom-style: hidden;
}

#pcjcopfmrn .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#pcjcopfmrn .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#pcjcopfmrn .gt_from_md > :first-child {
  margin-top: 0;
}

#pcjcopfmrn .gt_from_md > :last-child {
  margin-bottom: 0;
}

#pcjcopfmrn .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#pcjcopfmrn .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#pcjcopfmrn .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#pcjcopfmrn .gt_row_group_first td {
  border-top-width: 2px;
}

#pcjcopfmrn .gt_row_group_first th {
  border-top-width: 2px;
}

#pcjcopfmrn .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#pcjcopfmrn .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#pcjcopfmrn .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#pcjcopfmrn .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#pcjcopfmrn .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#pcjcopfmrn .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#pcjcopfmrn .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#pcjcopfmrn .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#pcjcopfmrn .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#pcjcopfmrn .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#pcjcopfmrn .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#pcjcopfmrn .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#pcjcopfmrn .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#pcjcopfmrn .gt_left {
  text-align: left;
}

#pcjcopfmrn .gt_center {
  text-align: center;
}

#pcjcopfmrn .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#pcjcopfmrn .gt_font_normal {
  font-weight: normal;
}

#pcjcopfmrn .gt_font_bold {
  font-weight: bold;
}

#pcjcopfmrn .gt_font_italic {
  font-style: italic;
}

#pcjcopfmrn .gt_super {
  font-size: 65%;
}

#pcjcopfmrn .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#pcjcopfmrn .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#pcjcopfmrn .gt_indent_1 {
  text-indent: 5px;
}

#pcjcopfmrn .gt_indent_2 {
  text-indent: 10px;
}

#pcjcopfmrn .gt_indent_3 {
  text-indent: 15px;
}

#pcjcopfmrn .gt_indent_4 {
  text-indent: 20px;
}

#pcjcopfmrn .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="r">
r
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p">
p
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="snp">
snp
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="var">
var
</th>
</tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #AF8DC3; color: #FFFFFF;">
-0.49
</td>
<td headers="p" class="gt_row gt_right">
0
</td>
<td headers="snp" class="gt_row gt_left">
Locus_2947
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA2
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BD9ECC; color: #000000;">
-0.43
</td>
<td headers="p" class="gt_row gt_right">
0
</td>
<td headers="snp" class="gt_row gt_left">
Locus_1278
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA2
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BFA1CE; color: #000000;">
-0.42
</td>
<td headers="p" class="gt_row gt_right">
0
</td>
<td headers="snp" class="gt_row gt_left">
Locus_2338
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA3
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BFA1CE; color: #000000;">
-0.42
</td>
<td headers="p" class="gt_row gt_right">
0
</td>
<td headers="snp" class="gt_row gt_left">
Locus_786
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA2
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BFA1CE; color: #000000;">
-0.42
</td>
<td headers="p" class="gt_row gt_right">
0
</td>
<td headers="snp" class="gt_row gt_left">
Locus_1318
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA2
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BFA1CE; color: #000000;">
-0.42
</td>
<td headers="p" class="gt_row gt_right">
0
</td>
<td headers="snp" class="gt_row gt_left">
Locus_771
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA2
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #C1A4CF; color: #000000;">
-0.41
</td>
<td headers="p" class="gt_row gt_right">
0
</td>
<td headers="snp" class="gt_row gt_left">
Locus_1374
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA2
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #A0D19B; color: #000000;">
0.40
</td>
<td headers="p" class="gt_row gt_right">
0
</td>
<td headers="snp" class="gt_row gt_left">
Locus_636
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA3
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #A4D39E; color: #000000;">
0.39
</td>
<td headers="p" class="gt_row gt_right">
0
</td>
<td headers="snp" class="gt_row gt_left">
Locus_1263
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA2
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #A4D39E; color: #000000;">
0.39
</td>
<td headers="p" class="gt_row gt_right">
0
</td>
<td headers="snp" class="gt_row gt_left">
Locus_166
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA3
</td>
</tr>
</tbody>
</table>
</div>
<p>Now, let’s run a partial RDA, imputing based on 3 sNMF clusters,
correcting for both geography and structure, and instead using the
Z-score method to detect outlier loci, including variance partitioning.
We’ll keep the number of PCs for population structure set to the
default, which is three axes (<code>nPC</code>):</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rda_do_everything.html">rda_do_everything</a></span><span class="op">(</span><span class="va">liz_vcf</span>, <span class="va">CA_env</span>, <span class="va">liz_coords</span>,</span>
<span>                             impute <span class="op">=</span> <span class="st">"structure"</span>,</span>
<span>                             K_impute <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                             model <span class="op">=</span> <span class="st">"full"</span>,</span>
<span>                             correctGEO <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                             correctPC <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                             nPC <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                             varpart <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                             outlier_method <span class="op">=</span> <span class="st">"z"</span>,</span>
<span>                             z <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<img src="RDA_vignette_files/figure-html/pRDA%20do%20everything-1.png" width="700" style="display: block; margin: auto;"><div id="sjwbwjpopq" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#sjwbwjpopq table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#sjwbwjpopq thead, #sjwbwjpopq tbody, #sjwbwjpopq tfoot, #sjwbwjpopq tr, #sjwbwjpopq td, #sjwbwjpopq th {
  border-style: none;
}

#sjwbwjpopq p {
  margin: 0;
  padding: 0;
}

#sjwbwjpopq .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#sjwbwjpopq .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#sjwbwjpopq .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#sjwbwjpopq .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#sjwbwjpopq .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#sjwbwjpopq .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#sjwbwjpopq .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#sjwbwjpopq .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#sjwbwjpopq .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#sjwbwjpopq .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#sjwbwjpopq .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#sjwbwjpopq .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#sjwbwjpopq .gt_spanner_row {
  border-bottom-style: hidden;
}

#sjwbwjpopq .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#sjwbwjpopq .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#sjwbwjpopq .gt_from_md > :first-child {
  margin-top: 0;
}

#sjwbwjpopq .gt_from_md > :last-child {
  margin-bottom: 0;
}

#sjwbwjpopq .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#sjwbwjpopq .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#sjwbwjpopq .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#sjwbwjpopq .gt_row_group_first td {
  border-top-width: 2px;
}

#sjwbwjpopq .gt_row_group_first th {
  border-top-width: 2px;
}

#sjwbwjpopq .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#sjwbwjpopq .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#sjwbwjpopq .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#sjwbwjpopq .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#sjwbwjpopq .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#sjwbwjpopq .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#sjwbwjpopq .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#sjwbwjpopq .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#sjwbwjpopq .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#sjwbwjpopq .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#sjwbwjpopq .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#sjwbwjpopq .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#sjwbwjpopq .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#sjwbwjpopq .gt_left {
  text-align: left;
}

#sjwbwjpopq .gt_center {
  text-align: center;
}

#sjwbwjpopq .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#sjwbwjpopq .gt_font_normal {
  font-weight: normal;
}

#sjwbwjpopq .gt_font_bold {
  font-weight: bold;
}

#sjwbwjpopq .gt_font_italic {
  font-style: italic;
}

#sjwbwjpopq .gt_super {
  font-size: 65%;
}

#sjwbwjpopq .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#sjwbwjpopq .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#sjwbwjpopq .gt_indent_1 {
  text-indent: 5px;
}

#sjwbwjpopq .gt_indent_2 {
  text-indent: 10px;
}

#sjwbwjpopq .gt_indent_3 {
  text-indent: 15px;
}

#sjwbwjpopq .gt_indent_4 {
  text-indent: 20px;
}

#sjwbwjpopq .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="gt_heading">
<td colspan="7" class="gt_heading gt_title gt_font_normal gt_bottom_border" style>
Variance partitioning
</td>
</tr>
<tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="Model">
Model
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="Inertia">
Inertia
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="R2">
R2
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="Adjusted R2">
Adjusted R2
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p (&amp;gt;F)">
p (&gt;F)
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="Prop. of explainable variance">
Prop. of explainable variance
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="Prop. of total variance">
Prop. of total variance
</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr>
<td headers="Model" class="gt_row gt_left">
Full model
</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #DDF1D8; color: #000000;">
67.82
</td>
<td headers="R2" class="gt_row gt_right">
0.43
</td>
<td headers="Adjusted R2" class="gt_row gt_right">
0.34
</td>
<td headers="p (&gt;F)" class="gt_row gt_right">
0.00
</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">
1.00
</td>
<td headers="Prop. of total variance" class="gt_row gt_right">
0.43
</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">
Pure enviro. model
</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #F5F7F5; color: #000000;">
4.82
</td>
<td headers="R2" class="gt_row gt_right">
0.03
</td>
<td headers="Adjusted R2" class="gt_row gt_right">
0.01
</td>
<td headers="p (&gt;F)" class="gt_row gt_right">
0.03
</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">
0.07
</td>
<td headers="Prop. of total variance" class="gt_row gt_right">
0.03
</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">
Pure pop. structure model
</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #ECF4EA; color: #000000;">
29.09
</td>
<td headers="R2" class="gt_row gt_right">
0.18
</td>
<td headers="Adjusted R2" class="gt_row gt_right">
0.16
</td>
<td headers="p (&gt;F)" class="gt_row gt_right">
0.00
</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">
0.43
</td>
<td headers="Prop. of total variance" class="gt_row gt_right">
0.18
</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">
Pure geography model
</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #F5F7F4; color: #000000;">
5.79
</td>
<td headers="R2" class="gt_row gt_right">
0.04
</td>
<td headers="Adjusted R2" class="gt_row gt_right">
0.01
</td>
<td headers="p (&gt;F)" class="gt_row gt_right">
0.00
</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">
0.09
</td>
<td headers="Prop. of total variance" class="gt_row gt_right">
0.04
</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">
Confounded variance
</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #ECF5EA; color: #000000;">
28.12
</td>
<td headers="R2" class="gt_row gt_right">
<br>
</td>
<td headers="Adjusted R2" class="gt_row gt_right">
<br>
</td>
<td headers="p (&gt;F)" class="gt_row gt_right">
<br>
</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">
0.41
</td>
<td headers="Prop. of total variance" class="gt_row gt_right">
0.18
</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">
Total unexplained variance
</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #CCE9C6; color: #000000;">
90.23
</td>
<td headers="R2" class="gt_row gt_right">
<br>
</td>
<td headers="Adjusted R2" class="gt_row gt_right">
<br>
</td>
<td headers="p (&gt;F)" class="gt_row gt_right">
<br>
</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">
<br>
</td>
<td headers="Prop. of total variance" class="gt_row gt_right">
0.57
</td>
</tr>
<tr>
<td headers="Model" class="gt_row gt_left">
Total inertia
</td>
<td headers="Inertia" class="gt_row gt_right" style="background-color: #7FBF7B; color: #000000;">
158.05
</td>
<td headers="R2" class="gt_row gt_right">
<br>
</td>
<td headers="Adjusted R2" class="gt_row gt_right">
<br>
</td>
<td headers="p (&gt;F)" class="gt_row gt_right">
<br>
</td>
<td headers="Prop. of explainable variance" class="gt_row gt_right">
<br>
</td>
<td headers="Prop. of total variance" class="gt_row gt_right">
1.00
</td>
</tr>
</tbody>
</table>
</div>
<img src="RDA_vignette_files/figure-html/pRDA%20do%20everything-2.png" width="700" style="display: block; margin: auto;"><div id="qyzhkxujot" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#qyzhkxujot table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#qyzhkxujot thead, #qyzhkxujot tbody, #qyzhkxujot tfoot, #qyzhkxujot tr, #qyzhkxujot td, #qyzhkxujot th {
  border-style: none;
}

#qyzhkxujot p {
  margin: 0;
  padding: 0;
}

#qyzhkxujot .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#qyzhkxujot .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#qyzhkxujot .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#qyzhkxujot .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#qyzhkxujot .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#qyzhkxujot .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#qyzhkxujot .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#qyzhkxujot .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#qyzhkxujot .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#qyzhkxujot .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#qyzhkxujot .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#qyzhkxujot .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#qyzhkxujot .gt_spanner_row {
  border-bottom-style: hidden;
}

#qyzhkxujot .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#qyzhkxujot .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#qyzhkxujot .gt_from_md > :first-child {
  margin-top: 0;
}

#qyzhkxujot .gt_from_md > :last-child {
  margin-bottom: 0;
}

#qyzhkxujot .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#qyzhkxujot .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#qyzhkxujot .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#qyzhkxujot .gt_row_group_first td {
  border-top-width: 2px;
}

#qyzhkxujot .gt_row_group_first th {
  border-top-width: 2px;
}

#qyzhkxujot .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#qyzhkxujot .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#qyzhkxujot .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#qyzhkxujot .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#qyzhkxujot .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#qyzhkxujot .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#qyzhkxujot .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#qyzhkxujot .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#qyzhkxujot .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#qyzhkxujot .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#qyzhkxujot .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#qyzhkxujot .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#qyzhkxujot .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#qyzhkxujot .gt_left {
  text-align: left;
}

#qyzhkxujot .gt_center {
  text-align: center;
}

#qyzhkxujot .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#qyzhkxujot .gt_font_normal {
  font-weight: normal;
}

#qyzhkxujot .gt_font_bold {
  font-weight: bold;
}

#qyzhkxujot .gt_font_italic {
  font-style: italic;
}

#qyzhkxujot .gt_super {
  font-size: 65%;
}

#qyzhkxujot .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#qyzhkxujot .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#qyzhkxujot .gt_indent_1 {
  text-indent: 5px;
}

#qyzhkxujot .gt_indent_2 {
  text-indent: 10px;
}

#qyzhkxujot .gt_indent_3 {
  text-indent: 15px;
}

#qyzhkxujot .gt_indent_4 {
  text-indent: 20px;
}

#qyzhkxujot .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="table gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="r">
r
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="p">
p
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="snp">
snp
</th>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="var">
var
</th>
</tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #AF8DC3; color: #FFFFFF;">
-0.47
</td>
<td headers="p" class="gt_row gt_right">
0.00
</td>
<td headers="snp" class="gt_row gt_left">
Locus_771
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA2
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #BB9CCB; color: #000000;">
-0.42
</td>
<td headers="p" class="gt_row gt_right">
0.00
</td>
<td headers="snp" class="gt_row gt_left">
Locus_263
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA1
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #CCB1D6; color: #000000;">
-0.35
</td>
<td headers="p" class="gt_row gt_right">
0.00
</td>
<td headers="snp" class="gt_row gt_left">
Locus_2767
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA1
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #B1DAAB; color: #000000;">
0.34
</td>
<td headers="p" class="gt_row gt_right">
0.00
</td>
<td headers="snp" class="gt_row gt_left">
Locus_1255
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA3
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #CEB4D7; color: #000000;">
-0.34
</td>
<td headers="p" class="gt_row gt_right">
0.00
</td>
<td headers="snp" class="gt_row gt_left">
Locus_2839
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA1
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #C4E5BE; color: #000000;">
0.29
</td>
<td headers="p" class="gt_row gt_right">
0.01
</td>
<td headers="snp" class="gt_row gt_left">
Locus_2053
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA1
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #DFC9E2; color: #000000;">
-0.27
</td>
<td headers="p" class="gt_row gt_right">
0.02
</td>
<td headers="snp" class="gt_row gt_left">
Locus_191
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA3
</td>
</tr>
<tr>
<td headers="r" class="gt_row gt_right" style="background-color: #E3CFE6; color: #000000;">
-0.25
</td>
<td headers="p" class="gt_row gt_right">
0.03
</td>
<td headers="snp" class="gt_row gt_left">
Locus_2321
</td>
<td headers="var" class="gt_row gt_left">
CA_rPCA1
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level3">
<h3 id="additional-documentation-and-citations">Additional documentation and citations<a class="anchor" aria-label="anchor" href="#additional-documentation-and-citations"></a>
</h3>
<hr>
<table class="table">
<colgroup>
<col width="26%">
<col width="38%">
<col width="34%">
</colgroup>
<thead><tr class="header">
<th></th>
<th>Citation/URL</th>
<th>Details</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Associated literature</td>
<td>
<a href="https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.13722" class="external-link">Capblancq
&amp; Forester 2021</a>; <a href="https://github.com/Capblancq/RDA-landscape-genomics" class="external-link">Github
repository for code</a>
</td>
<td>Paper describing RDA methodology; also provides walkthrough and
description of <code>rdadapt()</code> function</td>
</tr>
<tr class="even">
<td>Associated literature</td>
<td><a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12906" class="external-link">Capblancq
et al. 2018</a></td>
<td>Description of p-value method of detecting outlier loci</td>
</tr>
<tr class="odd">
<td>R package documentation</td>
<td><a href="https://cran.r-project.org/web/packages/vegan/index.html" class="external-link">Oksanen
et al. 2022.</a></td>
<td>algatr uses the <code><a href="https://rdrr.io/pkg/vegan/man/cca.html" class="external-link">rda()</a></code> and <code><a href="https://rdrr.io/pkg/vegan/man/ordistep.html" class="external-link">ordiR2step()</a></code>
functions in the vegan package</td>
</tr>
<tr class="even">
<td>Associated literature</td>
<td><a href="https://onlinelibrary.wiley.com/doi/full/10.1111/mec.14584" class="external-link">Forester
et al. 2018</a></td>
<td>Review of genotype-environment association methods</td>
</tr>
<tr class="odd">
<td>Associated tutorial</td>
<td><a href="https://popgen.nescent.org/2018-03-27_RDA_GEA.html" class="external-link">Forester et
al.</a></td>
<td>Walkthrough of performing an RDA with the Z-scores outlier detection
method</td>
</tr>
<tr class="even">
<td>Associated vignette</td>
<td><code><a href="https://cran.rstudio.com/web/packages/vegan/vignettes/intro-vegan.pdf" class="external-link">vignette("intro-vegan")</a></code></td>
<td>Introduction to ordination using the vegan package (including using
the <code><a href="https://rdrr.io/pkg/vegan/man/cca.html" class="external-link">rda()</a></code> function</td>
</tr>
<tr class="odd">
<td>Associated vignette</td>
<td><code><a href="https://cran.rstudio.com/web/packages/vegan/vignettes/partitioning.pdf" class="external-link">vignette("partitioning")</a></code></td>
<td>Introduction to variance partitioning in vegan</td>
</tr>
</tbody>
</table>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Anusha Bishop, Anne Chambers, Ian Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
