<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="algatr">
<title>TESS • algatr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="TESS">
<meta property="og:description" content="algatr">
<meta property="og:image" content="https://thewanglab.github.io/algatr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">algatr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference Manual</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <a class="dropdown-item" href="../articles/data_processing_vignette.html">Genetic data processing</a>
    <a class="dropdown-item" href="../articles/enviro_data_vignette.html">Environmental data processing</a>
    <a class="dropdown-item" href="../articles/masking_vignette.html">Masking</a>
    <a class="dropdown-item" href="../articles/gen_dist_vignette.html">Genetic distances</a>
    <a class="dropdown-item" href="../articles/TESS_vignette.html">TESS</a>
    <a class="dropdown-item" href="../articles/MMRR_vignette.html">MMRR</a>
    <a class="dropdown-item" href="../articles/GDM_vignette.html">GDM</a>
    <a class="dropdown-item" href="../articles/RDA_vignette.html">RDA</a>
    <a class="dropdown-item" href="../articles/LFMM_vignette.html">LFMM</a>
    <a class="dropdown-item" href="../articles/wingen_vignette.html">Wingen</a>
    <a class="dropdown-item" href="../articles/alazygatr_vignette.html">Alazygatr</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>TESS</h1>
            
      
      
      <div class="d-none name"><code>TESS_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install packages</span></span>
<span><span class="fu"><a href="../reference/tess_packages.html">tess_packages</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="tess">TESS<a class="anchor" aria-label="anchor" href="#tess"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thewanglab.github.io/algatr/">algatr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">wingen</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tess3r</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/raster" class="external-link">raster</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dnychka/fieldsRPackage" class="external-link">fields</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/AndySouth/rworldmap/" class="external-link">rworldmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">automap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># for plotting (from TESS)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"http://membres-timc.imag.fr/Olivier.Francois/Conversion.R"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"http://membres-timc.imag.fr/Olivier.Francois/POPSutilities.R"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Loading fields"</span></span>
<span><span class="co">#&gt; [1] "Loading RColorBrewer"</span></span></code></pre></div>
<p><strong>If using TESS, please cite the following: Caye K., Deist
T.M., Martins H., Michel O., François O. (2016) TESS3: fast inference of
spatial population structure and genome scans for selection. Molecular
Ecology Resources 16(2):540-548. DOI:
10.1111/1755-0998.12471.</strong></p>
<p>TESS is a method to estimate population structure using ancestry
coefficient estimates. As with STRUCTURE, one first describes genetic
variation by assigning individuals according to numbers of clusters (K
values). Given the “best” number of K values, individuals are then given
proportions to which they belong to each K. These proportions correspond
to ancestry coefficients, and can be interpreted as the proportion of an
individual’s ancestry belonging to different ancestral groups. Keep in
mind that you may want to prune out any SNPs that may be in linkage
disequilibrium prior to running TESS so as not to bias or overinflate
the significance of results. Please refer to the data processing
vignette for information on how this can be done in algatr using the
<code><a href="../reference/ld_prune.html">ld_prune()</a></code> function.</p>
<p>Unlike STRUCTURE, TESS is spatially explicit, taking into account
coordinate data; thus, ancestry coefficient estimates incorporate
knowledge of the sampling space. TESS3 (<a href="https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12471" class="external-link">Caye
et al. 2016</a>) was later developed as an extension of the TESS
algorithm, with modifications to the underlying statistical clustering
algorithm to increase computational speed.</p>
<p>For additional information on the original development and
implementation of the algorithm used by TESS, see <a href="https://academic.oup.com/genetics/article/174/2/805/6061653?login=false" class="external-link">François
et al. 2006</a> and <a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1471-8286.2007.01769.x" class="external-link">Chen
et al. 2007</a>. Finally, our code primarily uses the tess3r package
(see <a href="https://bcm-uga.github.io/TESS3_encho_sen/index.html" class="external-link">here</a> for
documentation).</p>
<div class="section level4">
<h4 id="read-in-and-process-input-data">Read in and process input data<a class="anchor" aria-label="anchor" href="#read-in-and-process-input-data"></a>
</h4>
<p>Running TESS3 requires three data files: a genotype dosage matrix
(the <code>gen</code> argument), coordinates for samples (the
<code>coords</code> argument), and environmental layers (the
<code>envlayers</code> argument). We can use a vcf and the
<code><a href="../reference/vcf_to_dosage.html">vcf_to_dosage()</a></code> function to convert a vcf to a dosage
matrix.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/load_algatr_example.html">load_algatr_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; ---------------- example dataset ----------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span> </span></span>
<span><span class="co"><span style="color: #0000BB;">#&gt; Objects loaded:</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_vcf*</span> <span style="color: #BBBB00;">vcfR object (1000 loci x 53 samples)</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_gendist*</span> <span style="color: #BBBB00;">genetic distance matrix (Plink Distance)</span> </span></span>
<span><span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; *liz_coords*</span> <span style="color: #00BB00;">dataframe with x and y coordinates</span> </span></span>
<span><span class="co"><span style="color: #BB00BB; font-weight: bold;">#&gt; *CA_env*</span> <span style="color: #BB00BB;">RasterStack with PC environmental layers</span> </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; -------------------------------------------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co"># Our code assumes that the first column is longitude and second is latitude; check this:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">liz_coords</span><span class="op">)</span></span>
<span><span class="co">#&gt;           x        y</span></span>
<span><span class="co">#&gt; 1 -120.3972 41.56120</span></span>
<span><span class="co">#&gt; 3 -116.8923 34.16940</span></span>
<span><span class="co">#&gt; 5 -124.0408 40.90450</span></span>
<span><span class="co">#&gt; 7 -118.5614 37.78830</span></span>
<span><span class="co">#&gt; 8 -119.7194 37.72386</span></span>
<span><span class="co">#&gt; 9 -121.8467 36.35440</span></span>
<span><span class="co"># Also, our code assumes that sample IDs from gendist and coords are the same order; be sure to check this before moving forward!</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert vcf to genotype matrix</span></span>
<span><span class="va">liz_dosage</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_to_dosage.html">vcf_to_dosage</a></span><span class="op">(</span><span class="va">liz_vcf</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required namespace: adegenet</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="process-environmental-data">Process environmental data<a class="anchor" aria-label="anchor" href="#process-environmental-data"></a>
</h4>
<p>algatr can create interpolated maps of the ancestry coefficient
estimates using a method known as kriging, which uses a spatially
explicit model for spatial interpolation. <em>N.B.: Be aware that tess3r
uses a different type of kriging to generate maps of ancestry
coefficients (algatr uses the <code><a href="https://rdrr.io/pkg/automap/man/autoKrige.html" class="external-link">autoKrige()</a></code> function within
the automap package), and tess3r performs this interpolation and
plotting in the same function, so there is no easy way to produce a
raster output.</em></p>
<p>To generate a map, we need a raster onto which we can map Q values.
We can either retrieve this, or generate one ourselves using wingen’s
<code><a href="https://rdrr.io/pkg/wingen/man/coords_to_raster.html" class="external-link">coords_to_raster()</a></code> function. Please refer to the wingen
documentation found <a href="https://github.com/AnushaPB/wingen" class="external-link">here</a>, and the
environmental data vignette, for further information on this
function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First, create a grid for kriging</span></span>
<span><span class="co"># We can use one environmental layer (PC1), aggregated (i.e., increased cell size) to increase computational speed</span></span>
<span><span class="va">krig_raster</span> <span class="op">&lt;-</span> <span class="fu">raster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">CA_env</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, fact <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># If you want to see the difference between the non-aggregated (original) and aggregated rasters:</span></span>
<span><span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">CA_env</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, col <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">mako</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, axes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/raster-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">krig_raster</span>, col <span class="op">=</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">mako</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, axes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/raster-2.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="k-selection">K selection<a class="anchor" aria-label="anchor" href="#k-selection"></a>
</h3>
<hr>
<p>algatr allows users with the option to test a range of K values using
the <code><a href="../reference/tess_ktest.html">tess_ktest()</a></code> function. K selection is often
accomplished by minimizing the cross entropy, but this can over-simplify
K selection. The <code><a href="../reference/tess_ktest.html">tess_ktest()</a></code> function allows users to
specify how K selection is done using the <code>K_selection</code>
argument; users can set the K values manually (<code>"manual"</code>) or
using an automatic approach described below (<code>"auto"</code>).
<code><a href="../reference/tess_ktest.html">tess_ktest()</a></code> runs TESS for each value of K within the
user-specified range, and outputs cross-validation scores for different
K values which allows a user to select the “best” K for their dataset.
Users have the option to select the algorithm with which TESS will be
run using the <code>tess_method</code> argument; options for the
argument are either projected least squares algorithm
(<code>"projected.ls"</code>; the default) or an alternating quadratic
programming algorithm (<code>"qp"</code>). Finally, as mentioned at the
beginning of the vignette, TESS is spatially explicit meaning that it
can take into account the distance between samples. The degree to which
users suspect spatial autocorrelation between data points can be
controlled using the spatial regularization (alpha) parameter, specified
with the <code>lambda</code> argument. Values closer to 1 samples
geographically closer are also suspected to be more genetically similar
while values closer to 0 imply that there is little spatial
autocorrelation.</p>
<div class="section level4">
<h4 id="manual-k-selection">Manual K selection<a class="anchor" aria-label="anchor" href="#manual-k-selection"></a>
</h4>
<p>The default K selection (<code>"manual"</code>) within
<code><a href="../reference/tess_ktest.html">tess_ktest()</a></code> runs TESS on a set of K values ranging from
1-10, and allows the user to select the “best” K value manually based on
cross-validation scores. Typically, researchers will select the “best” K
value by minimizing cross-validation scores.</p>
</div>
<div class="section level4">
<h4 id="automatic-k-selection">Automatic K selection<a class="anchor" aria-label="anchor" href="#automatic-k-selection"></a>
</h4>
<p>The other K selection method available is automatic K selection by
specifying <code>"auto"</code> within the <code><a href="../reference/tess_ktest.html">tess_ktest()</a></code>
function. This method is based on not only minimizing cross-entropy
scores (which is typically done), but minimizes the slope of the line
that connects cross-entropy scores between K values. In this way, the
best K is selected once an “elbow” is visible, or when cross-entropy
scores begin to plateau and the slope approaches 0. This function makes
use of the approach described <a href="https://chazhyseni.github.io/NALgen/post/determining_bestk/" class="external-link">here</a>.</p>
<p>The resulting object contains <code>tess3_obj</code>, which provides
the cross-validation scores for the different K values that were tested.
Let’s do the automatic K-value testing procedure once again for K values
1-10. As you can see, the <code><a href="../reference/tess_ktest.html">tess_ktest()</a></code> function outputs a
plot with cross-validation scores for each K value, with a red dashed
line indicating the best K value obtained from the automatic K selection
procedure described above.</p>
<p>For this vignette, we’ll use automatic K selection.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Best K is 3; this provides a more reasonable estimate for the "best" K compared to manual selection above</span></span>
<span><span class="va">tess3_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tess_ktest.html">tess_ktest</a></span><span class="op">(</span><span class="va">liz_dosage</span>, <span class="va">liz_coords</span>, Kvals <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, ploidy <span class="op">=</span> <span class="fl">2</span>, K_selection <span class="op">=</span> <span class="st">"auto"</span><span class="op">)</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/auto%20K%20selection-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="k-selection-results-from-test_ktest">K selection results from <code>test_ktest()</code><a class="anchor" aria-label="anchor" href="#k-selection-results-from-test_ktest"></a>
</h4>
<p>The <code>tess3_result</code> object contains results for the
best-supported K value, including:</p>
<ul>
<li><p><code>K</code>: The value of the best-supported K (3, in this
case)</p></li>
<li><p><code>tess3_obj</code>: Results from the cross-validation
analysis of all values of K</p></li>
<li><p><code>coords</code>: Sampling coordinates</p></li>
<li><p><code>Kvals</code>: The range of K values that were
tested</p></li>
<li><p><code>grid</code>: The RasterLayer upon which ancestry
coefficients are mapped; <code>NULL</code> in this case</p></li>
<li><p><code>pops</code>: Population assignments (determined based on
the maximum Q value) for each individual for the best K value</p></li>
</ul>
</div>
<div class="section level4">
<h4 id="running-tess-with-no-k-selection">Running TESS with no K selection<a class="anchor" aria-label="anchor" href="#running-tess-with-no-k-selection"></a>
</h4>
<p>If you want to run TESS without any K selection, you could just run
TESS as normal by doing the following:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tess3_obj_noK</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tess3r/man/tess3.html" class="external-link">tess3</a></span><span class="op">(</span><span class="va">liz_dosage</span>, coord <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/coerce.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">liz_coords</span><span class="op">)</span>, K <span class="op">=</span> <span class="fl">3</span>, method <span class="op">=</span> <span class="st">"projected.ls"</span>, ploidy <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; == Computing spectral decomposition of graph laplacian matrix: done</span></span>
<span><span class="co">#&gt; ==Main loop with 1 threads: done</span></span></code></pre></div>
<p>The resulting object from the above run is a typical TESS output
format (it is also contained within the above <code>tess3_obj</code>
object); see tess3r vignette for details <a href="https://bcm-uga.github.io/TESS3_encho_sen/articles/main-vignette.html" class="external-link">here</a>
to read more.</p>
</div>
</div>
<div class="section level3">
<h3 id="extracting-tess-results">Extracting TESS results<a class="anchor" aria-label="anchor" href="#extracting-tess-results"></a>
</h3>
<hr>
<p>Now that we know which K value we want to use (K = 3) from our
automatic K selection results, we can move forward with TESS. We need to
extract the tess3 object from our results, and create a Q-matrix with
ancestry coefficient values from K = 3 using the <code><a href="https://rdrr.io/pkg/tess3r/man/qmatrix.html" class="external-link">qmatrix()</a></code>
function with which we can visualize results.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get TESS object and best K from results</span></span>
<span><span class="va">tess3_obj</span> <span class="op">&lt;-</span> <span class="va">tess3_result</span><span class="op">$</span><span class="va">tess3_obj</span></span>
<span><span class="va">bestK</span> <span class="op">&lt;-</span> <span class="va">tess3_result</span><span class="op">[[</span><span class="st">"K"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Get Qmatrix with ancestry coefficients</span></span>
<span><span class="va">qmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tess3r/man/qmatrix.html" class="external-link">qmatrix</a></span><span class="op">(</span><span class="va">tess3_obj</span>, K <span class="op">=</span> <span class="va">bestK</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># qmat contains ancestry coefficient values for each individual (row) and each K value (column)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">qmat</span><span class="op">)</span></span>
<span><span class="co">#&gt;              [,1]         [,2]         [,3]</span></span>
<span><span class="co">#&gt; [1,] 9.999821e-01 8.946409e-06 8.946409e-06</span></span>
<span><span class="co">#&gt; [2,] 9.085585e-06 9.999818e-01 9.085585e-06</span></span>
<span><span class="co">#&gt; [3,] 7.212317e-01 1.298649e-01 1.489034e-01</span></span>
<span><span class="co">#&gt; [4,] 1.083884e-01 6.331817e-02 8.282934e-01</span></span>
<span><span class="co">#&gt; [5,] 8.999576e-01 1.000326e-01 9.841997e-06</span></span>
<span><span class="co">#&gt; [6,] 9.126593e-01 9.847514e-06 8.733083e-02</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="krige-q-values">Krige Q values<a class="anchor" aria-label="anchor" href="#krige-q-values"></a>
</h3>
<hr>
<p>The <code><a href="../reference/tess_krig.html">tess_krig()</a></code> function will take in ancestry
coefficient values (in the Q-matrix) and will krige the values based on
the raster provided (<code>krig_raster</code> from above). This will
produce a Raster* type object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">krig_admix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tess_krig.html">tess_krig</a></span><span class="op">(</span><span class="va">qmat</span>, <span class="va">liz_coords</span>, <span class="va">krig_raster</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-tess-results">Visualizing TESS results<a class="anchor" aria-label="anchor" href="#visualizing-tess-results"></a>
</h3>
<hr>
<div class="section level4">
<h4 id="bar-plot-of-q-values">Bar plot of Q values<a class="anchor" aria-label="anchor" href="#bar-plot-of-q-values"></a>
</h4>
<p>A typical representation for population structure results is a
STRUCTURE-style bar plot, in which each stacked bar represents an
individual and the proportion of stacked color represents the proportion
of ancestry assigned to that individual for each cluster (or K value).
In our case, we have K=3, so there will be three colors representing
each of these K values.</p>
<p>algatr has a custom function, <code><a href="../reference/tess_barplot.html">tess_barplot()</a></code>, to
generate these plots from our TESS results. This automatically sort
individuals based on their Q values (although this can be modified using
the <code>sort_by_Q</code> argument, if so desired); the outputted
“order” message indicates how individuals are ordered in the bar
plot.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tess_barplot.html">tess_barplot</a></span><span class="op">(</span><span class="va">qmat</span><span class="op">)</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/barplot-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; $order</span></span>
<span><span class="co">#&gt;  [1]  4 13 20 28 29 31 32 35 36 37 52 53  1  3  5  6  7  8  9 11 12 14 15 16 17</span></span>
<span><span class="co">#&gt; [26] 18 21 22 23 26 27 30 33 34 38 39 42 43 44 45 46 47 48  2 10 19 24 25 40 41</span></span>
<span><span class="co">#&gt; [51] 49 50 51</span></span></code></pre>
<p>We can also build bar plots using ggplot2 with the
<code><a href="../reference/tess_ggbarplot.html">tess_ggbarplot()</a></code> function.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tess_ggbarplot.html">tess_ggbarplot</a></span><span class="op">(</span><span class="va">qmat</span>, legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/ggbarplot-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="building-maps-in-tess3-using-tess_ggplot">Building maps in TESS3 using <code>tess_ggplot()</code><a class="anchor" aria-label="anchor" href="#building-maps-in-tess3-using-tess_ggplot"></a>
</h4>
<p>Now let’s explore how Q values (ancestry coefficients) are mapped.
The <code><a href="../reference/tess_ggplot.html">tess_ggplot()</a></code> function will take in the kriged
admixture values and sampling coordinates (if the user wants points
mapped), and provides several options for the plot method (with the
<code>plot_method</code> argument):</p>
<ul>
<li><p><code>"maxQ"</code> plots only the maximum Q value for each cell
(this is the default)</p></li>
<li><p><code>"allQ"</code> plots all Q-values that are greater than a
user-specified <code>minQ</code> value</p></li>
<li><p><code>"maxQ_poly"</code> plots maxQ as polygons for each K-value
instead of continuous Q values</p></li>
<li><p><code>"allQ_poly"</code> plots allQ as polygons for each K-value
instead of continuous Q values</p></li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, pty <span class="op">=</span> <span class="st">"s"</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/tess_ggplot.html">tess_ggplot</a></span><span class="op">(</span><span class="va">krig_admix</span>, plot_method <span class="op">=</span> <span class="st">"maxQ"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Using alpha for a discrete variable is not advised.</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/basic%20ggplot-1.png" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tess_ggplot.html">tess_ggplot</a></span><span class="op">(</span><span class="va">krig_admix</span>, plot_method <span class="op">=</span> <span class="st">"allQ"</span>, minQ <span class="op">=</span> <span class="fl">0.20</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Using alpha for a discrete variable is not advised.</span></span>
<span><span class="co">#&gt; $plot</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/basic%20ggplot-2.png" width="480" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $legend</span></span>
<span><span class="co">#&gt; Warning: Using alpha for a discrete variable is not advised.</span></span></code></pre>
<p><img src="TESS_vignette_files/figure-html/basic%20ggplot-3.png" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tess_ggplot.html">tess_ggplot</a></span><span class="op">(</span><span class="va">krig_admix</span>, plot_method <span class="op">=</span> <span class="st">"maxQ_poly"</span><span class="op">)</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/basic%20ggplot-4.png" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tess_ggplot.html">tess_ggplot</a></span><span class="op">(</span><span class="va">krig_admix</span>, plot_method <span class="op">=</span> <span class="st">"allQ_poly"</span>, minQ <span class="op">=</span> <span class="fl">0.20</span><span class="op">)</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/basic%20ggplot-5.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="extended-plotting">Extended plotting<a class="anchor" aria-label="anchor" href="#extended-plotting"></a>
</h4>
<p>In many cases, a user may want more customizability in mapping their
TESS results. If users want more control over color, they can use the
<code>ggplot_fill</code> argument. Also, if users they want x and y axes
to be displayed, this can be set using the <code>plot_axes</code>
argument:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tess_ggplot.html">tess_ggplot</a></span><span class="op">(</span><span class="va">krig_admix</span>, plot_method <span class="op">=</span> <span class="st">"maxQ"</span>, ggplot_fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#bd9dac"</span>, <span class="st">"#257b94"</span>, <span class="st">"#476e9e"</span><span class="op">)</span><span class="op">)</span>, plot_axes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/extended%20mapping-1.png" width="480" style="display: block; margin: auto;"></p>
<p>We may also want to plot individual layers (K values), which we can
using <code><a href="../reference/tess_plot_allK.html">tess_plot_allK()</a></code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/pkg/terra/man/dimensions.html" class="external-link">nlyr</a></span><span class="op">(</span><span class="va">krig_admix</span><span class="op">)</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, oma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tess_plot_allK.html">tess_plot_allK</a></span><span class="op">(</span><span class="va">krig_admix</span>, col_breaks <span class="op">=</span> <span class="fl">20</span>, legend.width <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/plot%20all%20K-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="plotting-with-default-tess3r-package-functions">Plotting with default tess3r package functions<a class="anchor" aria-label="anchor" href="#plotting-with-default-tess3r-package-functions"></a>
</h3>
<p>You can, of course, also plot TESS results using the tess3r package
defaults and not those provided with algatr. The TESS function for a
barplot is <code><a href="https://rdrr.io/pkg/terra/man/barplot.html" class="external-link">barplot()</a></code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="va">qmat</span>, sort.by.Q <span class="op">=</span> <span class="cn">TRUE</span>, border <span class="op">=</span> <span class="cn">NA</span>, space <span class="op">=</span> <span class="fl">0</span>, xlab <span class="op">=</span> <span class="st">"Individuals"</span>, ylab <span class="op">=</span> <span class="st">"Ancestry coefficients"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Use CreatePalette() to define color palettes.</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/tess%20default%20barplot-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; $order</span></span>
<span><span class="co">#&gt;  [1]  4 13 20 28 29 31 32 35 36 37 52 53  1  3  5  6  7  8  9 11 12 14 15 16 17</span></span>
<span><span class="co">#&gt; [26] 18 21 22 23 26 27 30 33 34 38 39 42 43 44 45 46 47 48  2 10 19 24 25 40 41</span></span>
<span><span class="co">#&gt; [51] 49 50 51</span></span></code></pre>
<p>You can also map kriged Q-values using tess3r with the
<code><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot()</a></code> function.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">qmat</span>,</span>
<span>  <span class="va">liz_coords</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"map.max"</span>,</span>
<span>  interpol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/tess3r/man/FieldsKrigModel.html" class="external-link">FieldsKrigModel</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Ancestry coefficients"</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"Longitude"</span>, ylab <span class="op">=</span> <span class="st">"Latitude"</span>,</span>
<span>  col.palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/tess3r/man/CreatePalette.html" class="external-link">CreatePalette</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">300</span>, <span class="fl">300</span><span class="op">)</span>, cex <span class="op">=</span> <span class="fl">.4</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: </span></span>
<span><span class="co">#&gt; Grid searches over lambda (nugget and sill variances) with  minima at the endpoints: </span></span>
<span><span class="co">#&gt;   (REML) Restricted maximum likelihood </span></span>
<span><span class="co">#&gt;    minimum at  right endpoint  lambda  =  0.003324997 (eff. df= 38.94999 )</span></span>
<span><span class="co">#&gt; Warning: </span></span>
<span><span class="co">#&gt; Grid searches over lambda (nugget and sill variances) with  minima at the endpoints: </span></span>
<span><span class="co">#&gt;   (REML) Restricted maximum likelihood </span></span>
<span><span class="co">#&gt;    minimum at  right endpoint  lambda  =  0.003324997 (eff. df= 38.94999 )</span></span>
<span><span class="co">#&gt; Warning: </span></span>
<span><span class="co">#&gt; Grid searches over lambda (nugget and sill variances) with  minima at the endpoints: </span></span>
<span><span class="co">#&gt;   (REML) Restricted maximum likelihood </span></span>
<span><span class="co">#&gt;    minimum at  right endpoint  lambda  =  0.003324997 (eff. df= 38.94999 )</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/unnamed-chunk-4-1.png" width="480" style="display: block; margin: auto;"></p>
<p>As you can see from the above, interpolation is done using a
different method in tess3r, which is why the resulting smoothed surface
differs from than produced using algatr.</p>
</div>
<div class="section level3">
<h3 id="running-tess-with-tess_do_everything">Running TESS with <code>tess_do_everything()</code><a class="anchor" aria-label="anchor" href="#running-tess-with-tess_do_everything"></a>
</h3>
<hr>
<p>The algatr package also has an option to run all of the above
functionality in a single function, <code><a href="../reference/tess_do_everything.html">tess_do_everything()</a></code>.
<em>N.B.: the <code><a href="../reference/tess_do_everything.html">tess_do_everything()</a></code> function does not
require a dosage matrix; it will do the conversion automatically if a
vcf is provided</em>. <strong>Please be aware that the
<code>do_everything()</code> functions are meant to be exploratory. We
do not recommend their use for final analyses unless certain they are
properly parameterized.</strong></p>
<p>The <code><a href="../reference/tess_do_everything.html">tess_do_everything()</a></code> function will print out the
barplot, the cross-validation scores for the range of K values tested,
and the kriged map (if a grid is provided). It will also print the order
in which individuals are ordered in the barplot. The resulting object
from this function contains:</p>
<ul>
<li><p><code>K</code>: The best K value based on user-specified K
selection</p></li>
<li><p><code>Qmatrix</code>: The matrix of individual ancestry
coefficients, or Q values</p></li>
<li><p><code>krig_admix</code>: A RasterBrick object containing the
kriged ancestry coefficient values for mapping; if no grid is provided,
the function will skip kriging</p></li>
<li><p><code>tess_results</code>: The tess3r object containing
cross-entropy scores for each K value tested</p></li>
<li><p><code>coords</code>: Sampling coordinates</p></li>
<li><p><code>Kvals</code>: Range of K values that were tested</p></li>
<li><p><code>grid</code>: The RasterLayer upon which Q values will be
mapped</p></li>
</ul>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># One could also use krig_raster as the grid object</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tess_do_everything.html">tess_do_everything</a></span><span class="op">(</span><span class="va">liz_vcf</span>, <span class="va">liz_coords</span>, grid <span class="op">=</span> <span class="va">CA_env</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, Kvals <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, K_selection <span class="op">=</span> <span class="st">"auto"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Please be aware: the do_everything functions are meant to be exploratory. We do not recommend their use for final analyses unless certain they are properly parameterized.</span></span>
<span><span class="co">#&gt; New names:</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">•</span> `` -&gt; `...1`</span></span></code></pre></div>
<p><img src="TESS_vignette_files/figure-html/do%20everything-1.png" width="480" style="display: block; margin: auto;"><img src="TESS_vignette_files/figure-html/do%20everything-2.png" width="480" style="display: block; margin: auto;"><img src="TESS_vignette_files/figure-html/do%20everything-3.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="additional-documentation-and-citations">Additional documentation and citations<a class="anchor" aria-label="anchor" href="#additional-documentation-and-citations"></a>
</h3>
<hr>
<table class="table">
<colgroup>
<col width="26%">
<col width="39%">
<col width="34%">
</colgroup>
<thead><tr class="header">
<th></th>
<th>Citation/URL</th>
<th>Details</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Main literature</td>
<td>
<a href="https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12471" class="external-link">Caye
et al. 2016</a>; vignette available <a href="https://bcm-uga.github.io/TESS3_encho_sen/articles/main-vignette.html" class="external-link">here</a>
</td>
<td>Citation for TESS3</td>
</tr>
<tr class="even">
<td>Associated literature</td>
<td><a href="https://academic.oup.com/genetics/article/174/2/805/6061653?login=false" class="external-link">François
et al. 2006</a></td>
<td>Details on algorithm used by TESS</td>
</tr>
<tr class="odd">
<td>Associated literature</td>
<td><a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1471-8286.2007.01769.x" class="external-link">Chen
et al. 2007</a></td>
<td>Details on algorithm used by TESS</td>
</tr>
<tr class="even">
<td>Blog post</td>
<td><a href="https://chazhyseni.github.io/NALgen/post/determining_bestk/" class="external-link">Automatic
K selection</a></td>
<td>Details on automatic K selection used by algatr</td>
</tr>
</tbody>
</table>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Anusha Bishop, Anne Chambers, Ian Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
