<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="algatr">
<title>Wingen • algatr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Wingen">
<meta property="og:description" content="algatr">
<meta property="og:image" content="https://thewanglab.github.io/algatr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">algatr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/data_processing_vignette.html">Genetic data processing</a>
    <a class="dropdown-item" href="../articles/enviro_data_vignette.html">Environmental data processing</a>
    <a class="dropdown-item" href="../articles/Masking_vignette.html">Masking</a>
    <a class="dropdown-item" href="../articles/gen_dist_vignette.html">Genetic distances</a>
    <a class="dropdown-item" href="../articles/TESS_vignette.html">TESS</a>
    <a class="dropdown-item" href="../articles/MMRR_vignette.html">MMRR_vignette</a>
    <a class="dropdown-item" href="../articles/GDM_vignette.html">GDM</a>
    <a class="dropdown-item" href="../articles/RDA_vignette.html">RDA</a>
    <a class="dropdown-item" href="../articles/LFMM_vignette.html">LFMM</a>
    <a class="dropdown-item" href="../articles/wingen_vignette.html">Wingen</a>
    <a class="dropdown-item" href="../articles/alazygatr_vignette.html">Alazygatr</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Wingen</h1>
            
      
      
      <div class="d-none name"><code>wingen_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install packages</span></span>
<span><span class="fu"><a href="../reference/wingen_packages.html">wingen_packages</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="wingen">wingen<a class="anchor" aria-label="anchor" href="#wingen"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thewanglab.github.io/algatr/">algatr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">wingen</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/raster" class="external-link">raster</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span></code></pre></div>
<p><strong>If using wingen, please cite the following: Bishop A.P.,
Chambers E.A., Wang I.J. (2023) Generating continuous maps of genetic
diversity using moving windows. Methods in Ecology and Evolution. DOI:
<a href="https://doi.org/10.1111/2041-210X.14090" class="external-link uri">https://doi.org/10.1111/2041-210X.14090</a>.</strong></p>
<p>wingen is a package that uses a moving window approach to create
continuous maps of genetic diversity. Please see <a href="https://doi.org/10.1111/2041-210X.14090" class="external-link">Bishop et al. 2023</a>
for the paper describing the method, and refer to the <a href="https://github.com/AnushaPB/wingen" class="external-link">wingen package</a> for further
documentation.</p>
<p>Briefly, wingen takes in genetic data (in the form of a vcf),
sampling coordinates, and a raster of a given study area. wingen then
moves a window (of a user-specified size) across the landscape, and
calculates genetic diversity for any samples that fall within this
window. The focal cell (the size of which is determined by the raster
layer inputted) is then assigned this resulting genetic diversity value,
and the window continues sliding across the landscape. Users are able to
specify not only window size, but can also aggregate cells to increase
map continuity and decrease computational time, and can perform
rarefaction of samples such that uneven sampling does not bias
calculations. The resulting wingen map can be smoothed using kriging,
and any areas that were undersampled (or fall outside an area of
interest, such as a species range, for example) are masked.</p>
<p>There are five key functions within wingen:</p>
<ul>
<li><p><code><a href="https://rdrr.io/pkg/wingen/man/preview_gd.html" class="external-link">preview_gd()</a></code> provides a reference point for what the
window and focal cell look like relative to the landscape; this function
also allows users to visualize the sample count in each cell given the
parameters</p></li>
<li><p><code><a href="https://rdrr.io/pkg/wingen/man/window_gd.html" class="external-link">window_gd()</a></code> to create a map of genetic
diversity</p></li>
<li><p><code><a href="https://rdrr.io/pkg/wingen/man/krig_gd.html" class="external-link">krig_gd()</a></code> for spatial interpolation (smoothing) of
the resulting map using kriging</p></li>
<li><p><code><a href="https://rdrr.io/pkg/wingen/man/mask_gd.html" class="external-link">mask_gd()</a></code> to mask out any undersampled areas (or
areas that are not of interest)</p></li>
<li><p><code><a href="https://rdrr.io/pkg/wingen/man/plot_gd.html" class="external-link">plot_gd()</a></code> to plot resulting genetic diversity map
and <code><a href="https://rdrr.io/pkg/wingen/man/plot_count.html" class="external-link">plot_count()</a></code> to plot the resulting sample count
map</p></li>
</ul>
<p><em>Note:</em> Coordinates and rasters used in wingen should be in a
projected (planar) coordinate system such that raster cells are of equal
sizes. Therefore, spherical systems (including latitude-longitude
coordinate systems) should be projected prior to use. An example of how
this can be accomplished is shown below. If no CRS is provided, a
warning will be given and wingen will assume the data are provided in a
projected system. Please refer to the wingen vignette for further
information.</p>
<div class="section level4">
<h4 id="read-in-and-process-input-data">Read in and process input data<a class="anchor" aria-label="anchor" href="#read-in-and-process-input-data"></a>
</h4>
<p>wingen requires a vcf, sampling coordinates, and a raster layer upon
which to move the sliding window across. We’ll also want the shape of
California for subsequent plotting, so we’ll save one of the
environmental PC layers as a SpatRaster object for this purpose.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/load_algatr_example.html">load_algatr_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; ---------------- example dataset ----------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span> </span></span>
<span><span class="co"><span style="color: #0000BB;">#&gt; Objects loaded:</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_vcf*</span> <span style="color: #BBBB00;">vcfR object (1000 loci x 53 samples)</span> </span></span>
<span><span class="co"><span style="color: #BBBB00; font-weight: bold;">#&gt; *liz_gendist*</span> <span style="color: #BBBB00;">genetic distance matrix (Plink Distance)</span> </span></span>
<span><span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; *liz_coords*</span> <span style="color: #00BB00;">dataframe with x and y coordinates</span> </span></span>
<span><span class="co"><span style="color: #BB00BB; font-weight: bold;">#&gt; *CA_env*</span> <span style="color: #BB00BB;">RasterStack with PC environmental layers</span> </span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; -------------------------------------------------</span></span></span>
<span><span class="co"><span style="color: #00BBBB; font-weight: bold;">#&gt; </span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="va">envlayer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="va">CA_env</span><span class="op">$</span><span class="va">CA_rPCA1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="generate-raster-layer-for-sliding-window">Generate raster layer for sliding window<a class="anchor" aria-label="anchor" href="#generate-raster-layer-for-sliding-window"></a>
</h4>
<p>Although we can use one of the PCs within our CA_env object as a
raster for wingen, the resolution of those layers is unnecessarily high
given our sampling localities, and running wingen (particularly the
kriging step) will take a very long time. To remedy this, we can
generate a raster layer using the <code><a href="https://rdrr.io/pkg/wingen/man/coords_to_raster.html" class="external-link">coords_to_raster()</a></code>
function in algatr by providing our sampling coordinates. Using this
function, we can specify the resolution of our raster with the
<code>res</code> argument.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liz_lyr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/wingen/man/coords_to_raster.html" class="external-link">coords_to_raster</a></span><span class="op">(</span><span class="va">liz_coords</span>, res <span class="op">=</span> <span class="fl">0.5</span>, buffer <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="moving-window-calculations-using-preview_gd-and-window_gd">Moving window calculations using <code>preview_gd()</code> and
<code>window_gd()</code><a class="anchor" aria-label="anchor" href="#moving-window-calculations-using-preview_gd-and-window_gd"></a>
</h3>
<hr>
<p>The <code><a href="https://rdrr.io/pkg/wingen/man/window_gd.html" class="external-link">window_gd()</a></code> function takes in a vcf file (the
<code>vcf</code> argument), sampling coordinates (<code>coords</code>
argument), and a RasterLayer for the window to move across (the
<code>lyr</code> argument), and the genetic diversity summary statistic
to calculate (the <code>stat</code> argument); choices are allelic
richness (<code>"allelic.richness"</code> or
<code>"biallelic.richness"</code>), nucleotide diversity
(<code>"pi"</code>), or average heterozygosity (<code>"Ho"</code>). For
more options for file types and statistics see the original wingen
vignette.</p>
<p>There are a number of additional arguments within this function, and
the wingen package documentation provides extensive explanations (and
guidelines for testing) these parameters. Briefly, <code>fact</code>
aggregates the input raster layer (by a given factor), <code>wdim</code>
specifies the window dimensions, <code>rarify</code> specifies whether
the user would like samples to be rarified (which then requires users to
specify <code>rarify_n</code> and <code>rarify_nit</code>), and finally
<code>parallel</code> specifies whether you would like to
parallelize.</p>
<div class="section level4">
<h4 id="preview-the-moving-window">Preview the moving window<a class="anchor" aria-label="anchor" href="#preview-the-moving-window"></a>
</h4>
<p>First, we can get an idea of what the size of the cell and moving
window look like using the <code><a href="https://rdrr.io/pkg/wingen/man/preview_gd.html" class="external-link">preview_gd()</a></code> function. We want
adjust the window size (using the <code>wdim</code> argument) to ensure
(to some extent) that our window size is capturing samples as it slides
across the landscape. Ideally, <code>wdim</code> should be set with the
study system in mind (e.g., the dispersal distance or neighborhood size
of the species). We will also specify that genetic diversity only be
calculated for windows that contain at least two samples
(<code>min_n</code>). This function will generate two figures: one that
allows us to visualize the size of our window compared to the raster
layer and sampling coordinates, and also a map of how many samples will
be included in each cell’s calculation (specified using
<code>sample_count</code>).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, oma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/wingen/man/preview_gd.html" class="external-link">preview_gd</a></span><span class="op">(</span><span class="va">liz_lyr</span>,</span>
<span>  <span class="va">liz_coords</span>,</span>
<span>  wdim <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  fact <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="wingen_vignette_files/figure-html/preview%20gd-1.png" width="672" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; class       : SpatRaster </span></span>
<span><span class="co">#&gt; dimensions  : 27, 25, 1  (nrow, ncol, nlyr)</span></span>
<span><span class="co">#&gt; resolution  : 0.5, 0.5  (x, y)</span></span>
<span><span class="co">#&gt; extent      : -126.5408, -114.0408, 30.372, 43.872  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#&gt; coord. ref. :  </span></span>
<span><span class="co">#&gt; source(s)   : memory</span></span>
<span><span class="co">#&gt; name        : sample_count </span></span>
<span><span class="co">#&gt; min value   :            0 </span></span>
<span><span class="co">#&gt; max value   :           10</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="run-the-moving-window">Run the moving window<a class="anchor" aria-label="anchor" href="#run-the-moving-window"></a>
</h4>
<p>Next, let’s run the moving window function (<code><a href="https://rdrr.io/pkg/wingen/man/window_gd.html" class="external-link">window_gd()</a></code>)
using the parameters specified above, calculating nucleotide diversity
(<code>stat = "pi"</code>) as our diversity metric. For simplicity’s
sake, we will not perform rarefaction of our samples
(<code>rarify = FALSE</code>, which is the default).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wgd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/wingen/man/window_gd.html" class="external-link">window_gd</a></span><span class="op">(</span><span class="va">liz_vcf</span>,</span>
<span>  <span class="va">liz_coords</span>,</span>
<span>  <span class="va">liz_lyr</span>,</span>
<span>  stat <span class="op">=</span> <span class="st">"pi"</span>,</span>
<span>  wdim <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  fact <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: vcfR</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    *****       ***   vcfR   ***       *****</span></span>
<span><span class="co">#&gt;    This is vcfR 1.14.0 </span></span>
<span><span class="co">#&gt;      browseVignettes('vcfR') # Documentation</span></span>
<span><span class="co">#&gt;      citation('vcfR') # Citation</span></span>
<span><span class="co">#&gt;    *****       *****      *****       *****</span></span>
<span><span class="co">#&gt; Loading required namespace: adegenet</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="visualizing-wingen-results">Visualizing wingen results<a class="anchor" aria-label="anchor" href="#visualizing-wingen-results"></a>
</h3>
<hr>
<div class="section level4">
<h4 id="plot-wingen-results-with-plot_gd-and-plot_count">Plot wingen results with <code>plot_gd()</code> and
<code>plot_count()</code><a class="anchor" aria-label="anchor" href="#plot-wingen-results-with-plot_gd-and-plot_count"></a>
</h4>
<p>wingen allows for plotting maps in two ways: <code><a href="https://rdrr.io/pkg/wingen/man/plot_gd.html" class="external-link">plot_gd()</a></code>
plots the genetic diversity layer, while <code><a href="https://rdrr.io/pkg/wingen/man/plot_count.html" class="external-link">plot_count()</a></code> plots
the sample count layer. We can also provide a background map using the
<code>bkg</code> argument.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, oma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Plot map of pi</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/wingen/man/plot_gd.html" class="external-link">plot_gd</a></span><span class="op">(</span><span class="va">wgd</span>, main <span class="op">=</span> <span class="st">"Moving window pi"</span>, bkg <span class="op">=</span> <span class="va">envlayer</span><span class="op">)</span></span>
<span><span class="co"># Plot sample count map</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/wingen/man/plot_count.html" class="external-link">plot_count</a></span><span class="op">(</span><span class="va">wgd</span>, main <span class="op">=</span> <span class="st">"Sample counts"</span><span class="op">)</span></span></code></pre></div>
<p><img src="wingen_vignette_files/figure-html/plot%20window-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="krige-results">Krige results<a class="anchor" aria-label="anchor" href="#krige-results"></a>
</h4>
<p>Kriging is a type of spatial interpolation that uses a spatial model
to provide estimates of genetic diversity from unsampled locations; in
the cae of wingen, it also allows for smoother maps of genetic
diversity. We can krige our moving window map using the
<code><a href="https://rdrr.io/pkg/wingen/man/krig_gd.html" class="external-link">krig_gd()</a></code> function by providing the results from the moving
window, and a raster layer upon which kriging is performed (an
“interpolation grid”); if no interpolation grid is provided, kriging
will be done on the moving window layer. The <code>index</code> argument
specifies which layer we want to krige across; in our case, we’ll krige
both layers 1 (genetic diversity) and 2 (sample counts) of the wgd
object because we will eventually want to visualize both kriged maps.
Kriging can be computationally intensive, but we can remedy this by
aggregating the raster layer upon which kriging is performed using
either the <code>agg_r</code> or <code>agg_grd</code> arguments
depending on whether we want to aggregate across the moving window layer
or the interpolation grid layer, respectively. In our case, given the
resolution of <code>lyr</code>, we want to disaggregate by a factor of 5
to get a smoother interpolated surface, so we’ll set
<code>disagg_grd = 5</code>. If only the moving window layer is provided
for kriging, we can use the <code>disagg_r</code> argument to set this
parameter.</p>
<p>The resulting object, <code>kgd</code>, is a RasterStack object
containing two layers: our genetic diversity statistic (pi), and our
sample count information (sample_count).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kgd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/wingen/man/krig_gd.html" class="external-link">krig_gd</a></span><span class="op">(</span><span class="va">wgd</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="va">liz_lyr</span>, disagg_grd <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; [using ordinary kriging]</span></span>
<span><span class="co">#&gt; [using ordinary kriging]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">kgd</span><span class="op">)</span></span>
<span><span class="co">#&gt;        pi           sample_count    </span></span>
<span><span class="co">#&gt;  Min.   :0.03367   Min.   : 0.0000  </span></span>
<span><span class="co">#&gt;  1st Qu.:0.06715   1st Qu.: 0.0000  </span></span>
<span><span class="co">#&gt;  Median :0.06728   Median : 0.0000  </span></span>
<span><span class="co">#&gt;  Mean   :0.06728   Mean   : 0.7127  </span></span>
<span><span class="co">#&gt;  3rd Qu.:0.06731   3rd Qu.: 0.7078  </span></span>
<span><span class="co">#&gt;  Max.   :0.11540   Max.   :10.0000</span></span></code></pre></div>
<p>Let’s look at the plots for the kriged results, and let’s avoid
including the background shape for now (i.e., no <code>bkg</code>
argument). Note that the overall shape of this kriged map is rectangular
because of the <code>lyr</code> raster shape.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, oma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Plot kriged map of pi</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/wingen/man/plot_gd.html" class="external-link">plot_gd</a></span><span class="op">(</span><span class="va">kgd</span>, main <span class="op">=</span> <span class="st">"Kriged pi"</span><span class="op">)</span></span>
<span><span class="co"># Plot kriged sample count map</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/wingen/man/plot_count.html" class="external-link">plot_count</a></span><span class="op">(</span><span class="va">kgd</span>, main <span class="op">=</span> <span class="st">"Kriged sample counts"</span><span class="op">)</span></span></code></pre></div>
<p><img src="wingen_vignette_files/figure-html/plot%20krige-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="mask-using-sample-counts">Mask using sample counts<a class="anchor" aria-label="anchor" href="#mask-using-sample-counts"></a>
</h4>
<p>Finally, we may want to mask out any areas that we’re not interested
in using the <code><a href="https://rdrr.io/pkg/wingen/man/mask_gd.html" class="external-link">mask_gd()</a></code> function. wingen provides several
ways to perform masking, but we will focus on two for this vignette. The
<code><a href="https://rdrr.io/pkg/wingen/man/mask_gd.html" class="external-link">mask_gd()</a></code> function is quite simple in that it takes in the
layer you want masked (<code>kgd</code>, in our case) and the layer you
want to mask to. Let’s first mask areas based on sample count. Recall
that above, we called <code><a href="https://rdrr.io/pkg/wingen/man/krig_gd.html" class="external-link">krig_gd()</a></code> using indices 1 and 2,
which told wingen to krige on both the genetic diversity (pi) and sample
count layers. We’ll make use of the kriged sample count layer now by
masking according to it. The <code>minval</code> argument specifies the
minimum value below which areas will be masked, meaning that any areas
that do not contain any samples will be masked. Let’s compare two maps
masked in this way with differing <code>minval</code> values.</p>
<p>As you can see, more areas are masked as we increase
<code>minval</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mgd_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/wingen/man/mask_gd.html" class="external-link">mask_gd</a></span><span class="op">(</span><span class="va">kgd</span>, <span class="va">kgd</span><span class="op">[[</span><span class="st">"sample_count"</span><span class="op">]</span><span class="op">]</span>, minval <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">mgd_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/wingen/man/mask_gd.html" class="external-link">mask_gd</a></span><span class="op">(</span><span class="va">kgd</span>, <span class="va">kgd</span><span class="op">[[</span><span class="st">"sample_count"</span><span class="op">]</span><span class="op">]</span>, minval <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, oma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/wingen/man/plot_gd.html" class="external-link">plot_gd</a></span><span class="op">(</span><span class="va">mgd_1</span>, main <span class="op">=</span> <span class="st">"Kriged &amp; masked pi"</span>, bkg <span class="op">=</span> <span class="va">envlayer</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/wingen/man/plot_gd.html" class="external-link">plot_gd</a></span><span class="op">(</span><span class="va">mgd_2</span>, main <span class="op">=</span> <span class="st">"Kriged &amp; masked pi"</span>, bkg <span class="op">=</span> <span class="va">envlayer</span><span class="op">)</span></span></code></pre></div>
<p><img src="wingen_vignette_files/figure-html/mask%20sample%20count-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="mask-using-state-boundaries">Mask using state boundaries<a class="anchor" aria-label="anchor" href="#mask-using-state-boundaries"></a>
</h4>
<p>Next, let’s mask out any areas outside of California state boundaries
using our <code>envlayer</code> object.</p>
<p>One thing to keep in mind is that if you are masking using a raster,
you need to ensure that pixels match between the object to be masked and
the object to use as a mask. If they aren’t the same, or if their
extents are different, you will either get an error or the map will look
wonky. To remedy this, we can use terra’s <code><a href="https://rdrr.io/pkg/terra/man/resample.html" class="external-link">resample()</a></code>
function. This function will resample a raster (here,
<code>envlayer</code>) based on an inputted raster (here,
<code>mgd_1</code>).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Resample envlayer based on masked layer</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/resample.html" class="external-link">resample</a></span><span class="op">(</span><span class="va">envlayer</span>, <span class="va">mgd_1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform masking</span></span>
<span><span class="va">mgd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/wingen/man/mask_gd.html" class="external-link">mask_gd</a></span><span class="op">(</span><span class="va">mgd_1</span>, <span class="va">r</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot masked map</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/wingen/man/plot_gd.html" class="external-link">plot_gd</a></span><span class="op">(</span><span class="va">mgd</span>, main <span class="op">=</span> <span class="st">"Kriged &amp; masked pi"</span>, bkg <span class="op">=</span> <span class="va">r</span><span class="op">)</span></span></code></pre></div>
<p><img src="wingen_vignette_files/figure-html/mask%20sd-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="running-wingen-with-wingen_do_everything">Running wingen with <code>wingen_do_everything()</code><a class="anchor" aria-label="anchor" href="#running-wingen-with-wingen_do_everything"></a>
</h3>
<hr>
<p>The algatr package also has an option to run all of the above
functionality in a single function, <code><a href="../reference/wingen_do_everything.html">wingen_do_everything()</a></code>.
If you want to take a look at the preview first, you can set
<code>preview = TRUE</code>; to krige, set <code>kriged = TRUE</code>,
and to mask, set <code>masked = TRUE</code>. If <code>preview</code> is
set to TRUE, the preview plot will be printed and a user can enter into
the console whether they would like to continue the wingen run with
existing parameter settings, or else halt the run. <strong>Please be
aware that the <code>do_everything()</code> functions are meant to be
exploratory. We do not recommend their use for final analyses unless
certain they are properly parameterized.</strong></p>
<p><strong>WARNING: Many wingen arguments are excluded from
<code><a href="../reference/wingen_do_everything.html">wingen_do_everything()</a></code> to reduce the number of arguments in
the function. Thus, we strongly advise against using this function for
actual analyses and encourage users to explore and test the individual
wingen functions given their datasets.</strong></p>
<p>Below, we’ll run wingen with the following:</p>
<ul>
<li><p>No preview generated before running wingen
(<code>preview = FALSE</code>)</p></li>
<li><p>Window size of 3x3 (<code>wdim = 3</code>), no aggregation factor
for moving window (<code>fact = 0</code>)</p></li>
<li><p>Pi as our genetic diversity statistic (<code>stat = "pi"</code>);
this is also the default</p></li>
<li><p>No rarefaction (<code>rarify = FALSE</code>)</p></li>
<li><p>Kriging (of both genetic diversity and sample count layers;
<code>kriged = TRUE</code> and <code>index = 1:2</code>) with a
disaggregation factor of 4 for kriging
(<code>disagg_grd = 4</code>)</p></li>
<li><p>Masking to California state boundaries
(<code>mask = envlayer</code>)</p></li>
<li><p>Plot kriged and masked sample count
(<code>plot_count = TRUE</code>)</p></li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wingen_do_everything.html">wingen_do_everything</a></span><span class="op">(</span></span>
<span>  gen <span class="op">=</span> <span class="va">liz_vcf</span>,</span>
<span>  lyr <span class="op">=</span> <span class="va">liz_lyr</span>,</span>
<span>  coords <span class="op">=</span> <span class="va">liz_coords</span>,</span>
<span>  wdim <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  fact <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  sample_count <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  preview <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  min_n <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  stat <span class="op">=</span> <span class="st">"pi"</span>,</span>
<span>  rarify <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  kriged <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  grd <span class="op">=</span> <span class="va">liz_lyr</span>,</span>
<span>  index <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,</span>
<span>  agg_grd <span class="op">=</span> <span class="cn">NULL</span>, disagg_grd <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  agg_r <span class="op">=</span> <span class="cn">NULL</span>, disagg_r <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  masked <span class="op">=</span> <span class="cn">TRUE</span>, mask <span class="op">=</span> <span class="va">envlayer</span>,</span>
<span>  bkg <span class="op">=</span> <span class="va">envlayer</span>, plot_count <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Please be aware: the do_everything functions are meant to be exploratory. We do not recommend their use for final analyses unless certain they are properly parameterized.</span></span>
<span><span class="co">#&gt; CRS do not match</span></span></code></pre></div>
<p><img src="wingen_vignette_files/figure-html/wingen%20do%20everything-1.png" width="700"><img src="wingen_vignette_files/figure-html/wingen%20do%20everything-2.png" width="700"><img src="wingen_vignette_files/figure-html/wingen%20do%20everything-3.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="additional-documentation-and-citations">Additional documentation and citations<a class="anchor" aria-label="anchor" href="#additional-documentation-and-citations"></a>
</h3>
<hr>
<table class="table">
<colgroup>
<col width="22%">
<col width="52%">
<col width="25%">
</colgroup>
<thead><tr class="header">
<th></th>
<th>Citation/URL</th>
<th>Details</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Manuscript with method</td>
<td><a href="https://doi.org/10.1111/2041-210X.14090" class="external-link">Bishop et
al. 2023</a></td>
<td>Paper describing wingen methodology and package</td>
</tr>
<tr class="even">
<td>R package documentation</td>
<td><a href="https://github.com/AnushaPB/wingen" class="external-link">Available on
Github</a></td>
<td>Code, vignette</td>
</tr>
</tbody>
</table>
<p>Retrieve wingen’s vignette:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># vignette("wingen-vignette")</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Anusha Bishop, Anne Chambers, Ian Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
